<html>
<head>
  <meta name="order" content="1" />
  <title>Data Description</title>
</head>

<body>
<b>
<a href="index.html">About Ensembl Variation</a> | 
Data Description |  
<a href="predicted_data.html">Predicted Data</a> |
<a href="database.html">Database Description</a> |  
<a href="/info/docs/api/variation/index.html">Perl API</a> |
<a href="vep/index.html">Variant Effect Predictor</a>
</b>

<div>
	<div style="float:right;margin-left:8px;margin-top:2px;background-color:#F2F2F2;color:#333;border-radius:5px">
    <div style="padding:5px;font-weight:bold;color:#FFF;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
		  <img src="/i/16/rev/info.png" style="vertical-align:top" />
		  Page content:
		</div> 
    <ul style="padding-left:25px;padding-right:5px">
      <li><a href="#sources">Species & Sources</a></li> 
      <li><a href="#displays">Displays</a></li> 
      <li><a href="#datatypes">Data types</a></li> 
      <li><a href="#classes">Variation classes</a></li>
      <li><a href="#indel">Insertion & Deletion coordinates</a></li>
      <li><a href="#populations">Populations</a></li>
      <li><a href="#variation_sets">Variation sets</a></li>
			<li><a href="#clin_significance">Clinical significance</a></li>
			<li><a href="#evidence_status">Evidence status</a></li>
      <li><a href="#quality_control">Quality control</a></li>
	  </ul>
  </div>

  <br />
  <h1>Ensembl Variation - Data description</h1>

  <p>
  Below is a description of the data we store in the databases for Ensembl Variation.<br />For several different species in Ensembl, 
  we import variation data (<span title="Single Nucleotide Polymorphisms">SNPs</span>, <span title="Copy Number Variations">CNVs</span>, allele frequencies, genotypes, etc) from a variety of 
  <a href="sources_documentation.html">sources</a> (e.g. dbSNP). Imported variants and alleles are subjected to a <a href="#quality_control">quality control</a> process to flag suspect data.
  <br />We classify the variants into different
  <a href="#classes">classes</a> and calculate the <a href="predicted_data.html">predicted consequence(s)</a>
	of the variant and we have also created variation <a href="#variation_sets">sets</a> to help people retrieve a specific
	group of variants from a particular dataset.<br />
	In human, we calculate the linkage disequilibrium for each variant, by <a href="#populations">population</a>.
  </p>
	
	<p>See some examples of imported data on the Ensembl website (Human):</p>
	<ul style="padding-left:0px;display:inline;list-style:none">
    <li style="display:inline">
		  <a href="/Homo_sapiens/Variation/Context?v=rs1333049"><img src="/i/96/var_genomic_context.png" class="portal" alt="Location of a variation in the genome" title="Location of a variation in the genome"/></a>
		</li>
		<li style="display:inline">
		  <a href="/Homo_sapiens/Variation/Population?v=rs699"><img src="/i/96/var_population_genetics.png" class="portal" alt="Population genotypes and frequencies of a variation" title="Population genotypes and frequencies of a variation"/></a>
		</li>
		<li style="display:inline">
		  <a href="/Homo_sapiens/Variation/Individual?v=rs699"><img src="/i/96/var_individual_genotypes.png" class="portal" alt="Individual genotypes of a variation" title="Individual genotypes of a variation"/></a>
	  </li>
		<li style="display:inline">
		  <a href="/Homo_sapiens/Variation/Populations?v=COSM191229"><img src="/i/96/var_sample_information.png" class="portal" alt="Sample information and frequencies of a somatic	variant" title="Sample information and frequencies of a somatic variant"/></a>
	  </li>
		<li style="display:inline">
		  <a href="/Homo_sapiens/Variation/Phenotype?v=rs1333049"><img src="/i/96/var_phenotype_data.png" class="portal" alt="Phenotype(s) associated with a variation" title="Phenotype(s) associated with a variation"/></a>
		</li>
		<li style="display:inline">
		  <a href="/Homo_sapiens/Variation/Sequence?v=rs699"><img src="/i/96/var_flanking_sequence.png" class="portal" alt="Genomic sequence flanking a variation" title="Genomic sequence flanking a variation"/></a>
		</li>
  </ul>
	<br /><br />
</div>
<div style="clear:both" />


<h2 id="sources"> Variation species and data sources </h2>

<p>
Ensembl stores variation data for the following species, but note that users can still use the 
<a href="vep/index.html">Variant Effect Predictor</a> on species for which we do not currently have 
a variation database.
</p>

<table><tr><td>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Bos_taurus.png" alt="Bos taurus" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Bos taurus</i> <a href="sources_documentation.html#bos_taurus">[sources]</a></span>
</div>
<div style="vertical-align:middle;margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Canis_familiaris.png" alt="Canis familiaris" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Canis familiaris</i> <a href="sources_documentation.html#canis_familiaris">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Danio_rerio.png" alt="Danio rerio" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Danio rerio</i> <a href="sources_documentation.html#danio_rerio">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Drosophila_melanogaster.png" alt="Drosophila melanogaster" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Drosophila melanogaster</i> <a href="sources_documentation.html#drosophila_melanogaster">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Equus_caballus.png" alt="Equus caballus" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Equus caballus</i> <a href="sources_documentation.html#equus_caballus">[sources]</a></span>
</div>
</td>
<td style="width:8px"></td>
<td>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Felis_catus.png" alt="Felis catus" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Felis catus</i> <a href="sources_documentation.html#felis_catus">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Gallus_gallus.png" alt="Gallus gallus" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Gallus gallus</i> <a href="sources_documentation.html#gallus_gallus">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Homo_sapiens.png" alt="Homo sapiens" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Homo sapiens</i> <a href="sources_documentation.html#homo_sapiens">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Macaca_mulatta.png" alt="Macaca mulatta" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Macaca mulatta</i> <a href="sources_documentation.html#macaca_mulatta">[sources]</a></span>
</div>	
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Monodelphis_domestica.png" alt="Monodelphis domestica" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Monodelphis domestica</i> <a href="sources_documentation.html#monodelphis_domestica">[sources]</a></span>
</div>
</td>
<td style="width:8px"></td>
<td>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Mus_musculus.png" alt="Mus musculus" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Mus musculus</i> <a href="sources_documentation.html#mus_musculus">[sources]</a></span>
</div>	
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Ornithorhynchus_anatinus.png" alt="Ornithorhynchus anatinus" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Ornithorhynchus anatinus</i> <a href="sources_documentation.html#ornithorhynchus_anatinus">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Pan_troglodytes.png" alt="Pan troglodytes" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Pan troglodytes</i> <a href="sources_documentation.html#pan_troglodytes">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Pongo_abelii.png" alt="Pongo abelii" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Pongo abelii</i> <a href="sources_documentation.html#pongo_abelii">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Rattus_norvegicus.png" alt="Rattus norvegicus" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Rattus norvegicus</i> <a href="sources_documentation.html#rattus_norvegicus">[sources]</a></span>
</div>
</td>
<td style="width:8px"></td>
<td>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Saccharomyces_cerevisiae.png" alt="Saccharomyces cerevisiae" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Saccharomyces cerevisiae</i> <a href="sources_documentation.html#saccharomyces_cerevisiae">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Sus_scrofa.png" alt="Sus scrofa" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Sus scrofa</i> <a href="sources_documentation.html#sus_scrofa">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Taeniopygia_guttata.png" alt="Taeniopygia guttata" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Taeniopygia guttata</i> <a href="sources_documentation.html#taeniopygia_guttata">[sources]</a></span>
</div>
<div style="margin-bottom:5px;padding-right:4px;overflow:hidden">
	<img src="/i/species/48/Tetraodon_nigroviridis.png" alt="Tetraodon nigroviridis" class="sp-thumb" style="float:none;margin-right:4px;vertical-align:middle" />
	<span><i>Tetraodon nigroviridis</i> <a href="sources_documentation.html#tetraodon_nigroviridis">[sources]</a></span>
</div>
</td></tr></table>

<p>
The majority of variants are imported from NCBI <a rel="external"
href="http://www.ncbi.nlm.nih.gov/projects/SNP/">dbSNP</a>. The data is
imported when it is released by dbSNP and incorporated into the next
Ensembl release. If dbSNP releases the data on a different assembly,
Ensembl will remap the variant positions onto the current assembly.
Data from projects like the <a rel="external" href="http://hapmap.ncbi.nlm.nih.gov/">HapMap Project</a> and <a
rel="external" href="http://www.1000genomes.org/">1000 Genomes Project</a> is
imported once it has been submitted to dbSNP.
</p>

<p>
Ensembl also includes data from other <a href="sources_documentation.html">sources</a>. To view data from these sources in the
browser go to a species Location page (e.g. <a href="/Homo_sapiens/Location/View?r=6:133017695-133161157">for human</a>), and click on
the <i>'Configure this page'</i> link on the left-hand side. The <i>'Variation'</i> and <i>'Somatic mutations'</i> sections contain a track list  of
all sources of variation data for that species.
</p>


<br />
<h2 id="displays">Variation displays</h2>

<p>
Variation data can be viewed in the browser through pages such as:
</p>

<ul>
<li><strong>Gene:</strong> Variation Table and Variation Image (for all variations in a gene) 
e.g. for example <a href="/Homo_sapiens/Gene/Variation_Gene/Table?db=core;g=ENSG00000159197">for all variants in KCNE2</a>. 
Structural Variation to see all structural variants overlapping the gene.</li>
<li><strong>Transcript:</strong> Population comparison, Comparison
image (for comparing variants in a transcript across different individual or strain
sequences) e.g. <a href="/Mus_musculus/Transcript/Population/Image?g=ENSMUSG00000041143;r=4:138528820-138586500;t=ENSMUST00000084208"> compare Tmco4 in different mouse strains </a> </li>
<li> <strong> Transcript: </strong> Sequence, protein: list of the coding variants in protein coordinates.</li>
<li><strong>Location:</strong> Region in Detail (Variations can be
drawn using &quot;Configure this page&quot; at the left. The menu allows
display of information in Ensembl databases along with external
sources in DAS format such as <a rel="external"
href="http://projects.tcag.ca/variation/">DGV loci</a>.)</li>
</ul>

<p>
Clicking on any variation on an Ensembl page will open a Variation tab
with information about the flanking sequence and source for the
selected variation. Links to linkage disequilibrium (LD) plots,
phenotype information (for human) from EGA, OMIM and <a rel="external"
href="http://www.genome.gov/">NHGRI</a> and Ensembl genes and
transcripts that include the variation can be found at the left of
this tab. You may also view multiple genome alignments of various
species, highlighting the
variation. <a href="/info/docs/compara/analyses.html#ancestral"
class="cp-external">Ancestral sequences</a> are included in this
display. 
</p>

<p>
Variation information can also be accessed
using <a href="/info/data/biomart.html"
class="cp-external">BioMart</a> (gene or variation database), and
the <a href="/info/docs/api/index.html" class="cp-external">Perl
API</a> (<a href="/info/docs/api/variation/index.html"
class="cp-external">variation API</a>).
</p>
</div>

<br />
<h2 id="datatypes">Data types</h2>
<p>
The Ensembl Variation database stores data imported from external sources and also data calculated on site.
</p>

<ul>
	<li><b>Data imported</b> from external sources (dbSNP, Sanger, DGVa, ...):
  	<ul>
  		<li>Variations <i>(SNPs, in-dels, insertion, deletion, ...)</i></li>
			<li>Structural variations <i>(copy number variation, tandem duplication, inversion, ...)</i></li>
			<li>Probes for copy number variations</li>
  		<li>Locations for variations and structural variations</li>
  		<li>Alleles</li>
  		<li>Populations</li>
  		<li>Genotypes</li>
			<li>Phenotypes</li>
  	</ul>
	</li>
	<li><b>Calculated data</b>: see the <a href="predicted_data.html">Predicted data page</a>.</li>
</ul>


<br />

<h2 id="classes">Variation classes</h2>

<p>
We call the class of a variation according to its component alleles and its mapping 
to the reference genome, and then display this information on the website. Internally 
we use <a href="http://www.sequenceontology.org/">Sequence Ontology</a> terms, but we 
map these to our own 'display' terms where common usage differs from the SO 
definition (e.g. our term SNP is closer to the SO term SNV). All the classes we call, 
along with their equivalent SO term are shown in the table below. We also differentiate 
somatic mutations from germline variations in the display term, prefixing the term with
'somatic'. API users can fetch either the SO term or the display term.
</p>

<br />


<table id="variation_classes" class="ss">
  <tr>
    <th style="width:8px;padding-left:0px;padding-right:0px;text-align:center">*</th>
    <th>SO term</th>
    <th>SO description</th>
    <th>SO accession</th>
    <th>Ensembl term</th>
    <th>Called for</th>
  </tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px"></td>
		<td rowspan="2">SNV</td>
		<td rowspan="2">SNVs are single nucleotide positions in genomic DNA at which different sequence alternatives exist.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001483">SO:0001483</a></td>
		<td>SNP</td>
		<td rowspan="2">Variation</td>
	</tr>
	<tr>
		<td>somatic_SNV</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px"></td>
		<td rowspan="2">indel</td>
		<td rowspan="2">A sequence alteration which included an insertion and a deletion, affecting 2 or more bases.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:1000032">SO:1000032</a></td>
		<td>indel</td>
		<td rowspan="2">Variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_indel</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px"></td>
		<td rowspan="2">substitution</td>
		<td rowspan="2">A sequence alteration where the length of the change in the variant is the same as that of the reference.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:1000002">SO:1000002</a></td>
		<td>substitution</td>
		<td rowspan="2">Variation</td>
	</tr>
	<tr>
		<td>somatic_substitution</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px"></td>
		<td rowspan="2">tandem_repeat</td>
		<td rowspan="2">Two or more adjcent copies of a region (of length greater than 1).</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000705">SO:0000705</a></td>
		<td>tandem_repeat</td>
		<td rowspan="2">Variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_tandem_repeat</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#99CCFF;border-top:1px solid #FFF"></td>
		<td rowspan="2">complex_structural_alteration</td>
		<td rowspan="2">A structural sequence alteration or rearrangement encompassing one or more genome fragments.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001784">SO:0001784</a></td>
		<td>Complex</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr>
		<td>somatic_Complex</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#0000CC;border-top:1px solid #FFF"></td>
		<td rowspan="2">copy_number_gain</td>
		<td rowspan="2">A sequence alteration whereby the copy number of a given regions is greater than the reference sequence.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001742">SO:0001742</a></td>
		<td>Gain</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_Gain</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#CC0000;border-top:1px solid #FFF"></td>
		<td rowspan="2">copy_number_loss</td>
		<td rowspan="2">A sequence alteration whereby the copy number of a given region is less than the reference sequence.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001743">SO:0001743</a></td>
		<td>Loss</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr>
		<td>somatic_Loss</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#000000;border-top:1px solid #FFF"></td>
		<td rowspan="2">copy_number_variation</td>
		<td rowspan="2">A variation that increases or decreases the copy number of a given region.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001019">SO:0001019</a></td>
		<td>CNV</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_CNV</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#000000;border-top:1px solid #FFF"></td>
		<td rowspan="2">duplication</td>
		<td rowspan="2">One or more nucleotides are added between two adjacent nucleotides in the sequence; the inserted sequence derives from, or is identical in sequence to, nucleotides adjacent to insertion point.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:1000035">SO:1000035</a></td>
		<td>Duplication</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr>
		<td>somatic_Duplication</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#B2B2B2;border-top:1px solid #FFF"></td>
		<td rowspan="2">interchromosomal_breakpoint</td>
		<td rowspan="2">A rearrangement breakpoint between two different chromosomes.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001873">SO:0001873</a></td>
		<td>Interchromosomal breakpoint</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_Interchromosomal breakpoint</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#B2B2B2;border-top:1px solid #FFF"></td>
		<td rowspan="2">intrachromosomal_breakpoint</td>
		<td rowspan="2">A rearrangement breakpoint within the same chromosome.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001874">SO:0001874</a></td>
		<td>Intrachromosomal breakpoint</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr>
		<td>somatic_Intrachromosomal breakpoint</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#9933FF;border-top:1px solid #FFF"></td>
		<td rowspan="2">inversion</td>
		<td rowspan="2">A continuous nucleotide sequence is inverted in the same position.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:1000036">SO:1000036</a></td>
		<td>inversion</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_inversion</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#732E00;border-top:1px solid #FFF"></td>
		<td rowspan="2">tandem_duplication</td>
		<td rowspan="2">A duplication consisting of 2 identical adjacent regions.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:1000173">SO:1000173</a></td>
		<td>Tandem duplication</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr>
		<td>somatic_Tandem duplication</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#C3A4FF;border-top:1px solid #FFF"></td>
		<td rowspan="2">translocation</td>
		<td rowspan="2">A region of nucleotide sequence that has translocated to a new position.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000199">SO:0000199</a></td>
		<td>translocation</td>
		<td rowspan="2">Structural variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_translocation</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#CC0000;border-top:1px solid #FFF"></td>
		<td rowspan="2">deletion</td>
		<td rowspan="2">The point at which one or more contiguous nucleotides were excised.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000159">SO:0000159</a></td>
		<td>deletion</td>
		<td rowspan="2">Variation<br />Structural variation</td>
	</tr>
	<tr>
		<td>somatic_deletion</td>
	</tr>

	<tr class="bg2">
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#FFCC00;border-top:1px solid #FFF"></td>
		<td rowspan="2">insertion</td>
		<td rowspan="2">The sequence of one or more nucleotides added between two adjacent nucleotides in the sequence.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000667">SO:0000667</a></td>
		<td>insertion</td>
		<td rowspan="2">Variation<br />Structural variation</td>
	</tr>
	<tr class="bg2">
		<td>somatic_insertion</td>
	</tr>

	<tr>
		<td rowspan="2" style="padding:0px;margin:0px;background-color:#B2B2B2;border-top:1px solid #FFF"></td>
		<td rowspan="2">sequence_alteration</td>
		<td rowspan="2">A sequence_alteration is a sequence_feature whose extent is the deviation from another sequence.</td>
		<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001059">SO:0001059</a></td>
		<td>sequence_alteration</td>
		<td rowspan="2">Variation<br />Structural variation</td>
	</tr>
	<tr>
		<td>somatic_sequence_alteration</td>
	</tr>

	<tr class="bg2">
		<td style="padding:0px;margin:0px;background-color:#4682B4;border-top:1px solid #FFF"></td>
		<td>probe</td>
		<td>A DNA sequence used experimentally to detect the presence or absence of a complementary nucleic acid.</td>
		<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000051">SO:0000051</a></td>
		<td>CNV_PROBE</td>
		<td>CNV probe</td>
	</tr></table>

<p>
<b>*</b> Corresponding colours for the Ensembl web displays (only for Structural variations). 
The colours are based on the <a rel="external" href="http://www.ncbi.nlm.nih.gov/dbvar/content/overview/">dbVar</a> displays.
<p>

<div class="js_panel" id="pie_chart_panel" style="margin-top:15px;margin-bottom:20px">
	<input class="panel_type" type="hidden" value="Piechart">
	<input class="graph_config" type="hidden" name="legendpos" value="'east'" />
	<input class="graph_dimensions" type="hidden" value="[90,80,80]" />
	
	<input type="hidden" class="graph_data" value="[[48416035,'SNV'],[3248758,'insertion'],[2542201,'deletion'],[559790,'sequence_alteration'],[395149,'indel'],[84469,'substitution'],[4561,'tandem_repeat']]" />

	<div class="pie_chart_classes" title="classes" style="width:400px;height:220px;border:1px solid #AAA;border-radius:10px;-moz-border-radius:10px;-webkit-border-radius:10px;-webkit-box-shadow:0 1px 3px #666;">
   	<h3 style="padding:10px 10px 5px">Human Variation class distribution - Ensembl 71</h4>
   	<div id="graphHolder0"></div>
	</div>
</div>
	

<h3 id="indel">Insertion and Deletion coordinates </h3>

<p> In Ensembl, an insertion is indicated by start coordinate = end coordinate +
1. For example, an insertion of 'C' between nucleotides 12600 and
12601 on the forward strand is indicated with start and end coordinates as
follows:</p>

<pre>
   12601     12600   
</pre>
<p>A deletion is indicated by the exact nucleotide coordinates. For
example, a three base pair deletion of nucleotides 12600, 12601, and
12602 of the reverse strand will have start and end coordinates of :</p>

<pre>
   12600     12602    
</pre>



<br />

<h2 id="populations">Populations</h2>
<p>
<p>For each variant we store population data (allele
frequencies, genotypes) available from different projects (e.g. 1000
Genomes, HapMap). You can see this on our variation pages, for
example: <a href="/Homo_sapiens/Variation/Population?v=rs1333049">rs1333049</a>.<br />
Below we list the population names associated with the 1000 Genomes Project and the HapMap Project:
</p>

<h4>Populations from the 1000 Genomes Project (Human)</h4>
<table id="1000_genomes" class="ss">

	<tr>
		<th style="width:200px">Name</th>
		<th>Size</th>
		<th>Description</th>
	</tr>

	<tr>
		<td>1000GENOMES:phase_1_ALL</td>
		<td>1092</td>
		<td>All individuals from phase 1 of the 1000 Genomes Project</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_AFR</td>
		<td>246</td>
		<td>All African individuals from phase 1 of the 1000 Genomes Project (YRI, LWK, ASW)</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_AMR</td>
		<td>181</td>
		<td>All American individuals from phase 1 of the 1000 Genomes Project</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_ASN</td>
		<td>286</td>
		<td>All East Asian individuals from phase 1 of the 1000 Genomes Project (CHB, JPT, CHS)</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_ASW</td>
		<td>61</td>
		<td>Americans of African Ancestry in SW USA</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_CEU</td>
		<td>85</td>
		<td>Utah Residents (CEPH) with Northern and Western European ancestry</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_CHB</td>
		<td>97</td>
		<td>Han Chinese in Bejing, China</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_CHS</td>
		<td>100</td>
		<td>Southern Han Chinese</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_CLM</td>
		<td>60</td>
		<td>Colombian from Medellian, Colombia</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_EUR</td>
		<td>379</td>
		<td>All European individuals from phase 1 of the 1000 Genomes Project (CEU, TSI, FIN, GBR, IBS)</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_FIN</td>
		<td>93</td>
		<td>Finnish in Finland</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_GBR</td>
		<td>89</td>
		<td>British in England and Scotland</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_IBS</td>
		<td>14</td>
		<td>Iberian population in Spain</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_JPT</td>
		<td>89</td>
		<td>Japanese in Tokyo, Japan</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_LWK</td>
		<td>97</td>
		<td>Luhya in Webuye, Kenya</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_MXL</td>
		<td>66</td>
		<td>Mexican Ancestry from Los Angeles USA</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_PUR</td>
		<td>55</td>
		<td>Puerto Ricans from Puerto Rica</td>
	</tr>
	<tr class="bg2">
		<td>1000GENOMES:phase_1_TSI</td>
		<td>98</td>
		<td>Toscani in Italy</td>
	</tr>
	<tr>
		<td>1000GENOMES:phase_1_YRI</td>
		<td>88</td>
		<td>Yoruba in Ibadan, Nigera</td>
	</tr>
</table>

<h4>Populations from the HapMap Project (Human)</h4>
<table id="hapmap" class="ss">

	<tr>
		<th style="width:200px">Name</th>
		<th>Size</th>
		<th>Description</th>
	</tr>

	<tr>
		<td>CSHL-HAPMAP:HAPMAP-ASW</td>
		<td>90</td>
		<td>African ancestry in Southwest USA. ASW is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr class="bg2">
		<td>CSHL-HAPMAP:HapMap-CEU</td>
		<td>180</td>
		<td>Utah residents with Northern and Western European ancestry from the CEPH collection. CEU is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr>
		<td>CSHL-HAPMAP:HAPMAP-CHB</td>
		<td>90</td>
		<td>Han Chinese in Beijing, China. CHB is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr class="bg2">
		<td>CSHL-HAPMAP:HAPMAP-CHD</td>
		<td>100</td>
		<td>Chinese in Metropolitan Denver, Colorado. CHD is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr>
		<td>CSHL-HAPMAP:HAPMAP-GIH</td>
		<td>100</td>
		<td>Gujarati Indians in Houston, Texas. GIH is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr class="bg2">
		<td>CSHL-HAPMAP:HapMap-HCB</td>
		<td>45</td>
		<td>45 unrelated Han Chinese in Beijing, China, representing one of the populations studied in the International HapMap project (<a href="http://www.hapmap.org"> http://www.hapmap.org</a>). See <a href="http://www.hapmap.org/citinghapmap.html.en">http://www.hapmap.org/citinghapmap.html.en</a> for further information about,this population and others studied in the project. <a href="http://www.hapmap.org/hapmappopulations.html.en">http://www.hapmap.org/hapmappopulations.html.en</a>,also has relevant information.</td>
	</tr>
	<tr>
		<td>CSHL-HAPMAP:HapMap-JPT</td>
		<td>91</td>
		<td>Japanese in Tokyo, Japan. JPT is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr class="bg2">
		<td>CSHL-HAPMAP:HAPMAP-LWK</td>
		<td>100</td>
		<td>Luhya in Webuye, Kenya. LWK is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr>
		<td>CSHL-HAPMAP:HAPMAP-MEX</td>
		<td>90</td>
		<td>Mexican ancestry in Los Angeles, California. MEX is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr class="bg2">
		<td>CSHL-HAPMAP:HAPMAP-MKK</td>
		<td>180</td>
		<td>Maasai in Kinyawa, Kenya. MKK is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr>
		<td>CSHL-HAPMAP:HAPMAP-TSI</td>
		<td>100</td>
		<td>Toscans in Italy. TSI is one of the 11 populations in HapMap phase 3.</td>
	</tr>
	<tr class="bg2">
		<td>CSHL-HAPMAP:HapMap-YRI</td>
		<td>180</td>
		<td>Yoruba in Ibadan, Nigeria. YRI is one of the 11 populations in HapMap phase 3.</td>
	</tr>
</table>

<p>This population data is stored in the sample table of the Variation database. You can see a description of the sample
table <a href="variation_schema.html#sample">here</a>.
</p>


<br />
<h2 id="variation_sets">Variation sets</h2>
    
<p>
We use the concept of variation sets to group variations that share some property together. For example, we have grouped the variations identified in the three different
1000 Genomes pilot studies into separate variation sets. The sets can be further subdivided into supersets and subsets to reflect hierarchical relationships between them. In the case
of the 1000 Genomes phase 1 sets, these are divided into subsets based on population. For example, the set representing variations identified in the 1000 Genomes phase 1 study is named 
'1000 Genomes - All' and has several subsets like: '1000 Genomes - AFR', '1000 Genomes - AMR', '1000 Genomes - ASN' and '1000 Genomes - EUR'. 
The variation sets can be displayed as separate tracks on the location view. This behaviour is controlled from the 'Variation' section on the configuration panel which is accessed by 
clicking the 'Configure this page' link in the left hand side navigation.
</p>

<p>
The sets are constructed during production and are stored in the database. The table below lists the available
variation sets in the Ensembl variation database (subsets are indicated by bullet points).
</p>

<h4>Variation sets common to all species</h4>
<table id="variation_set_table" class="ss">

	<tr>
		<th>Name</th>
		<th>Short name</th>
		<th>Description</th>
	</tr>

	<tr>
		<td>All failed variations</td>
		<td>fail_all</td>
		<td>Variations that have failed the Ensembl QC checks</td>
	</tr>
</table>
<br />
<h4>Variation sets specific to Human</h4>
<table id="human_variation_set_table" class="ss">

	<tr>
		<th>Name</th>
		<th>Short name</th>
		<th>Description</th>
	</tr>
	<tr>
		<td>1000 Genomes - All</td>
		<td>1kg</td>
		<td>Variants genotyped by the 1000 Genomes project (phase 1)</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - AFR</li></ul></td>
		<td>1kg_afr</td>
		<td>Variants genotyped in African individuals by the 1000 Genomes project (phase 1)</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - AFR - common</li></ul></td>
		<td>1kg_afr_com</td>
		<td>Variants genotyped in African individuals by the 1000 Genomes project (phase 1) with frequency of at least 1%</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - AMR</li></ul></td>
		<td>1kg_amr</td>
		<td>Variants genotyped in admixed American individuals by the 1000 Genomes project (phase 1)</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - AMR - common</li></ul></td>
		<td>1kg_amr_com</td>
		<td>Variants genotyped in admixed American individuals by the 1000 Genomes project (phase 1) with frequency of at least 1%</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - ASN</li></ul></td>
		<td>1kg_asn</td>
		<td>Variants genotyped in East Asian individuals by the 1000 Genomes project (phase 1)</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - ASN - common</li></ul></td>
		<td>1kg_asn_com</td>
		<td>Variants genotyped in East Asian individuals by the 1000 Genomes project (phase 1) with frequency of at least 1%</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - All - common</li></ul></td>
		<td>1kg_com</td>
		<td>Variants genotyped by the 1000 Genomes project (phase 1) with frequency of at least 1%</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - EUR</li></ul></td>
		<td>1kg_eur</td>
		<td>Variants genotyped in European individuals by the 1000 Genomes project (phase 1)</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - EUR - common</li></ul></td>
		<td>1kg_eur_com</td>
		<td>Variants genotyped in European individuals by the 1000 Genomes project (phase 1) with frequency of at least 1%</td>
	</tr>
	<tr>
		<td>1000 Genomes - High coverage - Trios</td>
		<td>1kg_hct</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data from two family trios (Pilot 2)</td>
	</tr>
	<tr class="bg2">
		<td>1000 Genomes - Low coverage</td>
		<td>1kg_lc</td>
		<td>Variations called by the 1000 Genomes project on low coverage sequence data from 179 unrelated individuals (Pilot 1)</td>
	</tr>
	<tr>
		<td>Affy GeneChip 500K</td>
		<td>Affy_500K</td>
		<td>Variants from the Affymetrix GeneChip Human Mapping 500K Array Set</td>
	</tr>
	<tr class="bg2">
		<td>Affy GenomeWideSNP_6.0</td>
		<td>Affy_SNP6</td>
		<td>Variants from the Affymetrix Genome-Wide Human SNP Array 6.0</td>
	</tr>
	<tr>
		<td>Illumina_Cardio-Metabo_Chip</td>
		<td>Cardio-Metabo_Chip</td>
		<td>Variants from the Illumina Cardio-Metabo_Chip genotyping array designed to target variants of interest for metabolic and cardiovascular disease traits</td>
	</tr>
	<tr class="bg2">
		<td>Illumina_Human610_Quad</td>
		<td>Human610_Quad</td>
		<td>Variants from the Illumina Human610_Quad v1_B whole genome genotyping array designed for association studies</td>
	</tr>
	<tr>
		<td>Illumina_HumanHap550</td>
		<td>HumanHap550</td>
		<td>Variants from the Illumina Human550 v3.0 whole genome genotyping array designed for association studies</td>
	</tr>
	<tr class="bg2">
		<td>Illumina_HumanHap650Y</td>
		<td>HumanHap650Y</td>
		<td>Variants from the Illumina HumanHap650Y v3.0 whole genome genotyping array designed for association studies</td>
	</tr>
	<tr>
		<td>Illumina_HumanOmni1-Quad</td>
		<td>HumanOmni1-Quad</td>
		<td>Variants from the Illumina HumanOmni1-Quad whole genome genotyping array designed for association studies</td>
	</tr>
	<tr class="bg2">
		<td>Illumina_HumanOmni2.5</td>
		<td>HumanOmni2.5</td>
		<td>Variants from the Illumina HumanOmni2.5 4v1 whole genome genotyping array designed for association studies</td>
	</tr>
	<tr>
		<td>Illumina_1M-duo</td>
		<td>Illumina_1M-duo</td>
		<td>Variants from the Illumina Human1M-duo v3 whole genome genotyping array designed for association studies</td>
	</tr>
	<tr class="bg2">
		<td>Illumina_Human660W-quad</td>
		<td>Illumina_660Q</td>
		<td>Variants from the Illumina Human660W-quad whole genome genotyping array designed for association studies</td>
	</tr>
	<tr>
		<td>Illumina_CytoSNP12v1</td>
		<td>Illumina_CytoSNP12v1</td>
		<td>Variants from the Illumina Cyto SNP-12 v1 whole genome SNP genotyping chip designed for cytogenetic analysis</td>
	</tr>
	<tr class="bg2">
		<td>ESP_6500</td>
		<td>esp_6500</td>
		<td>Variants from the NHLBI Exome Sequencing Project</td>
	</tr>
	<tr>
		<td>All HapMap</td>
		<td>hapmap</td>
		<td>Variations which have been assayed by The International HapMap Project [http://hapmap.ncbi.nlm.nih.gov/]</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">HapMap - CEU</li></ul></td>
		<td>hapmap_ceu</td>
		<td>Variations which have been assayed by The International HapMap Project from CEU individuals</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">HapMap - HCB</li></ul></td>
		<td>hapmap_hcb</td>
		<td>Variations which have been assayed by The International HapMap Project from HCB individuals</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">HapMap - JPT</li></ul></td>
		<td>hapmap_jpt</td>
		<td>Variations which have been assayed by The International HapMap Project from JPT individuals</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">HapMap - YRI</li></ul></td>
		<td>hapmap_yri</td>
		<td>Variations which have been assayed by The International HapMap Project from YRI individuals</td>
	</tr>
	<tr class="bg2">
		<td>Anonymous Korean</td>
		<td>ind_ak1</td>
		<td>Variants genotyped in an anonymous Korean individual</td>
	</tr>
	<tr>
		<td>Misha Angrist</td>
		<td>ind_angrist</td>
		<td>Variants genotyped in Misha Angrist</td>
	</tr>
	<tr class="bg2">
		<td>Henry Louis Gates Jr</td>
		<td>ind_gates_jr</td>
		<td>Variants genotyped in Henry Louis Gates Jr</td>
	</tr>
	<tr>
		<td>Henry Louis Gates Sr</td>
		<td>ind_gates_sr</td>
		<td>Variants genotyped in Henry Louis Gates Sr</td>
	</tr>
	<tr class="bg2">
		<td>Rosalynn Gill</td>
		<td>ind_gill</td>
		<td>Variants genotyped in Rosalynn Gill</td>
	</tr>
	<tr>
		<td>Anonymous Irish Male</td>
		<td>ind_irish</td>
		<td>Variants genotyped in an anonymous Irish Male</td>
	</tr>
	<tr class="bg2">
		<td>Marjolein Kriek</td>
		<td>ind_kriek</td>
		<td>Variants genotyped in Marjolein Kriek</td>
	</tr>
	<tr>
		<td>Stephen Quake</td>
		<td>ind_quake</td>
		<td>Variants genotyped in Stephen Quake</td>
	</tr>
	<tr class="bg2">
		<td>Saqqaq</td>
		<td>ind_saqqaq</td>
		<td>Variants genotyped in a Palaeo-Eskimo Saqqaq individual</td>
	</tr>
	<tr>
		<td>Saqqaq HC</td>
		<td>ind_saqqaq_hc</td>
		<td>Variants genotyped in a Palaeo-Eskimo Saqqaq individual (high confidence SNPs)</td>
	</tr>
	<tr class="bg2">
		<td>Seong-Jin Kim</td>
		<td>ind_sjk</td>
		<td>Variants genotyped in Seong-Jin Kim</td>
	</tr>
	<tr>
		<td>ENSEMBL:Venter</td>
		<td>ind_venter</td>
		<td>Variants genotyped in Craig Venter</td>
	</tr>
	<tr class="bg2">
		<td>ENSEMBL:Watson</td>
		<td>ind_watson</td>
		<td>Variants genotyped in James Watson</td>
	</tr>
	<tr>
		<td>YanHang</td>
		<td>ind_yh</td>
		<td>Variants genotyped in a Han Chinese individual (YanHuang Project)</td>
	</tr>
	<tr class="bg2">
		<td>All phenotype-associated variants</td>
		<td>ph_variants</td>
		<td>Variants that have been associated with a phenotype</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">COSMIC phenotype variants</li></ul></td>
		<td>ph_cosmic</td>
		<td>Phenotype annotations of somatic mutations found in human cancers from the COSMIC project</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Clinical variations from dbSNP</li></ul></td>
		<td>precious</td>
		<td>Variations that belong to a reserved or "precious" set of clinically associated SNPs from dbSNP [http://www.ncbi.nlm.nih.gov/projects/SNP/]</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">HGMD-PUBLIC variants</li></ul></td>
		<td>ph_hgmd_pub</td>
		<td>Variants annotated by HGMD</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Johnson & O'Donnell phenotype variants</li></ul></td>
		<td>ph_johnson_et_al</td>
		<td>Johnson & O'Donnell 'An Open Access Database of Genome-wide Association Results' PMID:19161620</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">NHGRI catalog phenotype variants</li></ul></td>
		<td>ph_nhgri</td>
		<td>Variants associated with phenotype data from the NHGRI GWAS catalog [http://www.genome.gov/gwastudies/]</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">OMIM phenotype variants</li></ul></td>
		<td>ph_omim</td>
		<td>Variations linked to entries in the Online Mendelian Inheritance in Man (OMIM) database</td>
	</tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Uniprot phenotype variants</li></ul></td>
		<td>ph_uniprot</td>
		<td>Variations with phenotype annotations provided by Uniprot</td>
	</tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">clinically associated</li></ul></td>
		<td>clin_assoc</td>
		<td>Variants described by ClinVar as being probable-pathogenic, pathogenic, drug-response or histocompatibility</td>
	</tr>
</table>




<br />
<h2 id="clin_significance">Clinical significance</h2>

<div id="dbsnp_clin_sig" style="float:left;margin-right:100px;">
  <b><span style="color:#333">dbSNP</span> clinical significance</b><br />
    <p>The clinical significance of a variant as reported by dbSNP.</p>

	<table class="ss" style="width:auto">
       <tr><th>Value</th><th>Example</th></tr>
    <tr><td>drug-response</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs671">rs671</a></td></tr>
    <tr class="bg2"><td>histocompatibility</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs1093">rs1093</a></td></tr>
    <tr><td>non-pathogenic</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs328">rs328</a></td></tr>
    <tr class="bg2"><td>other</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs662">rs662</a></td></tr>
    <tr><td>pathogenic</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs268">rs268</a></td></tr>
    <tr class="bg2"><td>probable-non-pathogenic</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs140598">rs140598</a></td></tr>
    <tr><td>probable-pathogenic</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs5935">rs5935</a></td></tr>
    <tr class="bg2"><td>unknown</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs721710">rs721710</a></td></tr>
    <tr><td>untested</td><td><a href="/Homo_sapiens/Variation/Explore?v=rs4601">rs4601</a></td></tr>

  </table>
</div>
<div id="dgva_clin_sig" style="float:left;margin-right:100px;">
  <b><span style="color:#333">DGVa</span> clinical significance</b><br />
    <p>The clinical significance of a structural variant as reported by DGVa.</p>

	<table class="ss" style="width:auto">
       <tr><th>Value</th><th>Example</th></tr>
    <tr class="bg2"><td>Benign</td><td><a href="/Homo_sapiens/StructuralVariation/Evidence?sv=nsv491544">nsv491544</a></td></tr>
    <tr><td>Not tested</td><td></td></tr>
    <tr class="bg2"><td>Pathogenic</td><td><a href="/Homo_sapiens/StructuralVariation/Evidence?sv=nsv483167">nsv483167</a></td></tr>
    <tr><td>Uncertain Significance</td><td><a href="/Homo_sapiens/StructuralVariation/Evidence?sv=nsv918063">nsv918063</a></td></tr>
    <tr class="bg2"><td>Uncertain Significance: likely benign</td><td><a href="/Homo_sapiens/StructuralVariation/Evidence?sv=nsv917698">nsv917698</a></td></tr>
    <tr><td>Uncertain Significance: likely pathogenic</td><td><a href="/Homo_sapiens/StructuralVariation/Evidence?sv=esv1791726">esv1791726</a></td></tr>

  </table>
</div>
<div style="clear:both" />


<br />
<h2 id="evidence_status">Evidence status </h2>
<p>
We provide a simple summary of the evidence supporting a variant as a guide to its potential reliability
</p>
<table id="evidence_status_table" class="ss" style="width:850px">
    <tr>
        <th>Name</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>Multiple_observations</td>
        <td>The variant has multiple independent dbSNP submissions, i.e. submissions with a different submitter handles or different discovery samples </td>
		</tr>
    <tr class="bg2">
        <td>Frequency </td>
        <td>The variant is reported to be polymorphic in at least one sample </td>

    </tr>
    <tr>
        <td>HapMap</td>
        <td>The variant is polymorphic in at least one HapMap panel (human only)
</td>

    </tr>
    <tr class="bg2">
        <td>1000 Genomes</td>
        <td>The variant was discovered in the 1000 genomes project (human only)</td>
    </tr>
        <tr>
        <td>Cited </td>
        <td>dbSNP holds a citation from PubMed for the variant.</td>

    </tr>
</table>


<br />
<h2 id="quality_control">Quality control </h2>

<p>
A quality control process is employed to check imported variation data. Suspect variations and alleles are flagged, but are not withheld from downstream annotation. Data failing the checks is available through the browser where failure reasons are prominently listed. The API does not extract failed data by default, unless the database adaptor is specifically configured to do so using 
<a href ="http://www.ensembl.org/info/docs/Doxygen/variation-api/classBio_1_1EnsEMBL_1_1Variation_1_1DBSQL_1_1DBAdaptor.html#add8230ebca14116047cf7d17499f9e3b">Bio::EnsEMBL::Variation::DBSQL::DBAdaptor::include_failed_variations()</a>;
<p>
 Variations for which dbSNP holds citations from PubMed are not submitted to the QC process so are not flagged as failed.
</p>

<h4>Failure reasons</h4>
<table id="quality_control_table" class="ss">

<tr class="bg2">
<th>QC Type</th>
<th>Reported failure reason</th>
<th>Checking process</th>
</tr>

<tr >
<td  rowspan="4" class="bg2"><b>Mapping checks</b></td>
<td >Variation does not map to the genome </td>
<td > Variations with flanking sequences which do not map to reference or non-reference genomic sequences are flagged as failed.</td>
</tr>

<tr>
<td class="bg2">Variation maps to more than 1 location</td>
<td class="bg2"> For variations with flanking sequences mapping to a reference sequence, the number of mappings within all reference sequences is counted and those mapping more than once are flagged as failed. (Variants with a single mapping to both X and Y within a PAR region are not failed.)
For variations with flanking sequences which do not map to a reference sequence, the number of mappings within all non-reference sequences is counted and those mapping more than once are flagged as failed. 
</td>
</tr>

<tr>
<td >Mapped position is not compatible with reported alleles </td>
<td >The length of the reported alleles is compared to that expected given the coordinates specified for the variation. If none of the alleles match the expected length, the variation is flagged as failed. </td>
</tr>
<tr>
<td class="bg2">None of the variant alleles match the reference allele </td>
<td class="bg2"> The sequence at the coordinates specified for the variation are extracted from the reference genome and compared to the dbSNP refSNP alleles. If the extracted sequence does not match the expected alleles, the variation is flagged as failed.</td>
</tr>

<tr>
<td  rowspan="4" class="bg2"><b>Checks on the alleles of refSNPs</b></td>
<td>Loci with no observed variant alleles in dbSNP   </td>
<td>Variations with dbSNP refSNP alleles reported as 'NOVARIATION' are flagged as failed.</td>
</tr>
<tr>
<td class="bg2">Variation has more than 3 different alleles </td>
<td class="bg2">Variations with all of A, T, G and C in the dbSNP refSNP alleles are flagged as failed.</td>
</tr>
<tr>
<td>Alleles contain ambiguity codes  </td>
<td>Variations with a IUPAC ambiguity code (eg.  M, Y, R, etc ) in the dbSNP refSNP alleles are reported as failed. </td>
</tr>
<tr>
<td class="bg2">Alleles contain non-nucleotide characters  </td>
<td class="bg2">Variations with unexpected characters in the dbSNP refSNP alleles are reported as failed.</td>
</tr>


<tr>
<td  rowspan="1" class="bg2"><b>Checks on the alleles in dbSNP submissions</b></td>
<td> Additional submitted allele data from dbSNP does not agree with the dbSNP refSNP alleles </td>
<td> Alleles from dbSNP submissions ( primarily frequency submissions, but also variant discovery submissions which have been merged in the dbSNP pipeline with the pre-existing refSNP variation) are checked against the dbSNP refSNP alleles and discrepant sets flagged as failed.  Often this will highlight a strand error in the submission of frequency information for a known variation. Tradditionally, the failure would be flagged at the variation level, in new database builds it is flagged at the allele submission level.
</td>
</tr>


<tr>
<td  rowspan="1" class="bg2"><b>External failure classification</b></td>
<td class="bg2"> Flagged as suspect by dbSNP </td>
<td class="bg2">Variations reported as being suspect by dbSNP due to being in probable paralogous regions are imported but flagged as failed (human only).</td>
</tr>
</table>
<br />



</body>
</html>
