<html>
<head>
<meta http-equiv="CONTENT-TYPE" content="text/html; charset=utf-8" />
<title>Variation Tables Description</title>

<script language="Javascript" type="text/javascript">
	// Function to show/hide the columns table
	function show_hide (param) {
		div   = document.getElementById('div_'+param);
		alink = document.getElementById('a_'+param);
		if (div.style.display=='inline') {
			div.style.display='none';
			alink.innerHTML='Show';
		}
		else {
			if (div.style.display=='none') {
				div.style.display='inline';
				alink.innerHTML='Hide';
			}
		}
	}
	
	// Function to show/hide all the tables
	function show_hide_all () {
		expand_flag = document.getElementById('expand');
		divs = document.getElementsByTagName('div');

		for(var i=0; i < divs.length; i++){
			param = divs[i].id.substring(4);
			div   = document.getElementById('div_'+param);
		  alink = document.getElementById('a_'+param);
			
			if (div && alink){
				if (expand_flag.value=='0') {
					div.style.display='inline';
					alink.innerHTML='Hide';
				}
				else {
					div.style.display='none';
					alink.innerHTML='Show';
				}
			}
		}
		if (expand_flag.value=='0') {
			expand_flag.value='1';
		}
		else {
			expand_flag.value='0';
		}
	}
</script>

</head>

<body>
<b>
<a href="index.html">About Ensembl Variation</a> |
<a href="data_description.html">Data Description</a> | 
<a href="predicted_data.html">Predicted Data</a> |
<a href="database.html">Database Description</a> | 
<a href="/info/docs/api/variation/index.html">Perl API</a> |
<a href="vep/index.html">Variant Effect Predictor</a>
</b>
<br /><br />

<h1>Ensembl Variation Tables Description</h1>

<p>
This document gives a high-level description of the tables that
make up the Ensembl variation schema. Tables are listed by alphabetical order, and the purpose of each table is explained. It is intended to
allow people to familiarise themselves with the schema when
encountering it for the first time, or when they need to use some
tables that they've not used before.
</p>

<div style="float:left">
  <p>
This document refers to version <span style="font-weight:bold;font-size:11pt;color:#000">71</span> of the Ensembl
variation schema. 
  </p>
  <p>
  The variation database schema diagram (PDF format) is available here:
  </p>
</div>
<div style="float:left;margin-left:5px">
  <a href="variation-database-schema.pdf">
    <img src="/img/variation_schema_thumb.png" style="border:1px solid navy" alt="Variation Schema Diagram" title="Variation Schema Diagram (PDF)"/>
  </a>
</div>
<div style="clear:both">
<br /><br />

<h3 id="top">List of the tables (sorted by group)</h3>
<div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#FF0000;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Attributes tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#attrib"><b>attrib</b></a></li>
        <li><a href="#attrib_set"><b>attrib_set</b></a></li>
        <li><a href="#attrib_type"><b>attrib_type</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#3CB371;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Failed tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#failed_allele"><b>failed_allele</b></a></li>
        <li><a href="#failed_description"><b>failed_description</b></a></li>
        <li><a href="#failed_structural_variation"><b>failed_structural_variation</b></a></li>
        <li><a href="#failed_variation"><b>failed_variation</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#FF8500;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Genotype tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#compressed_genotype_region"><b>compressed_genotype_region</b></a></li>
        <li><a href="#compressed_genotype_var"><b>compressed_genotype_var</b></a></li>
        <li><a href="#individual_genotype_multiple_bp"><b>individual_genotype_multiple_bp</b></a></li>
        <li><a href="#population_genotype"><b>population_genotype</b></a></li>
        <li><a href="#read_coverage"><b>read_coverage</b></a></li>
        <li><a href="#tmp_individual_genotype_single_bp"><b>tmp_individual_genotype_single_bp</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#BC5CEC;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Metadata tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#meta"><b>meta</b></a></li>
        <li><a href="#meta_coord"><b>meta_coord</b></a></li>
      </ul>  </div>
  <div style="clear:both" />
</div>

<div>
  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#000;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Other tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#allele_code"><b>allele_code</b></a></li>
        <li><a href="#coord_system"><b>coord_system</b></a></li>
        <li><a href="#genotype_code"><b>genotype_code</b></a></li>
        <li><a href="#phenotype"><b>phenotype</b></a></li>
        <li><a href="#seq_region"><b>seq_region</b></a></li>
        <li><a href="#submitter_handle"><b>submitter_handle</b></a></li>
        <li><a href="#subsnp_handle"><b>subsnp_handle</b></a></li>
        <li><a href="#tagged_variation_feature"><b>tagged_variation_feature</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#22949b;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Phenotype tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#phenotype_feature"><b>phenotype_feature</b></a></li>
        <li><a href="#phenotype_feature_attrib"><b>phenotype_feature_attrib</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#1E90FF;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Protein tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#protein_function_predictions"><b>protein_function_predictions</b></a></li>
        <li><a href="#translation_md5"><b>translation_md5</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#FF8500;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Sample tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#individual"><b>individual</b></a></li>
        <li><a href="#individual_population"><b>individual_population</b></a></li>
        <li><a href="#individual_type"><b>individual_type</b></a></li>
        <li><a href="#population"><b>population</b></a></li>
        <li><a href="#population_structure"><b>population_structure</b></a></li>
        <li><a href="#sample"><b>sample</b></a></li>
        <li><a href="#sample_synonym"><b>sample_synonym</b></a></li>
      </ul>  </div>
  <div style="clear:both" />
</div>

<div>
  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#7CFC00;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Source/study tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#associate_study"><b>associate_study</b></a></li>
        <li><a href="#source"><b>source</b></a></li>
        <li><a href="#study"><b>study</b></a></li>
        <li><a href="#study_variation"><b>study_variation</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#01D4F7;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Structural variation tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#structural_variation"><b>structural_variation</b></a></li>
        <li><a href="#structural_variation_association"><b>structural_variation_association</b></a></li>
        <li><a href="#structural_variation_feature"><b>structural_variation_feature</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#FF4DC8;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Variation effect tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#motif_feature_variation"><b>motif_feature_variation</b></a></li>
        <li><a href="#regulatory_feature_variation"><b>regulatory_feature_variation</b></a></li>
        <li><a href="#transcript_variation"><b>transcript_variation</b></a></li>
      </ul>  </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#FFD700;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Variation set tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#variation_set"><b>variation_set</b></a></li>
        <li><a href="#variation_set_structural_variation"><b>variation_set_structural_variation</b></a></li>
        <li><a href="#variation_set_structure"><b>variation_set_structure</b></a></li>
        <li><a href="#variation_set_variation"><b>variation_set_variation</b></a></li>
      </ul>  </div>
  <div style="clear:both" />
</div>

<div>
  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#000;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Variation tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#allele"><b>allele</b></a></li>
        <li><a href="#variation"><b>variation</b></a></li>
        <li><a href="#variation_feature"><b>variation_feature</b></a></li>
        <li><a href="#variation_synonym"><b>variation_synonym</b></a></li>
      </ul>  </div>

  <div style="clear:both" />
</div>
  <input type="button" onclick="show_hide_all()" class="fbutton" value="Show/hide all"/>
  <input type="hidden" id="expand" value="0" />
  
<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #FF0000">
  <h2 style="display:inline;color:#000">Attributes tables</h2>
</div>
<p style="width:800px">These tables define the variation attributes data.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF0000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="attrib" style="font-size:11pt;font-weight:bold">attrib</span></td>
      <td style="width:180px;text-align:right"><a id="a_attrib" style="cursor:pointer;text-decoration:underline" onclick="show_hide('attrib')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Defines various attributes used elsewhere in the database</p>

  <div id="div_attrib" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>attrib_id</b></td><td>INT(11)</td><td>0</td><td>Primary key</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>attrib_type_id</b></td><td>SMALLINT(5)</td><td>0</td><td>Key into the <a href="#attrib_type">attrib_type</a> table, identifies the type of this attribute</td><td>unique key: type_val_idx</td></tr>
      <tr class="bg1"><td><b>value</b></td><td>TEXT</td><td></td><td>The value of this attribute</td><td>unique key: type_val_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#attrib_type">attrib_type</a></li>
    <li><a href="#attrib_set">attrib_set</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF0000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="attrib_set" style="font-size:11pt;font-weight:bold">attrib_set</span></td>
      <td style="width:180px;text-align:right"><a id="a_attrib_set" style="cursor:pointer;text-decoration:underline" onclick="show_hide('attrib_set')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Groups related attributes together</p>

  <div id="div_attrib_set" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>attrib_set_id</b></td><td>INT(11)</td><td>0</td><td>Primary key</td><td>unique key: set_idx</td></tr>
      <tr class="bg2"><td><b>attrib_id</b></td><td>INT(11)</td><td>0</td><td>Key of an attribute in this set</td><td>unique key: set_idx<br />key: attrib_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#attrib">attrib</a></li>
    <li><a href="#attrib_type">attrib_type</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF0000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="attrib_type" style="font-size:11pt;font-weight:bold">attrib_type</span></td>
      <td style="width:180px;text-align:right"><a id="a_attrib_type" style="cursor:pointer;text-decoration:underline" onclick="show_hide('attrib_type')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Defines the set of possible attribute types used in the attrib table</p>

  <div id="div_attrib_type" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>attrib_type_id</b></td><td>SMALLINT(5)</td><td>0</td><td>Primary key</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>code</b></td><td>VARCHAR(20)</td><td>''</td><td>A short codename for this type (indexed, so should be used for lookups)</td><td>unique key: code_idx</td></tr>
      <tr class="bg1"><td><b>name</b></td><td>VARCHAR(255)</td><td>''</td><td>The name of this type</td><td></td></tr>
      <tr class="bg2"><td><b>description</b></td><td>TEXT</td><td></td><td>Longer description of this type</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#attrib">attrib</a></li>
    <li><a href="#attrib_set">attrib_set</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #3CB371">
  <h2 style="display:inline;color:#000">Failed tables</h2>
</div>
<p style="width:800px">These tables define the list of variants/alleles flagged as "failed" in the Variation pipeline.<br /> The list of reasons for a variation being flagged as failed is available in the <a href="data_description.html#quality_control">Quality Control documentation</a>.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#3CB371"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="failed_allele" style="font-size:11pt;font-weight:bold">failed_allele</span></td>
      <td style="width:180px;text-align:right"><a id="a_failed_allele" style="cursor:pointer;text-decoration:underline" onclick="show_hide('failed_allele')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Contains alleles that did not pass the Ensembl filters</p>

  <div id="div_failed_allele" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>failed_allele_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>allele_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#allele">allele</a> table.</td><td>unique key: allele_idx</td></tr>
      <tr class="bg1"><td><b>failed_description_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#failed_description">failed_description</a> table.</td><td>unique key: allele_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#failed_description">failed_description</a></li>
    <li><a href="#allele">allele</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#3CB371"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="failed_description" style="font-size:11pt;font-weight:bold">failed_description</span></td>
      <td style="width:180px;text-align:right"><a id="a_failed_description" style="cursor:pointer;text-decoration:underline" onclick="show_hide('failed_description')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains descriptions of reasons for a variation being flagged as failed.</p>

  <div id="div_failed_description" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>failed_description_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>description</b></td><td>text</td><td></td><td>Text containing the reason why the Variation has been flagged as failed. e.g. "Variation does not map to the genome".</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#failed_variation">failed_variation</a></li>
    <li><a href="#failed_allele">failed_allele</a></li>
    <li><a href="#failed_structural_variation">failed_structural_variation</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#3CB371"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="failed_structural_variation" style="font-size:11pt;font-weight:bold">failed_structural_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_failed_structural_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('failed_structural_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">For various reasons it may be necessary to store information about a structural variation that has failed quality checks (mappings) in the Structural Variation pipeline. This table acts as a flag for such failures.</p>

  <div id="div_failed_structural_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>failed_structural_variation_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>structural_variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#structural_variation">structural_variation</a> table.</td><td>unique key: structural_variation_idx</td></tr>
      <tr class="bg1"><td><b>failed_description_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#failed_description">failed_description</a> table.</td><td>unique key: structural_variation_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#failed_description">failed_description</a></li>
    <li><a href="#structural_variation">structural_variation</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#3CB371"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="failed_variation" style="font-size:11pt;font-weight:bold">failed_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_failed_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('failed_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">For various reasons it may be necessary to store information about a variation that has failed quality checks in the Variation pipeline. This table acts as a flag for such failures.</p>

  <div id="div_failed_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>failed_variation_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>unique key: variation_idx</td></tr>
      <tr class="bg1"><td><b>failed_description_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#failed_description">failed_description</a> table.</td><td>unique key: variation_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#failed_description">failed_description</a></li>
    <li><a href="#variation">variation</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #FF8500">
  <h2 style="display:inline;color:#000">Genotype tables</h2>
</div>
<p style="width:800px">These tables define the genotype data at the individual and population levels.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="compressed_genotype_region" style="font-size:11pt;font-weight:bold">compressed_genotype_region</span></td>
      <td style="width:180px;text-align:right"><a id="a_compressed_genotype_region" style="cursor:pointer;text-decoration:underline" onclick="show_hide('compressed_genotype_region')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table holds genotypes compressed using the pack() method in Perl. These genotypes are mapped to particular genomic locations rather than variation objects. The data have been compressed to reduce table size and increase the speed of the web code when retrieving strain slices and LD data. Only data from resequenced and individuals used for LD calculations are included in this table</p>

  <div id="div_compressed_genotype_region" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the sample table.</td><td>key: sample_idx</td></tr>
      <tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. ers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
      <tr class="bg1"><td><b>seq_region_start</b></td><td>int(11)</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
      <tr class="bg2"><td><b>seq_region_end</b></td><td>int(11)</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint(4)</td><td></td><td>The orientation of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>genotypes</b></td><td>blob</td><td></td><td>Encoded representation of the genotype data:<br />Each row in the compressed table stores genotypes from one individual in one fixed-size region of the genome (arbitrarily defined as 100 Kb). The compressed string (using Perl's pack method) consisting of a repeating triplet of elements: a distance in base pairs from the previous genotype; a variation dbID; a genotype_code_id identifier.<br />For example, a given row may have a start position of 1000, indicating the chromosomal position of the first genotype in this row. The unpacked genotypes field then may contain the following elements:<br />0, 1, 1, 20, 2, 5, 35, 3, 3, ...<br />The first genotype has a position of 1000 + 0 = 1000, and corresponds to the variation with the identifier 1 and genotype_code corresponding to A and G.<br />The second genotype has a position of 1000 + 20 = 1020, variation_id 2 and genotype_code representing C and C.<br />The third genotype similarly has a position of 1055, and so on.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
    <li><a href="#seq_region">seq_region</a></li>
    <li><a href="#variation">variation</a></li>
    <li><a href="#genotype_code">genotype_code</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="compressed_genotype_var" style="font-size:11pt;font-weight:bold">compressed_genotype_var</span></td>
      <td style="width:180px;text-align:right"><a id="a_compressed_genotype_var" style="cursor:pointer;text-decoration:underline" onclick="show_hide('compressed_genotype_var')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table holds genotypes compressed using the pack() method in Perl. These genotypes are mapped directly to variation objects. The data have been compressed to reduce table size. All genotypes in the database are included in this table (included duplicates of those genotypes contained in the compressed_genotype_region table). This table is optimised for retrieval from </p>

  <div id="div_compressed_genotype_var" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_id</b></td><td>int(11)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
      <tr class="bg2"><td><b>subsnp_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the <a href="#subsnp_handle">subsnp_handle</a> table.</td><td>key: subsnp_idx</td></tr>
      <tr class="bg1"><td><b>genotypes</b></td><td>blob</td><td></td><td>Encoded representation of the genotype data:<br />Each row in the compressed table stores genotypes from one subsnp of a variation (or one variation if no subsnp is defined). The compressed string (using Perl's pack method) consisting of a repeating pair of elements: an internal sample_id corresponding to an individual; a genotype_code_id identifier.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
    <li><a href="#variation">variation</a></li>
    <li><a href="#genotype_code">genotype_code</a></li>
    <li><a href="#subsnp_handle">subsnp_handle</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="individual_genotype_multiple_bp" style="font-size:11pt;font-weight:bold">individual_genotype_multiple_bp</span></td>
      <td style="width:180px;text-align:right"><a id="a_individual_genotype_multiple_bp" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual_genotype_multiple_bp')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table holds uncompressed genotypes for given variations.</p>

  <div id="div_individual_genotype_multiple_bp" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
      <tr class="bg2"><td><b>subsnp_id</b></td><td>int(15)</td><td></td><td>Foreign key references to the <a href="#subsnp_handle">subsnp_handle</a> table.</td><td>key: subsnp_idx</td></tr>
      <tr class="bg1"><td><b>allele_1</b></td><td>varchar(25000)</td><td></td><td>One of the alleles of the genotype, e.g. "TAG".</td><td></td></tr>
      <tr class="bg2"><td><b>allele_2</b></td><td>varchar(25000)</td><td></td><td>The other allele of the genotype.</td><td></td></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td>key: sample_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
    <li><a href="#variation">variation</a></li>
    <li><a href="#subsnp_handle">subsnp_handle</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="population_genotype" style="font-size:11pt;font-weight:bold">population_genotype</span></td>
      <td style="width:180px;text-align:right"><a id="a_population_genotype" style="cursor:pointer;text-decoration:underline" onclick="show_hide('population_genotype')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores genotypes and frequencies for variations in given populations.</p>

  <div id="div_population_genotype" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>population_genotype_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_id</b></td><td>int(11)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
      <tr class="bg1"><td><b>subsnp_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the subsnp_handle table.</td><td>key: subsnp_idx</td></tr>
      <tr class="bg2"><td><b>genotype_code_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key reference to the <a href="#genotype_code">genotype_code</a> table.</td><td></td></tr>
      <tr class="bg1"><td><b>frequency</b></td><td>float</td><td>NULL</td><td>Frequency of the genotype in the population.</td><td></td></tr>
      <tr class="bg2"><td><b>sample_id</b></td><td>int(10)</td><td>NULL</td><td>Foreign key references to the <a href="#population">population</a> table.</td><td>key: sample_idx</td></tr>
      <tr class="bg1"><td><b>count</b></td><td>int(10)</td><td>NULL</td><td>Number of individuals who have this genotype, in this population.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#population">population</a></li>
    <li><a href="#variation">variation</a></li>
    <li><a href="#subsnp_handle">subsnp_handle</a></li>
    <li><a href="#genotype_code">genotype_code</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="read_coverage" style="font-size:11pt;font-weight:bold">read_coverage</span></td>
      <td style="width:180px;text-align:right"><a id="a_read_coverage" style="cursor:pointer;text-decoration:underline" onclick="show_hide('read_coverage')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores the read coverage in the resequencing of individuals. Each row contains an individual ID, chromosomal coordinates and a read coverage level.</p>

  <div id="div_read_coverage" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. ers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: seq_region_idx</td></tr>
      <tr class="bg2"><td><b>seq_region_start</b></td><td>int</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: seq_region_idx</td></tr>
      <tr class="bg1"><td><b>seq_region_end</b></td><td>int</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>level</b></td><td>tinyint</td><td></td><td>Minimum number of reads.</td><td></td></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
    <li><a href="#seq_region">seq_region</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="tmp_individual_genotype_single_bp" style="font-size:11pt;font-weight:bold">tmp_individual_genotype_single_bp</span></td>
      <td style="width:180px;text-align:right"><a id="a_tmp_individual_genotype_single_bp" style="cursor:pointer;text-decoration:underline" onclick="show_hide('tmp_individual_genotype_single_bp')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">his table is only needed for create master schema when run healthcheck system. Needed for other species, but human, so keep it.</p>

  <div id="div_tmp_individual_genotype_single_bp" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
      <tr class="bg2"><td><b>subsnp_id</b></td><td>int(15)</td><td></td><td>Foreign key references to the <a href="#subsnp_handle">subsnp_handle</a> table.</td><td>key: subsnp_idx</td></tr>
      <tr class="bg1"><td><b>allele_1</b></td><td>char(1)</td><td></td><td>One of the alleles of the genotype, e.g. "TAG".</td><td></td></tr>
      <tr class="bg2"><td><b>allele_2</b></td><td>char(1)</td><td></td><td>The other allele of the genotype.</td><td></td></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td>key: sample_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
    <li><a href="#variation">variation</a></li>
    <li><a href="#subsnp_handle">subsnp_handle</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #BC5CEC">
  <h2 style="display:inline;color:#000">Metadata tables</h2>
</div>
<p style="width:800px">These tables define some metadata information.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#BC5CEC"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="meta" style="font-size:11pt;font-weight:bold">meta</span></td>
      <td style="width:180px;text-align:right"><a id="a_meta" style="cursor:pointer;text-decoration:underline" onclick="show_hide('meta')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores various metadata relating to the database, generally used by the Ensembl web code.</p>

  <div id="div_meta" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>meta_id</b></td><td>INT(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>species_id</b></td><td>INT</td><td>1</td><td>...</td><td>unique key: species_key_value_idx<br />key: species_value_idx</td></tr>
      <tr class="bg1"><td><b>meta_key</b></td><td>varchar( 40 )</td><td></td><td>Name of the meta entry, e.g. "schema_version".</td><td>unique key: species_key_value_idx</td></tr>
      <tr class="bg2"><td><b>meta_value</b></td><td>varchar( 255 )</td><td></td><td>Corresponding value of the key, e.g. "61".</td><td>unique key: species_key_value_idx<br />key: species_value_idx</td></tr>
    </table>
  </div>

<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#BC5CEC"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="meta_coord" style="font-size:11pt;font-weight:bold">meta_coord</span></td>
      <td style="width:180px;text-align:right"><a id="a_meta_coord" style="cursor:pointer;text-decoration:underline" onclick="show_hide('meta_coord')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table gives the coordinate system used by various tables in the database.</p>

  <div id="div_meta_coord" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>table_name</b></td><td>VARCHAR(40)</td><td></td><td>Name of the feature table, e.g. "variation_feature".</td><td>unique</td></tr>
      <tr class="bg2"><td><b>coord_system_id</b></td><td>INT(10)</td><td></td><td>Foreign key to core database coord_system table refers to coordinate system that features from this table can be found in.</td><td>unique</td></tr>
      <tr class="bg1"><td><b>max_length</b></td><td>INT</td><td></td><td>Maximun length of the feature. </td><td></td></tr>
    </table>
  </div>

<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #000">
  <h2 style="display:inline;color:#000">Other tables</h2>
</div>
<p style="width:800px">These tables define the other data associated with a variation.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="allele_code" style="font-size:11pt;font-weight:bold">allele_code</span></td>
      <td style="width:180px;text-align:right"><a id="a_allele_code" style="cursor:pointer;text-decoration:underline" onclick="show_hide('allele_code')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores the relationship between the internal allele identifiers and the alleles themselves.</p>

  <div id="div_allele_code" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>allele_code_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>allele</b></td><td>varchar(60000)</td><td>NULL</td><td>String representing the allele. Has a unique constraint on the first 1000 characters (max allowed by MySQL).</td><td>unique key: allele_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#allele">allele</a></li>
    <li><a href="#genotype_code">genotype_code</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="coord_system" style="font-size:11pt;font-weight:bold">coord_system</span></td>
      <td style="width:180px;text-align:right"><a id="a_coord_system" style="cursor:pointer;text-decoration:underline" onclick="show_hide('coord_system')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Stores information about the available co-ordinate systems for the species identified through the species_id field. Note that for each species, there must be one co-ordinate system that has the attribute "top_level" and one that has the attribute "sequence_level".</p>

  <div id="div_coord_system" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>coord_system_id</b></td><td>INT(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary   key</td></tr>
      <tr class="bg2"><td><b>species_id</b></td><td>INT(10)</td><td>1</td><td>Indentifies the species for multi-species databases.</td><td>unique    key: rank_idx<br />unique    key: name_idx<br />key: species_idx</td></tr>
      <tr class="bg1"><td><b>name</b></td><td>VARCHAR(40)</td><td></td><td>Co-oridinate system name, e.g. 'chromosome', 'contig', 'scaffold' etc.</td><td>unique    key: name_idx</td></tr>
      <tr class="bg2"><td><b>version</b></td><td>VARCHAR(255)</td><td>NULL</td><td>Assembly.</td><td>unique    key: name_idx</td></tr>
      <tr class="bg1"><td><b>rank</b></td><td>INT</td><td></td><td>Co-oridinate system rank.</td><td>unique    key: rank_idx</td></tr>
      <tr class="bg2"><td><b>attrib</b></td><td>SET('default_version', 'sequence_level')</td><td></td><td>Co-oridinate system attrib (e.g. "top_level", "sequence_level").</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#seq_region">seq_region</a></li>
    <li><a href="#meta_coord">meta_coord</a></li>
    <li><a href="#meta">meta</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="genotype_code" style="font-size:11pt;font-weight:bold">genotype_code</span></td>
      <td style="width:180px;text-align:right"><a id="a_genotype_code" style="cursor:pointer;text-decoration:underline" onclick="show_hide('genotype_code')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores genotype codes as multiple rows of allele_code identifiers, linked by genotype_code_id and ordered by haplotype_id.</p>

  <div id="div_genotype_code" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>genotype_code_id</b></td><td>int(11)</td><td></td><td>Internal identifier.</td><td>key: genotype_code_id</td></tr>
      <tr class="bg2"><td><b>allele_code_id</b></td><td>int(11)</td><td></td><td>Foreign key reference to <a href="#allele_code">allele_code</a> table.</td><td>key: allele_code_id</td></tr>
      <tr class="bg1"><td><b>haplotype_id</b></td><td>tinyint(2)</td><td></td><td>Sorting order of the genotype's alleles.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#allele_code">allele_code</a></li>
    <li><a href="#population_genotype">population_genotype</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="phenotype" style="font-size:11pt;font-weight:bold">phenotype</span></td>
      <td style="width:180px;text-align:right"><a id="a_phenotype" style="cursor:pointer;text-decoration:underline" onclick="show_hide('phenotype')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores details of the phenotypes associated with phenotype_features.</p>

  <div id="div_phenotype" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>phenotype_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>stable_id</b></td><td>varchar(255)</td><td>NULL</td><td>Ensembl stable identifier for the phenotype</td><td>key: stable_idx</td></tr>
      <tr class="bg1"><td><b>name</b></td><td>varchar(50)</td><td>NULL</td><td>Phenotype short name. e.g. "CAD".</td><td>unique key: name_idx</td></tr>
      <tr class="bg2"><td><b>description</b></td><td>varchar(255)</td><td>NULL</td><td>varchar		Phenotype long name. e.g. "Coronary Artery Disease".</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#phenotype_feature">phenotype_feature</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="seq_region" style="font-size:11pt;font-weight:bold">seq_region</span></td>
      <td style="width:180px;text-align:right"><a id="a_seq_region" style="cursor:pointer;text-decoration:underline" onclick="show_hide('seq_region')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores the relationship between Ensembl's internal coordinate system identifiers and traditional chromosome names.</p>

  <div id="div_seq_region" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>seq_region_id</b></td><td>INT(10)</td><td></td><td>Primary key. Foreign key references seq_region in core db. Refers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>VARCHAR(40)</td><td></td><td>The name of this sequence region.</td><td>unique key: name_cs_idx</td></tr>
      <tr class="bg1"><td><b>coord_system_id</b></td><td>INT(10)</td><td></td><td>Foreign key references to the <a href="#coord_system">coord_system</a> table.</td><td>unique key: name_cs_idx<br />key: cs_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_feature">variation_feature</a></li>
    <li><a href="#compressed_genotype_region">compressed_genotype_region</a></li>
    <li><a href="#read_coverage">read_coverage</a></li>
    <li><a href="#coord_system">coord_system</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="submitter_handle" style="font-size:11pt;font-weight:bold">submitter_handle</span></td>
      <td style="width:180px;text-align:right"><a id="a_submitter_handle" style="cursor:pointer;text-decoration:underline" onclick="show_hide('submitter_handle')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table holds a short string to distinguish data submitters</p>

  <div id="div_submitter_handle" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>handle_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>handle</b></td><td>varchar(25)</td><td></td><td>Short string assigned to the data submitter. </td><td>unique</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#allele ">allele </a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="subsnp_handle" style="font-size:11pt;font-weight:bold">subsnp_handle</span></td>
      <td style="width:180px;text-align:right"><a id="a_subsnp_handle" style="cursor:pointer;text-decoration:underline" onclick="show_hide('subsnp_handle')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains the SubSNP(ss) ID and the name of the submitter handle of dbSNP.</p>

  <div id="div_subsnp_handle" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>subsnp_id</b></td><td>int(11)</td><td></td><td>Primary key. It corresponds to the subsnp identifier (ssID) from dbSNP.<br />This ssID is stored in this table without the "ss" prefix. e.g. "120258606" instead of "ss120258606".</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>handle</b></td><td>varchar(20)</td><td></td><td>The name of the dbSNP handler who submitted the ssID.<br />Name of the synonym (a different <b>sample_id</b>).</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#allele">allele</a></li>
    <li><a href="#failed_variation">failed_variation</a></li>
    <li><a href="#population_genotype">population_genotype</a></li>
    <li><a href="#sample">sample</a></li>
    <li><a href="#variation_synonym">variation_synonym</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="tagged_variation_feature" style="font-size:11pt;font-weight:bold">tagged_variation_feature</span></td>
      <td style="width:180px;text-align:right"><a id="a_tagged_variation_feature" style="cursor:pointer;text-decoration:underline" onclick="show_hide('tagged_variation_feature')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table lists variation feature IDs that are tagged by another variation feature ID. Tag pairs are defined as having an r<sup>2</sup> &gt; 0.99.</p>

  <div id="div_tagged_variation_feature" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_feature_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation_feature">variation_feature</a> table.</td><td>key: tag_idx</td></tr>
      <tr class="bg2"><td><b>tagged_variation_feature_id</b></td><td>int(10)</td><td>NULL</td><td>Foreign key references to the <a href="#variation_feature">variation_feature</a> table.</td><td>key: tagged_idx</td></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#sample">sample</a> table.</td><td>key: sample_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_feature">variation_feature</a></li>
    <li><a href="#population">population</a></li>
    <li><a href="#sample">sample</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #22949b">
  <h2 style="display:inline;color:#000">Phenotype tables</h2>
</div>
<p style="width:800px">These tables store information linking entities (variants, genes, QTLs) with phenotypes and other annotations.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#22949b"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="phenotype_feature" style="font-size:11pt;font-weight:bold">phenotype_feature</span></td>
      <td style="width:180px;text-align:right"><a id="a_phenotype_feature" style="cursor:pointer;text-decoration:underline" onclick="show_hide('phenotype_feature')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores information linking entities (variants, genes, QTLs) and phenotypes.</p>

  <div id="div_phenotype_feature" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>phenotype_feature_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>phenotype_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the <a href="#phenotype">phenotype</a> table.</td><td>key: phenotype_idx</td></tr>
      <tr class="bg1"><td><b>source_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the <a href="#source">source</a> table.</td><td></td></tr>
      <tr class="bg2"><td><b>study_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the <a href="#study">study</a> table.</td><td></td></tr>
      <tr class="bg1"><td><b>type</b></td><td>enum('Gene','Variation','StructuralVariation','SupportingStructuralVariation','QTL','RegulatoryFeature')</td><td>NULL</td><td>Type of object associated.</td><td>key: object_idx<br />key: type_idx</td></tr>
      <tr class="bg2"><td><b>object_id</b></td><td>varchar(255)</td><td>NULL</td><td>Stable identifier for associated object.</td><td>key: object_idx</td></tr>
      <tr class="bg1"><td><b>is_significant</b></td><td>tinyint(1)</td><td>'1'</td><td>Flag indicating if the association is statistically significant in the given study.</td><td></td></tr>
      <tr class="bg2"><td><b>seq_region_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. Refers to the seq_region which this feature is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
      <tr class="bg1"><td><b>seq_region_start</b></td><td>int(11)</td><td>NULL</td><td>The start position of the feature on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
      <tr class="bg2"><td><b>seq_region_end</b></td><td>int(11)</td><td>NULL</td><td>The end position of the feature on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
      <tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint(4)</td><td>NULL</td><td>The orientation of the feature on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation">variation</a></li>
    <li><a href="#phenotype">phenotype</a></li>
    <li><a href="#source">source</a></li>
    <li><a href="#study">study</a></li>
    <li><a href="#seq_region">seq_region</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#22949b"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="phenotype_feature_attrib" style="font-size:11pt;font-weight:bold">phenotype_feature_attrib</span></td>
      <td style="width:180px;text-align:right"><a id="a_phenotype_feature_attrib" style="cursor:pointer;text-decoration:underline" onclick="show_hide('phenotype_feature_attrib')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores additional information on a given phenotype/object association. It is styled as an attrib table to allow for a variety of fields to be populated across different object types.</p>

  <div id="div_phenotype_feature_attrib" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>phenotype_feature_id</b></td><td>int(11)</td><td></td><td>Foreign key, references to the <a href="#phenotype_feature">phenotype_feature</a> table.</td><td>key: phenotype_feature_idx</td></tr>
      <tr class="bg2"><td><b>attrib_type_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the <a href="#attrib_type">attrib_type</a> table.</td><td>key: type_value_idx</td></tr>
      <tr class="bg1"><td><b>value</b></td><td>varchar(255)</td><td>NULL</td><td>The value of the attribute.</td><td>key: type_value_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#phenotype_feature">phenotype_feature</a></li>
    <li><a href="#attrib_type">attrib_type</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #1E90FF">
  <h2 style="display:inline;color:#000">Protein tables</h2>
</div>
<p style="width:800px">These tables define the protein prediction data.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#1E90FF"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="protein_function_predictions" style="font-size:11pt;font-weight:bold">protein_function_predictions</span></td>
      <td style="width:180px;text-align:right"><a id="a_protein_function_predictions" style="cursor:pointer;text-decoration:underline" onclick="show_hide('protein_function_predictions')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Contains encoded protein function predictions for every protein-coding transcript in this species.</p>

  <div id="div_protein_function_predictions" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>translation_md5_id</b></td><td>int(11)</td><td></td><td>Identifies the MD5 hash corresponding to the protein sequence to which  these predictions apply</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>analysis_attrib_id</b></td><td>int(11)</td><td></td><td>Identifies the analysis (sift, polyphen etc.) that produced these predictions </td><td>primary key</td></tr>
      <tr class="bg1"><td><b>prediction_matrix</b></td><td>mediumblob</td><td></td><td>A compressed binary string containing the predictions for all possible  amino acid substitutions in this protein. See the explanation <a href="predicted_data.html#nsSNP">here</a></td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#translation_md5">translation_md5</a></li>
    <li><a href="#attrib">attrib</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#1E90FF"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="translation_md5" style="font-size:11pt;font-weight:bold">translation_md5</span></td>
      <td style="width:180px;text-align:right"><a id="a_translation_md5" style="cursor:pointer;text-decoration:underline" onclick="show_hide('translation_md5')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Maps a hex MD5 hash of a translation sequence to an ID used for the protein function predictions</p>

  <div id="div_translation_md5" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>translation_md5_id</b></td><td>int(11)</td><td></td><td>Primary key</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>translation_md5</b></td><td>char(32)</td><td></td><td>Hex MD5 hash of a translation sequence</td><td>unique key: md5_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#protein_function_predictions">protein_function_predictions</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #FF8500">
  <h2 style="display:inline;color:#000">Sample tables</h2>
</div>
<p style="width:800px">These tables define the individual and population information.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="individual" style="font-size:11pt;font-weight:bold">individual</span></td>
      <td style="width:180px;text-align:right"><a id="a_individual" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Stores information about an identifiable individual, including gender and the identifiers of the individual's parents (if known).</p>

  <div id="div_individual" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier. See the <a href="#sample">sample</a> table. Corresponds to the individual ID.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>gender</b></td><td>enum('Male', 'Female', 'Unknown')</td><td>'Unknown'</td><td>The sex of this individual.</td><td></td></tr>
      <tr class="bg1"><td><b>father_individual_sample_id</b></td><td>int(10)</td><td></td><td>Self referential ID, the father of this individual if known.</td><td></td></tr>
      <tr class="bg2"><td><b>mother_individual_sample_id</b></td><td>int(10)</td><td></td><td>Self referential ID, the mother of this individual if known.</td><td></td></tr>
      <tr class="bg1"><td><b>individual_type_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual_type">individual_type</a> table.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#sample">sample</a></li>
    <li><a href="#individual_type">individual_type</a></li>
    <li><a href="#individual_population">individual_population</a></li>
    <li><a href="#individual_genotype_multiple_bp">individual_genotype_multiple_bp</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="individual_population" style="font-size:11pt;font-weight:bold">individual_population</span></td>
      <td style="width:180px;text-align:right"><a id="a_individual_population" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual_population')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table resolves the many-to-many relationship between the individual and population tables; i.e. samples may belong to more than one population. Hence it is composed of rows of individual and population identifiers.</p>

  <div id="div_individual_population" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>individual_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#individual">individual</a> table.</td><td>key: individual_sample_idx</td></tr>
      <tr class="bg2"><td><b>population_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#population">population</a> table.</td><td>key: population_sample_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
    <li><a href="#population">population</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="individual_type" style="font-size:11pt;font-weight:bold">individual_type</span></td>
      <td style="width:180px;text-align:right"><a id="a_individual_type" style="cursor:pointer;text-decoration:underline" onclick="show_hide('individual_type')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table gives a deatailed description for each of the possible individual types: fully_inbred, partly_inbred, outbred, mutant  </p>

  <div id="div_individual_type" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>individual_type_id</b></td><td>int(0)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Short name of the individual type. e.g. "fully_inbred","mutant".</td><td></td></tr>
      <tr class="bg1"><td><b>description</b></td><td>text</td><td></td><td>Long name of the individual type.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="population" style="font-size:11pt;font-weight:bold">population</span></td>
      <td style="width:180px;text-align:right"><a id="a_population" style="cursor:pointer;text-decoration:underline" onclick="show_hide('population')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">A table consisting simply of sample_ids representing populations; all data relating to the populations are stored in separate tables (see below).<br />A population may be an ethnic group (e.g. caucasian, hispanic), assay group (e.g. 24 europeans), strain, phenotypic group (e.g. blue eyed, diabetes) etc. Populations may be composed of other populations by defining relationships in the population_structure table.</p>

  <div id="div_population" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#sample">sample</a> table. Corresponds to the population ID.</td><td>primary key</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#sample">sample</a></li>
    <li><a href="#sample_synonym">sample_synonym</a></li>
    <li><a href="#individual_population">individual_population</a></li>
    <li><a href="#population_structure">population_structure</a></li>
    <li><a href="#population_genotype">population_genotype</a></li>
    <li><a href="#allele">allele</a></li>
    <li><a href="#tagged_variation_feature">tagged_variation_feature</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="population_structure" style="font-size:11pt;font-weight:bold">population_structure</span></td>
      <td style="width:180px;text-align:right"><a id="a_population_structure" style="cursor:pointer;text-decoration:underline" onclick="show_hide('population_structure')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores hierarchical relationships between populations by relating them as populations and sub-populations.</p>

  <div id="div_population_structure" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>super_population_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the population table.</td><td>unique<br />key: sub_pop_sample_idx</td></tr>
      <tr class="bg2"><td><b>sub_population_sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the population table.</td><td>unique<br />key: sub_pop_sample_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#population">population</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="sample" style="font-size:11pt;font-weight:bold">sample</span></td>
      <td style="width:180px;text-align:right"><a id="a_sample" style="cursor:pointer;text-decoration:underline" onclick="show_hide('sample')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Sample is used as a generic catch-all term to cover individuals, populations and strains; it contains a name and description, as well as a size if applicable to the population.<br /> Some population examples are available <a href="data_description.html#populations">here</a>.</p>

  <div id="div_sample" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the sample (can be an individual or a population name).</td><td>key: name_idx</td></tr>
      <tr class="bg1"><td><b>size</b></td><td>int</td><td></td><td>Number of individual in the sample.</td><td></td></tr>
      <tr class="bg2"><td><b>description</b></td><td>text</td><td></td><td>Description of the sample.</td><td></td></tr>
      <tr class="bg1"><td><b>display</b></td><td>enum('REFERENCE',<br />'DEFAULT',<br />'DISPLAYABLE',<br />'UNDISPLAYABLE',<br />'LD',<br />'MARTDISPLAYABLE')</td><td>'UNDISPLAYABLE'</td><td>Information used by the website: samples with little information are filtered from some web displays.</td><td></td></tr>
      <tr class="bg2"><td><b>freqs_from_gts</b></td><td>tinyint(1)</td><td>'0'</td><td>Flag indicating if the population frequencies can be retrieved from the allele table (0) or from the individual genotypes (1).</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#individual">individual</a></li>
    <li><a href="#population">population</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF8500"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="sample_synonym" style="font-size:11pt;font-weight:bold">sample_synonym</span></td>
      <td style="width:180px;text-align:right"><a id="a_sample_synonym" style="cursor:pointer;text-decoration:underline" onclick="show_hide('sample_synonym')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Used to store alternative names for populations when data comes from multiple sources.</p>

  <div id="div_sample_synonym" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>sample_synonym_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>sample_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#sample">sample</a> table.</td><td>key: sample_idx</td></tr>
      <tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td>key: </td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the synonym (a different <b>sample_id</b>).</td><td>key: </td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#sample">sample</a></li>
    <li><a href="#population">population</a></li>
    <li><a href="#source">source</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #7CFC00">
  <h2 style="display:inline;color:#000">Source/study tables</h2>
</div>
<p style="width:800px">These tables define the variation source and study information.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#7CFC00"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="associate_study" style="font-size:11pt;font-weight:bold">associate_study</span></td>
      <td style="width:180px;text-align:right"><a id="a_associate_study" style="cursor:pointer;text-decoration:underline" onclick="show_hide('associate_study')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains identifiers of associated studies (e.g. NHGRI and EGA studies with the same pubmed identifier).</p>

  <div id="div_associate_study" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>study1_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#study">study</a> table.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>study2_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#study">study</a> table.</td><td>primary key</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#study">study</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#7CFC00"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="source" style="font-size:11pt;font-weight:bold">source</span></td>
      <td style="width:180px;text-align:right"><a id="a_source" style="cursor:pointer;text-decoration:underline" onclick="show_hide('source')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains details of the source from which a variation is derived. Most commonly this is NCBI's dbSNP; other sources include SNPs called by Ensembl.<br /> You can see the complete list, by species, <a href="sources_documentation.html">here</a>.</p>

  <div id="div_source" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(24)</td><td></td><td>Name of the source. e.g. "dbSNP"</td><td></td></tr>
      <tr class="bg1"><td><b>version</b></td><td>int</td><td></td><td>Version number of the source (if available). e.g. "132"</td><td></td></tr>
      <tr class="bg2"><td><b>description</b></td><td>varchar(255)</td><td></td><td>Description of the source.</td><td></td></tr>
      <tr class="bg1"><td><b>url</b></td><td>varchar(255)</td><td></td><td>URL of the source.</td><td></td></tr>
      <tr class="bg2"><td><b>type</b></td><td>ENUM('chip','lsdb')</td><td>NULL</td><td>Define the type of the source, e.g. 'chip'</td><td></td></tr>
      <tr class="bg1"><td><b>somatic_status</b></td><td>ENUM ('germline','somatic','mixed')</td><td>'germline'</td><td>Indicates if this source includes somatic or germline mutations, or a mixture</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation">variation</a></li>
    <li><a href="#variation_synonym">variation_synonym</a></li>
    <li><a href="#variation_feature">variation_feature</a></li>
    <li><a href="#sample_synonym">sample_synonym</a></li>
    <li><a href="#structural_variation">structural_variation</a></li>
    <li><a href="#study">study</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#7CFC00"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="study" style="font-size:11pt;font-weight:bold">study</span></td>
      <td style="width:180px;text-align:right"><a id="a_study" style="cursor:pointer;text-decoration:underline" onclick="show_hide('study')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains details of the studies. The studies information can come from internal studies (DGVa, EGA) or from external studies (Uniprot, NHGRI, ...).</p>

  <div id="div_study" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>study_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td>key: source_idx</td></tr>
      <tr class="bg1"><td><b>name</b></td><td>varchar(255)</td><td>null</td><td>Name of the study. e.g. "EGAS00000000001"</td><td></td></tr>
      <tr class="bg2"><td><b>description</b></td><td>varchar(255)</td><td>NULL</td><td>Description of the study.</td><td></td></tr>
      <tr class="bg1"><td><b>url</b></td><td>varchar(255)</td><td>NULL</td><td>URL to find the study data (http or ftp).</td><td></td></tr>
      <tr class="bg2"><td><b>external_reference</b></td><td>varchar(255)</td><td>NULL</td><td>The pubmed/id or project name associated with this study.</td><td></td></tr>
      <tr class="bg1"><td><b>study_type</b></td><td>varchar(255)</td><td>NULL</td><td>Displays the type of the study (e.g. genome-wide association study, control-set, case-set, curated, ...).</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#source">source</a></li>
    <li><a href="#structural_variation">structural_variation</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#7CFC00"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="study_variation" style="font-size:11pt;font-weight:bold">study_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_study_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('study_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Links a variation to a study</p>

  <div id="div_study_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key, foreign key references variation</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>study_id</b></td><td>int(10)</td><td></td><td>Primary key, foreign key references study</td><td>primary key</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation ">variation </a></li>
    <li><a href="#study  ">study  </a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #01D4F7">
  <h2 style="display:inline;color:#000">Structural variation tables</h2>
</div>
<p style="width:800px">These tables define the structural variation data.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#01D4F7"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="structural_variation" style="font-size:11pt;font-weight:bold">structural_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_structural_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('structural_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores information about structural variation.</p>

  <div id="div_structural_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>structural_variation_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_name</b></td><td>varchar(255)</td><td>NULL</td><td>The external identifier or name of the variation. e.g. "esv9549".</td><td>key: name_idx</td></tr>
      <tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td>key: source_idx</td></tr>
      <tr class="bg2"><td><b>study_id</b></td><td>int(10)</td><td>NULL</td><td>Foreign key references to the <a href="#study">study</a> table.	</td><td>key: study_idx</td></tr>
      <tr class="bg1"><td><b>class_attrib_id</b></td><td>int(10)</td><td>0</td><td>Foreign key references to the <a href="#attrib">attrib</a> table. Defines the type of structural variant.<br />  The list of structural variation classes is available <a href="data_description.html#classes">here</a>.</td><td>key: attrib_idx</td></tr>
      <tr class="bg2"><td><b>validation_status</b></td><td>ENUM('validated','not validated','high quality')</td><td></td><td>Validation status of the variant.</td><td></td></tr>
      <tr class="bg1"><td><b>is_evidence</b></td><td>TINYINT(4)</td><td>0</td><td>Flag indicating if the structural variation is a supporting evidence (1) or not (0).</td><td></td></tr>
      <tr class="bg2"><td><b>somatic</b></td><td>TINYINT(1)</td><td>0</td><td>Flags whether this structural variation is known to be somatic or not</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#source">source</a></li>
    <li><a href="#study">study</a></li>
    <li><a href="#attrib">attrib</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#01D4F7"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="structural_variation_association" style="font-size:11pt;font-weight:bold">structural_variation_association</span></td>
      <td style="width:180px;text-align:right"><a id="a_structural_variation_association" style="cursor:pointer;text-decoration:underline" onclick="show_hide('structural_variation_association')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores the associations between structural variations and their supporting evidences.</p>

  <div id="div_structural_variation_association" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>structural_variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#structural_variation">structural_variation</a> table.</td><td>primary key<br />key: structural_variation_idx</td></tr>
      <tr class="bg2"><td><b>supporting_structural_variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#structural_variation">structural_variation</a> table.</td><td>primary key<br />key: supporting_structural_variation_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#structural_variation">structural_variation</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#01D4F7"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="structural_variation_feature" style="font-size:11pt;font-weight:bold">structural_variation_feature</span></td>
      <td style="width:180px;text-align:right"><a id="a_structural_variation_feature" style="cursor:pointer;text-decoration:underline" onclick="show_hide('structural_variation_feature')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores information about structural variation features (i.e. mappings of structural variations to genomic locations).</p>

  <div id="div_structural_variation_feature" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>structural_variation_feature_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. Refers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
      <tr class="bg1"><td><b>outer_start</b></td><td>int</td><td></td><td>The 5' outer bound position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>seq_region_start</b></td><td>int</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
      <tr class="bg1"><td><b>inner_start</b></td><td>int</td><td></td><td>The 5' inner bound position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>inner_end</b></td><td>int</td><td></td><td>The 3' inner bound position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg1"><td><b>seq_region_end</b></td><td>int</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
      <tr class="bg2"><td><b>outer_end</b></td><td>int</td><td></td><td>The 3' outer bound position of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint</td><td></td><td>The orientation of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>structural_variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#structural_variation">structural_variation</a> table.</td><td>key: structural_variation_idx</td></tr>
      <tr class="bg1"><td><b>variation_name</b></td><td>varchar(255)</td><td></td><td>A denormalisation taken from the structural_variation table. This is the name or identifier that is used for displaying the feature (e.g. "esv9549").</td><td></td></tr>
      <tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td>key: source_idx</td></tr>
      <tr class="bg1"><td><b>class_attrib_id</b></td><td>int(10)</td><td>0</td><td>Foreign key references to the <a href="#attrib">attrib</a> table. Defines the type of structural variant.<br />  The list of structural variation classes is available <a href="data_description.html#classes">here</a>.</td><td>key: attrib_idx</td></tr>
      <tr class="bg2"><td><b>allele_string</b></td><td>longtext</td><td>NULL</td><td>The variant allele, where known.</td><td></td></tr>
      <tr class="bg1"><td><b>is_evidence</b></td><td>tinyint(1)</td><td>0</td><td>Flag indicating if the structural variation is a supporting evidence (1) or not (0).</td><td></td></tr>
      <tr class="bg2"><td><b>variation_set_id</b></td><td>SET (<br />'1','2','3','4','5','6','7','8',<br />'9','10','11','12','13','14','15','16',<br />'17','18','19','20','21','22','23','24',<br />'25','26','27','28','29','30','31','32',<br />'33','34','35','36','37','38','39','40',<br />'41','42','43','44','45','46','47','48',<br />'49','50','51','52','53','54','55','56',<br />'57','58','59','60','61','62','63','64'<br /> )</td><td>''</td><td>The structural variation feature can belong to a <a href="#variation_set">variation_set</a>.</td><td>key: variation_set_idx</td></tr>
      <tr class="bg1"><td><b>somatic</b></td><td>TINYINT(1)</td><td>0</td><td>Flags whether this structural variation is known to be somatic or not</td><td></td></tr>
      <tr class="bg2"><td><b>breakpoint_order</b></td><td>TINYINT(4)</td><td>NULL</td><td>Defines the order of the breakpoints when several events/mutation occured for a structural variation (e.g. somatic mutations)</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#structural_variation">structural_variation</a></li>
    <li><a href="#source">source</a></li>
    <li><a href="#seq_region">seq_region</a></li>
    <li><a href="#attrib">attrib</a></li>
    <li><a href="#variation_set">variation_set</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #FF4DC8">
  <h2 style="display:inline;color:#000">Variation effect tables</h2>
</div>
<p style="width:800px">These tables define the variation effect prediction data in different Ensembl features.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF4DC8"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="motif_feature_variation" style="font-size:11pt;font-weight:bold">motif_feature_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_motif_feature_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('motif_feature_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table relates a single allele of a variation_feature to a motif feature (see Regulation documentation). It contains the consequence of the allele.</p>

  <div id="div_motif_feature_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>motif_feature_variation_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_feature_id</b></td><td>int(11)</td><td></td><td>Foreign key references to the <a href="#variation_feature">variation_feature</a> table.</td><td>key: variation_feature_idx</td></tr>
      <tr class="bg1"><td><b>feature_stable_id</b></td><td>varchar(128)</td><td>NULL</td><td>Foreign key to regulation databases. Unique stable id of related regulatory_feature.</td><td>key: somatic_feature_idx</td></tr>
      <tr class="bg2"><td><b>motif_feature_id</b></td><td>int(11)</td><td></td><td>Foreign key to regulation databases. Internal id of related motif_feature.</td><td></td></tr>
      <tr class="bg1"><td><b>allele_string</b></td><td>text</td><td></td><td>Shows the reference sequence and variant sequence of this allele.</td><td></td></tr>
      <tr class="bg2"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>Flags if the associated variation is known to be somatic.</td><td>key: somatic_feature_idx</td></tr>
      <tr class="bg1"><td><b>consequence_types</b></td><td>set (<br />'splice_acceptor_variant',<br />'splice_donor_variant',<br />'stop_lost',<br />'coding_sequence_variant',<br />'missense_variant',<br />'stop_gained',<br />'synonymous_variant',<br />'frameshift_variant',<br />'nc_transcript_variant',<br />'non_coding_exon_variant',<br />'mature_miRNA_variant',<br />'NMD_transcript_variant',<br />'5_prime_UTR_variant',<br />'3_prime_UTR_variant',<br />'incomplete_terminal_codon_variant',<br />'intron_variant',<br />'splice_region_variant',<br />'downstream_gene_variant',<br />'upstream_gene_variant',<br />'initiator_codon_variant',<br />'stop_retained_variant',<br />'inframe_insertion',<br />'inframe_deletion',<br />'transcript_ablation',<br />'transcript_fusion',<br />'transcript_amplification',<br />'transcript_translocation',<br />'TF_binding_site_variant',<br />'TFBS_ablation',<br />'TFBS_fusion',<br />'TFBS_amplification',<br />'TFBS_translocation',<br />'regulatory_region_variant',<br />'regulatory_region_ablation',<br />'regulatory_region_fusion',<br />'regulatory_region_amplification',<br />'regulatory_region_translocation',<br />'feature_elongation',<br />'feature_truncation'<br /> )</td><td></td><td>The consequence(s) of the variant allele on this motif_feature.<br /> The list of consequence descriptions is available <a href="predicted_data.html#consequences">here</a>.</td><td>key: consequence_type_idx</td></tr>
      <tr class="bg2"><td><b>motif_name</b></td><td>text</td><td></td><td>The display label of the motif.</td><td></td></tr>
      <tr class="bg1"><td><b>motif_start</b></td><td>int(11)</td><td></td><td>The start position of the variation in the motif.</td><td></td></tr>
      <tr class="bg2"><td><b>motif_end</b></td><td>int(11)</td><td></td><td>The end position of the variation in the motif.</td><td></td></tr>
      <tr class="bg1"><td><b>motif_score_delta</b></td><td>float</td><td>NULL</td><td>The deviation from the score (that is derived from alignment software (e.g. MOODS)) caused by the variation.</td><td></td></tr>
      <tr class="bg2"><td><b>in_informative_position</b></td><td>tinyint(1)</td><td>0</td><td>Flags if the variation is in an informative position.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_feature">variation_feature</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF4DC8"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="regulatory_feature_variation" style="font-size:11pt;font-weight:bold">regulatory_feature_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_regulatory_feature_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('regulatory_feature_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table relates a single allele of a variation_feature to a regulatory feature (see Regulation documentation). It contains the consequence of the allele.</p>

  <div id="div_regulatory_feature_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>regulatory_feature_variation_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_feature_id</b></td><td>int(11)</td><td></td><td>Foreign key references to the <a href="#variation_feature">variation_feature</a> table.</td><td>key: variation_feature_idx</td></tr>
      <tr class="bg1"><td><b>feature_stable_id</b></td><td>varchar(128)</td><td>NULL</td><td>Foreign key to regulation databases. Unique stable id of related regulatory_feature.</td><td>key: somatic_feature_idx</td></tr>
      <tr class="bg2"><td><b>feature_type</b></td><td>text</td><td></td><td>The name of the feature type.</td><td></td></tr>
      <tr class="bg1"><td><b>allele_string</b></td><td>text</td><td></td><td>Shows the reference sequence and variant sequence of this allele.</td><td></td></tr>
      <tr class="bg2"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>Flags if the associated variation is known to be somatic.</td><td>key: somatic_feature_idx</td></tr>
      <tr class="bg1"><td><b>consequence_types</b></td><td>set (<br />'splice_acceptor_variant',<br />'splice_donor_variant',<br />'stop_lost',<br />'coding_sequence_variant',<br />'missense_variant',<br />'stop_gained',<br />'synonymous_variant',<br />'frameshift_variant',<br />'nc_transcript_variant',<br />'non_coding_exon_variant',<br />'mature_miRNA_variant',<br />'NMD_transcript_variant',<br />'5_prime_UTR_variant',<br />'3_prime_UTR_variant',<br />'incomplete_terminal_codon_variant',<br />'intron_variant',<br />'splice_region_variant',<br />'downstream_gene_variant',<br />'upstream_gene_variant',<br />'initiator_codon_variant',<br />'stop_retained_variant',<br />'inframe_insertion',<br />'inframe_deletion',<br />'transcript_ablation',<br />'transcript_fusion',<br />'transcript_amplification',<br />'transcript_translocation',<br />'TF_binding_site_variant',<br />'TFBS_ablation',<br />'TFBS_fusion',<br />'TFBS_amplification',<br />'TFBS_translocation',<br />'regulatory_region_variant',<br />'regulatory_region_ablation',<br />'regulatory_region_fusion',<br />'regulatory_region_amplification',<br />'regulatory_region_translocation',<br />'feature_elongation',<br />'feature_truncation'<br /> )</td><td></td><td>The consequence(s) of the variant allele on this regulatory feature.<br /> The list of consequence descriptions is available <a href="predicted_data.html#consequences">here</a>.</td><td>key: consequence_type_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_feature">variation_feature</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FF4DC8"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="transcript_variation" style="font-size:11pt;font-weight:bold">transcript_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_transcript_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('transcript_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table relates a single allele of a variation_feature to a transcript (see Core documentation). It contains the consequence of the allele e.g. intron_variant, non_synonymous_codon, stop_lost etc, along with the change in amino acid in the resulting protein if applicable.</p>

  <div id="div_transcript_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>transcript_variation_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>feature_stable_id</b></td><td>varchar(128)</td><td>NULL</td><td>Foreign key to core databases. Unique stable id of related transcript.</td><td>key: somatic_feature_idx</td></tr>
      <tr class="bg1"><td><b>variation_feature_id</b></td><td>int(11)</td><td></td><td>Foreign key references to the <a href="#variation_feature">variation_feature</a> table.</td><td>key: variation_feature_idx</td></tr>
      <tr class="bg2"><td><b>allele_string</b></td><td>text</td><td></td><td>Shows the reference sequence and variant sequence of this allele</td><td></td></tr>
      <tr class="bg1"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>Flags if the associated variation is known to be somatic</td><td>key: somatic_feature_idx</td></tr>
      <tr class="bg2"><td><b>consequence_types</b></td><td>set (<br />'splice_acceptor_variant',<br />'splice_donor_variant',<br />'stop_lost',<br />'coding_sequence_variant',<br />'missense_variant',<br />'stop_gained',<br />'synonymous_variant',<br />'frameshift_variant',<br />'nc_transcript_variant',<br />'non_coding_exon_variant',<br />'mature_miRNA_variant',<br />'NMD_transcript_variant',<br />'5_prime_UTR_variant',<br />'3_prime_UTR_variant',<br />'incomplete_terminal_codon_variant',<br />'intron_variant',<br />'splice_region_variant',<br />'downstream_gene_variant',<br />'upstream_gene_variant',<br />'initiator_codon_variant',<br />'stop_retained_variant',<br />'inframe_insertion',<br />'inframe_deletion',<br />'transcript_ablation',<br />'transcript_fusion',<br />'transcript_amplification',<br />'transcript_translocation',<br />'TFBS_ablation',<br />'TFBS_fusion',<br />'TFBS_amplification',<br />'TFBS_translocation',<br />'regulatory_region_ablation',<br />'regulatory_region_fusion',<br />'regulatory_region_amplification',<br />'regulatory_region_translocation',<br />'feature_elongation',<br />'feature_truncation'<br /> )</td><td></td><td>The consequence(s) of the variant allele on this transcript.<br /> The list of consequence descriptions is available <a href="predicted_data.html#consequences">here</a>. </td><td>key: consequence_type_idx</td></tr>
      <tr class="bg1"><td><b>cds_start</b></td><td>int(11)</td><td></td><td>The start position of variation in cds coordinates.</td><td></td></tr>
      <tr class="bg2"><td><b>cds_end</b></td><td>int(11)</td><td></td><td>The end position of variation in cds coordinates.</td><td></td></tr>
      <tr class="bg1"><td><b>cdna_start</b></td><td>int(11)</td><td></td><td>The start position of variation in cdna coordinates.</td><td></td></tr>
      <tr class="bg2"><td><b>cdna_end</b></td><td>int(11)</td><td></td><td>The end position of variation in cdna coordinates.</td><td></td></tr>
      <tr class="bg1"><td><b>translation_start</b></td><td>int(11)</td><td></td><td>The start position of variation on peptide.</td><td></td></tr>
      <tr class="bg2"><td><b>translation_end</b></td><td>int(11)</td><td></td><td>The end position of variation on peptide.</td><td></td></tr>
      <tr class="bg1"><td><b>distance_to_transcript</b></td><td>int(11)</td><td></td><td>Only for upstream or downstream variants, it gives the distance from the start or the end of the transcript</td><td></td></tr>
      <tr class="bg2"><td><b>codon_allele_string</b></td><td>text</td><td></td><td>The reference and variant codons</td><td></td></tr>
      <tr class="bg1"><td><b>pep_allele_string</b></td><td>text</td><td></td><td>The reference and variant peptides</td><td></td></tr>
      <tr class="bg2"><td><b>hgvs_genomic</b></td><td>text</td><td></td><td>HGVS representation of this allele with respect to the genomic sequence</td><td></td></tr>
      <tr class="bg1"><td><b>hgvs_transcript</b></td><td>text</td><td></td><td>HGVS representation of this allele with respect to the [coding or non-coding] transcript</td><td></td></tr>
      <tr class="bg2"><td><b>hgvs_protein</b></td><td>text</td><td></td><td>HGVS representation of this allele with respect to the protein</td><td></td></tr>
      <tr class="bg1"><td><b>polyphen_prediction</b></td><td>enum('unknown', 'benign', 'possibly damaging', 'probably damaging')</td><td>NULL</td><td>The PolyPhen prediction for the effect of this allele on the protein</td><td></td></tr>
      <tr class="bg2"><td><b>polyphen_score</b></td><td>float</td><td>NULL</td><td>The PolyPhen score corresponding to the prediction </td><td></td></tr>
      <tr class="bg1"><td><b>sift_prediction</b></td><td>enum('tolerated', 'deleterious')</td><td>NULL</td><td>The SIFT prediction for the effect of this allele on the protein </td><td></td></tr>
      <tr class="bg2"><td><b>sift_score</b></td><td>float</td><td>NULL</td><td>The SIFT score corresponsing to this prediction</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_feature">variation_feature</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #FFD700">
  <h2 style="display:inline;color:#000">Variation set tables</h2>
</div>
<p style="width:800px">These tables define the variation and structural variation set data. The complete list of variation sets with their descriptions is available <a href="data_description.html#variation_sets">here</a>.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FFD700"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="variation_set" style="font-size:11pt;font-weight:bold">variation_set</span></td>
      <td style="width:180px;text-align:right"><a id="a_variation_set" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_set')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table containts the name of sets and subsets of variations stored in the database. It usually represents the name of the project or subproject where a group of variations has been identified.</p>

  <div id="div_variation_set" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_set_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>VARCHAR(255)</td><td></td><td>Name of the set e.g. "Phenotype-associated variations".</td><td>key: name_idx</td></tr>
      <tr class="bg1"><td><b>description</b></td><td>TEXT</td><td></td><td>Description of the set.</td><td></td></tr>
      <tr class="bg2"><td><b>short_name_attrib_id</b></td><td>INT(10)</td><td>NULL</td><td>Foreign key references to the <a href="#attrib">attrib</a> table. Short name used for web purpose.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_set_variation">variation_set_variation</a></li>
    <li><a href="#variation_set_structure">variation_set_structure</a></li>
    <li><a href="#attrib">attrib</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FFD700"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="variation_set_structural_variation" style="font-size:11pt;font-weight:bold">variation_set_structural_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_variation_set_structural_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_set_structural_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">A table for mapping structural variations to variation_sets.</p>

  <div id="div_variation_set_structural_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>structural_variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#structural_variation">structural_variation</a> table.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_set_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_set">variation_set</a> table.</td><td>primary key</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#structural_variation">structural_variation</a></li>
    <li><a href="#variation_set">variation_set</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FFD700"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="variation_set_structure" style="font-size:11pt;font-weight:bold">variation_set_structure</span></td>
      <td style="width:180px;text-align:right"><a id="a_variation_set_structure" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_set_structure')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores hierarchical relationships between variation sets by relating them as variation sets and variation subsets.</p>

  <div id="div_variation_set_structure" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_set_super</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_set">variation_set</a> table.</td><td>primary key<br />key: sub_idx</td></tr>
      <tr class="bg2"><td><b>variation_set_sub</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_set">variation_set</a> table.</td><td>primary key<br />key: sub_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_set">variation_set</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#FFD700"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="variation_set_variation" style="font-size:11pt;font-weight:bold">variation_set_variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_variation_set_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_set_variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">A table for mapping variations to variation_sets.</p>

  <div id="div_variation_set_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation">variation</a> table.</td><td>primary key<br />key: variation_set_idx</td></tr>
      <tr class="bg2"><td><b>variation_set_id</b></td><td>int(10)</td><td></td><td>Primary key. Foreign key references to the <a href="#variation_set">variation_set</a> table.</td><td>primary key<br />key: variation_set_idx</td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation">variation</a></li>
    <li><a href="#variation_set">variation_set</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #000">
  <h2 style="display:inline;color:#000">Variation tables</h2>
</div>
<p style="width:800px">These tables define the central variation data.</p>
<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="allele" style="font-size:11pt;font-weight:bold">allele</span></td>
      <td style="width:180px;text-align:right"><a id="a_allele" style="cursor:pointer;text-decoration:underline" onclick="show_hide('allele')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores information about each of a variation's alleles, along with population frequencies.</p>

  <div id="div_allele" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>allele_id</b></td><td>int(11)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_id</b></td><td>int(11)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
      <tr class="bg1"><td><b>subsnp_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the <a href="#subsnp_handle">subsnp_handle</a> table.</td><td>key: subsnp_idx</td></tr>
      <tr class="bg2"><td><b>allele_code_id</b></td><td>int(11)</td><td></td><td>Foreign key reference to <a href="#allele_code">allele_code</a> table.</td><td></td></tr>
      <tr class="bg1"><td><b>frequency</b></td><td>float</td><td>NULL</td><td>Frequency of this allele in the sample.</td><td></td></tr>
      <tr class="bg2"><td><b>sample_id</b></td><td>int(11)</td><td>NULL</td><td>Foreign key references to the <a href="#sample">sample</a> table.</td><td>key: sample_idx</td></tr>
      <tr class="bg1"><td><b>count</b></td><td>int(11)</td><td>NULL</td><td>Number of individuals in the sample where this allele is found.</td><td></td></tr>
      <tr class="bg2"><td><b>frequency_submitter_handle</b></td><td>int(10)</td><td>NULL</td><td>dbSNP handle for submitter of frequency data [may be different to submitter of observed variant]</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation">variation</a></li>
    <li><a href="#population">population</a></li>
    <li><a href="#subsnp_handle">subsnp_handle</a></li>
    <li><a href="#allele_code">allele_code</a></li>
    <li><a href="#sample">sample</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="variation" style="font-size:11pt;font-weight:bold">variation</span></td>
      <td style="width:180px;text-align:right"><a id="a_variation" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This is the schema's generic representation of a variation, defined as a genetic feature that varies between individuals of the same species.  The most common type is the single nucleotide variation (SNP) though the schema also accommodates copy number variations (CNVs) and structural variations (SVs).<br /> In Ensembl, a variation is defined by its flanking sequence rather than its mapped location on a chromosome; a variation may in fact have multiple mappings across a genome,  although this fails our <a href="data_description.html#quality_control">Quality Control</a>.<br />  This table stores a variation's name (commonly an ID of the form e.g. rs123456, assigned by dbSNP), along with a validation status and ancestral (or reference) allele.</p>

  <div id="div_variation" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#source">source</a> table.</td><td>key: source_idx</td></tr>
      <tr class="bg1"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the variation. e.g. "rs1333049".</td><td>unique</td></tr>
      <tr class="bg2"><td><b>validation_status</b></td><td>SET('cluster','freq',<br />'submitter','doublehit',<br />'hapmap','1000Genome',<br />'failed','precious')</td><td></td><td>Variant discovery method and validation from dbSNP. See the status descriptions <a href="data_description.html#validation_status">here</a>.</td><td></td></tr>
      <tr class="bg1"><td><b>ancestral_allele</b></td><td>varchar(255)</td><td>NULL</td><td>Taken from dbSNP to show ancestral allele for the variation.</td><td></td></tr>
      <tr class="bg2"><td><b>flipped</b></td><td>tinyint(1)</td><td>NULL</td><td>This is set to 1 if the variant is flipped from the negative to the positive strand during import.</td><td></td></tr>
      <tr class="bg1"><td><b>class_attrib_id</b></td><td>int(10)</td><td>0</td><td>Class of the variation, key into the <a href="#attrib">attrib</a> table.<br /> The list of variation classes is available <a href="data_description.html#classes">here</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>flags whether this variation is known to be somatic or not</td><td></td></tr>
      <tr class="bg1"><td><b>minor_allele</b></td><td>char(50)</td><td>NULL</td><td>The minor allele of this variant, as reported by dbSNP</td><td></td></tr>
      <tr class="bg2"><td><b>minor_allele_freq</b></td><td>float</td><td>NULL</td><td>The 'global' frequency of the minor allele of this variant, as reported by dbSNP</td><td></td></tr>
      <tr class="bg1"><td><b>minor_allele_count</b></td><td>int(10)</td><td>NULL</td><td>The number of samples the minor allele of this variant is found in, as reported by dbSNP</td><td></td></tr>
      <tr class="bg2"><td><b>clinical_significance_attrib_id</b></td><td>int(10)</td><td>NULL</td><td>An attrib_id identifying the clinical significance of this variant, as reported by dbSNP.<br />  The list of clinical significances is available <a href="data_description.html#clin_significance">here</a>.</td><td></td></tr>
      <tr class="bg1"><td><b>evidence</b></td><td>SET('Multiple_observations', 'Frequency','HapMap','1000Genomes', 'Cited')</td><td></td><td>A summary of the evidence supporting a variant as a guide to its potential reliability.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation_synonym">variation_synonym</a></li>
    <li><a href="#failed_variation">failed_variation</a></li>
    <li><a href="#variation_feature">variation_feature</a></li>
    <li><a href="#allele">allele</a></li>
    <li><a href="#individual_genotype_multiple_bp">individual_genotype_multiple_bp</a></li>
    <li><a href="#compressed_genotype_var">compressed_genotype_var</a></li>
    <li><a href="#attrib">attrib</a></li>
    <li><a href="#source">source</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="variation_feature" style="font-size:11pt;font-weight:bold">variation_feature</span></td>
      <td style="width:180px;text-align:right"><a id="a_variation_feature" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_feature')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table represents mappings of variations to genomic locations. It stores an allele string representing the different possible alleles that are found at that locus e.g. "A/T" for a SNP, as well as a "worst case" consequence of the mutation. It also acts as part of the relationship between variations and transcripts.</p>

  <div id="div_variation_feature" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_feature_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>seq_region_id</b></td><td>int(10)</td><td></td><td>Foreign key references <a href="#seq_region">seq_region</a> in core db. Refers to the seq_region which this variant is on, which may be a chromosome, a clone, etc...</td><td>key: pos_idx</td></tr>
      <tr class="bg1"><td><b>seq_region_start</b></td><td>int</td><td></td><td>The start position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
      <tr class="bg2"><td><b>seq_region_end</b></td><td>int</td><td></td><td>The end position of the variation on the <a href="#seq_region">seq_region</a>.</td><td>key: pos_idx</td></tr>
      <tr class="bg1"><td><b>seq_region_strand</b></td><td>tinyint</td><td></td><td>The orientation of the variation on the <a href="#seq_region">seq_region</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the <a href="#variation">variation</a> table.</td><td>key: variation_idx</td></tr>
      <tr class="bg1"><td><b>allele_string</b></td><td>varchar(50000)</td><td></td><td>This is a denormalised string taken from the alleles in the allele table associated with this variation. The reference allele (i.e. one on the reference genome comes first).</td><td></td></tr>
      <tr class="bg2"><td><b>variation_name</b></td><td>varchar(255)</td><td></td><td>A denormalisation taken from the variation table. This is the name or identifier that is used for displaying the feature.</td><td></td></tr>
      <tr class="bg1"><td><b>map_weight</b></td><td>int</td><td></td><td>The number of times that this variation has mapped to the genome. This is a denormalisation as this particular feature is one example of a mapped location. This can be used to limit the the features that come back from a query.</td><td></td></tr>
      <tr class="bg2"><td><b>flags</b></td><td>SET('genotyped')</td><td></td><td>Flag to filter the selection of variations.</td><td></td></tr>
      <tr class="bg1"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the source table.</td><td></td></tr>
      <tr class="bg2"><td><b>validation_status</b></td><td>SET(<br />'cluster',<br />'freq',<br />'submitter',<br />'doublehit',<br />'hapmap',<br />'1000Genome',<br />'precious'<br /> )</td><td></td><td>Variant discovery method and validation from dbSNP. See the status descriptions <a href="data_description.html#validation_status">here</a>.</td><td></td></tr>
      <tr class="bg1"><td><b>consequence_types</b></td><td>SET (<br />'intergenic_variant',<br />'splice_acceptor_variant',<br />'splice_donor_variant',<br />'stop_lost',<br />'coding_sequence_variant',<br />'missense_variant',<br />'stop_gained',<br />'synonymous_variant',<br />'frameshift_variant',<br />'nc_transcript_variant',<br />'non_coding_exon_variant',<br />'mature_miRNA_variant',<br />'NMD_transcript_variant',<br />'5_prime_UTR_variant',<br />'3_prime_UTR_variant',<br />'incomplete_terminal_codon_variant',<br />'intron_variant',<br />'splice_region_variant',<br />'downstream_gene_variant',<br />'upstream_gene_variant',<br />'initiator_codon_variant',<br />'stop_retained_variant',<br />'inframe_insertion',<br />'inframe_deletion',<br />'transcript_ablation',<br />'transcript_fusion',<br />'transcript_amplification',<br />'transcript_translocation',<br />'TFBS_ablation',<br />'TFBS_fusion',<br />'TFBS_amplification',<br />'TFBS_translocation',<br />'regulatory_region_ablation',<br />'regulatory_region_fusion',<br />'regulatory_region_amplification',<br />'regulatory_region_translocation',<br />'feature_elongation',<br />'feature_truncation',<br />'regulatory_region_variant',<br />'TF_binding_site_variant'<br /> )</td><td>'intergenic_variant'</td><td>The SO term(s) of all unique observed consequence types of this variation feature.<br /> The list of consequence descriptions is available <a href="predicted_data.html#consequences">here</a>.</td><td>key: consequence_type_idx</td></tr>
      <tr class="bg2"><td><b>variation_set_id</b></td><td>SET (<br />'1','2','3','4','5','6','7','8',<br />'9','10','11','12','13','14','15','16',<br />'17','18','19','20','21','22','23','24',<br />'25','26','27','28','29','30','31','32',<br />'33','34','35','36','37','38','39','40',<br />'41','42','43','44','45','46','47','48',<br />'49','50','51','52','53','54','55','56',<br />'57','58','59','60','61','62','63','64'<br /> )</td><td>''</td><td>The variation feature can belong to a <a href="#variation_set">variation_set</a>.</td><td>key: variation_set_idx</td></tr>
      <tr class="bg1"><td><b>class_attrib_id</b></td><td>int(10)</td><td>0</td><td>Class of the variation, key in the <a href="#attrib">attrib</a> table.<br /> The list of variation classes is available <a href="data_description.html#classes">here</a>.</td><td></td></tr>
      <tr class="bg2"><td><b>somatic</b></td><td>tinyint(1)</td><td>0</td><td>Flags whether this variation_feature is somatic or germline</td><td></td></tr>
      <tr class="bg1"><td><b>minor_allele</b></td><td>char(50)</td><td>NULL</td><td>The minor allele of this variant, as reported by dbSNP</td><td></td></tr>
      <tr class="bg2"><td><b>minor_allele_freq</b></td><td>float</td><td>NULL</td><td>The 'global' frequency of the minor allele of this variant, as reported by dbSNP</td><td></td></tr>
      <tr class="bg1"><td><b>minor_allele_count</b></td><td>int(10)</td><td>NULL</td><td>The number of samples the minor allele of this variant is found in, as reported by dbSNP</td><td></td></tr>
      <tr class="bg2"><td><b>alignment_quality</b></td><td>double</td><td>NULL</td><td>Quality of alignment for variants mapped by flanks rather than position justified.</td><td></td></tr>
      <tr class="bg1"><td><b>evidence</b></td><td>SET('Multiple_observations', 'Frequency','HapMap','1000Genomes', 'Cited')</td><td></td><td>A summary of the evidence supporting a variant as a guide to its potential reliability.</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#variation">variation</a></li>
    <li><a href="#tagged_variation_feature">tagged_variation_feature</a></li>
    <li><a href="#transcript_variation">transcript_variation</a></li>
    <li><a href="#seq_region">seq_region</a></li>
    <li><a href="#attrib">attrib</a></li>
    <li><a href="#variation_set">variation_set</a></li>
  </ul>


<br />
  <table style="border: 2px groove #CCCCCC;background-color:#FAFAFF">
    <tr style="vertical-align:middle">
      <td style="padding:0px;width:10px;background-color:#000"></td>
      <td style="width:2px"></td>
      <td style="width:420px;text-align:left;height:10px"><span id="variation_synonym" style="font-size:11pt;font-weight:bold">variation_synonym</span></td>
      <td style="width:180px;text-align:right"><a id="a_variation_synonym" style="cursor:pointer;text-decoration:underline" onclick="show_hide('variation_synonym')">Show</a> columns
        &nbsp;<b>|</b>&nbsp;<a href="#top">[Back to top]</a></td>
    </tr>
  </table>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table allows for a variation to have multiple IDs, generally given by multiple sources.</p>

  <div id="div_variation_synonym" style="display:none">
    <table style="border:1px outset #222222">
      <tr class="bg3 center"><th style="width:180px">Column</th><th style="width:150px">Type</th><th style="width:100px">Default value</th><th style="width:400px">Description</th><th style="width:150px">Index</th></tr>
      <tr class="bg1"><td><b>variation_synonym_id</b></td><td>int(10)</td><td></td><td>Primary key, internal identifier.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>variation_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the variation table.</td><td>key: variation_idx</td></tr>
      <tr class="bg1"><td><b>subsnp_id</b></td><td>int(15)</td><td></td><td>Foreign key references to the subsnp_handle table.</td><td>key: subsnp_idx</td></tr>
      <tr class="bg2"><td><b>source_id</b></td><td>int(10)</td><td></td><td>Foreign key references to the source table.</td><td>unique<br />key: source_idx</td></tr>
      <tr class="bg1"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Name of the synonym variation. e.g. 'rs1333049'. The corresponding variation ID of this variation is different from the one stored in the column variation_id.</td><td>unique</td></tr>
      <tr class="bg2"><td><b>moltype</b></td><td>varchar(50)</td><td></td><td>...</td><td></td></tr>
    </table>
  </div>
  <p style="padding-top:5px"><b>See also:</b></p>
  <ul>
    <li><a href="#source">source</a></li>
    <li><a href="#variation">variation</a></li>
    <li><a href="#subsnp_handle">subsnp_handle</a></li>
  </ul>



</body>
</html>
