<html>
<head>
	<meta name="order" content="2" />
  <meta name="navigation" content="Database description">

	<title>Database Description</title>
</head>

<body>

<h1>Ensembl Variation Database Description</h1>

<a href="/info/docs/variation/index.html">About Variation Data</a> | 
<a href="/info/docs/variation/database.html">Database Description</a> |
<a href="/info/docs/variation/sources_documentation.html">Variation Sources</a> |
<a href="/info/docs/variation/variation_schema.html">Variation Tables Description</a> |
<a href="/info/docs/api/variation/index.html">Perl API</a>

<h2 id="db_desc">Database Description</h2>

<p>
The Ensembl Variation database stores two types of variation data, depending it comes from an external source or data it is calculated on site.
</p>

<ul>
<li>Data imported from external sources (dbSNP, Sanger,...):
  <ul>
  	<li>SNPs, in-dels</li>
		<li>Structural variations</li>
		<li>Copy number variations</li>
  	<li>Locations for variations</li>
  	<li>Alleles</li>
  	<li>Populations</li>
  	<li>Genotypes</li>
		<li>Phenotypes</li>
  </ul>
</li>

<li>Calculated data:
  <ul>
  	<li>Consequence of variations in transcripts (synonymous, non-synonymous,...)</li>
  	<li>Linkage disequilibrium information</li>
  	<li>Tagged variations</li>
  	<li>Read coverage data</li>
  </ul>
</li>

</ul>


<br />
<h2 id="db_schema">Database Schema</h2>

<p>
The tables of the database are described in the <a href="/info/docs/variation/variation_schema.html">Variation Tables Description</a> page.<br />
The database schema is described in the following <a href="/info/docs/variation/variation-database-schema.pdf">pdf</a>. 
</p>


<br />
<h2 id="db_load">Database Load</h2>
<p>
The variations databases can be loaded using dumped files from the <a href="/info/data/ftp/index.html">Ensembl
FTP</a>.<br />
For Ensembl 61, it takes a couple of minutes to load the largest tables for Human on our servers, e.g:
</p>
<table id="loading_times" class="ss" style="width:350px;">
<tr><th>Table</th><th>Loading time</th></tr>
<tr class="bg1"><td>variation</td><td style="text-align:right">24 minutes</td></tr>
<tr class="bg2"><td>variation_feature</td><td style="text-align:right">13 minutes</td></tr>
<tr class="bg1"><td>flanking_sequence</td><td style="text-align:right">3 minutes</td></tr>
<tr class="bg2"><td>compressed_genotype_single_bp</td><td style="text-align:right">13 minutes</td></tr>
<tr class="bg1"><td>population_genotype</td><td style="text-align:right">30 minutes</td></tr>
</table>
<p>
The load of the largest table in the Human variation database (<b>allele</b> table) takes almost 3 hours.
</p>
<p>
See below some settings of our server:
</p>
<table id="mysql_settings" class="ss" style="width:350px;">
<tr><th>Variable</th><th>Value</th></tr>
<tr class="bg1"><td>myisam_data_pointer_size</td><td>6</td></tr>
<tr class="bg2"><td>myisam_max_sort_file_size</td><td>9223372036853727232</td></tr>
<tr class="bg1"><td>myisam_recover_options</td><td>OFF</td></tr>
<tr class="bg2"><td>myisam_repair_threads</td><td>1</td></tr>
<tr class="bg1"><td>myisam_sort_buffer_size</td><td>67108864</td></tr>
<tr class="bg2"><td>myisam_stats_method</td><td>nulls_unequal</td></tr>
<tr class="bg1"><td>myisam_use_mmap</td><td>OFF</td></tr>
</table>


<br />
<h2 id="support">Ensembl Software Support</h2>

<p>
Ensembl is an open project and we would like to encourage correspondence and discussions on any subject on any aspect of Ensembl.
Please see the
<a href="http://www.ensembl.org/info/about/contact/">Ensembl Contacts</a> page for suitable options getting in touch with us.
</p>

</body>
</html>
