<html>
<head>
<meta name="order" content="1" />
<title>Variation</title>
</head>

<body>

<h1>Ensembl Variation</h1>

<a href="/info/docs/variation/index.html">About Variation Data</a> | 
<a href="/info/docs/variation/database.html">Database Description</a> |
<a href="/info/docs/variation/sources_documentation.html">Variation Sources</a> |
<a href="/info/docs/variation/variation_schema.html">Variation Tables Description</a> |
<a href="/info/docs/api/variation/index.html">Perl API</a> |
<a href="/info/docs/variation/vep/index.html">Variant Effect Predictor</a>

<br /><br />
<p>
Ensembl stores both structural variants and sequence variants including
Single Nucleotide Polymorphisms (SNPs), insertions, deletions and somatic mutations
for several species and makes this data available on the website, via the Perl API,
and provides direct access to the underlying databases.
</p>

<p>
This data is also integrated with other sections of Ensembl, in particular we predict
the effects of variants on the Ensembl transcripts for each species.
</p>

<h2 id="sources"> Variation species and data sources </h2>
<p>
Ensembl stores variation data for the following species, but note that users can still use the 
<a href="#VEP">Variant Effect Predictor</a> on species for which we do not currently have 
a variation database.
</p>
<table><tr><td>
<ul>
	<li><i>Bos taurus</i> <a href="sources_documentation.html#bos_taurus">[sources]</a></li>
	<li><i>Canis familiaris</i> <a href="sources_documentation.html#canis_familiaris">[sources]</a></li>
	<li><i>Danio rerio</i> <a href="sources_documentation.html#danio_rerio">[sources]</a></li>
	<li><i>Drosophila melanogaster</i> <a href="sources_documentation.html#drosophila_melanogaster">[sources]</a></li>
	<li><i>Equus caballus</i> <a href="sources_documentation.html#equus_caballus">[sources]</a></li>
	<li><i>Felis catus</i> <a href="sources_documentation.html#felis_catus">[sources]</a></li>
	<li><i>Gallus gallus</i> <a href="sources_documentation.html#gallus_gallus">[sources]</a></li>
	<li><i>Homo sapiens</i> <a href="sources_documentation.html#homo_sapiens">[sources]</a></li>
	<li><i>Monodelphis domestica</i> <a href="sources_documentation.html#monodelphis_domestica">[sources]</a></li>
</ul>
</td><td>
<ul>
	<li><i>Mus musculus</i> <a href="sources_documentation.html#mus_musculus">[sources]</a></li>
	<li><i>Ornithorhynchus anatinus</i> <a href="sources_documentation.html#ornithorhynchus_anatinus">[sources]</a></li>
	<li><i>Pan troglodytes</i> <a href="sources_documentation.html#pan_troglodytes">[sources]</a></li>
	<li><i>Pongo abelii</i> <a href="sources_documentation.html#pongo_abelii">[sources]</a></li>
	<li><i>Rattus norvegicus</i> <a href="sources_documentation.html#rattus_norvegicus">[sources]</a></li>
	<li><i>Saccharomyces cerevisiae</i> <a href="sources_documentation.html#saccharomyces_cerevisiae">[sources]</a></li>
	<li><i>Sus scrofa</i> <a href="sources_documentation.html#sus_scrofa">[sources]</a></li>
	<li><i>Taeniopygia guttata</i> <a href="sources_documentation.html#taeniopygia_guttata">[sources]</a></li>
	<li><i>Tetraodon nigroviridis</i> <a href="sources_documentation.html#tetraodon_nigroviridis">[sources]</a></li>
</ul>
</td></tr></table>

<p>
The majority of variants are imported from NCBI <a rel="external"
href="http://www.ncbi.nlm.nih.gov/projects/SNP/">dbSNP</a>. The data is
imported when it is released by dbSNP and incorporated into the next
Ensembl release. If dbSNP releases the data on a different assembly,
Ensembl will remap the variant positions onto the current assembly.
Data from projects like the <a rel="external" href="http://hapmap.ncbi.nlm.nih.gov/">HapMap Project</a> and <a
rel="external" href="http://www.1000genomes.org/">1000 Genomes Project</a> is
imported once it has been submitted to dbSNP.
</p>

<p>
Ensembl also includes data from other <a href="sources_documentation.html">sources</a>. To view data from these sources in the
browser go to a species Location page (e.g. <a href="/Homo_sapiens/Location/View?r=6:133017695-133161157">for human</a>), and click on
the <i>'Configure this page'</i> link on the left-hand side. The <i>'Germline
variation'</i> and <i>'Somatic mutations'</i> sections contain a track list  of
all sources of variation data for that species.
</p>


<br />
<h2 id="displays">Variation displays</h2>
<p>
Variation data can be viewed in the browser through pages such as:
</p>

<ul>
<li><strong>Gene:</strong> Variation Table and Variation Image (for all variations in a gene) 
e.g. for example <a href="/Homo_sapiens/Gene/Variation_Gene/Table?db=core;g=ENSG00000159197">for all variants in KCNE2</a>. 
Structural Variation to see all structural variants overlapping the gene.</li>
<li><strong>Transcript:</strong> Population comparison, Comparison
image (for comparing variants in a transcript across different individual or strain
sequences) e.g. <a href="/Mus_musculus/Transcript/Population/Image?g=ENSMUSG00000041143;r=4:138528820-138586500;t=ENSMUST00000084208"> compare Tmco4 in different mouse strains </a> </li>
<li> <strong> Transcript: </strong> Sequence, protein: list of the coding variants in protein coordinates.</li>
<li><strong>Location:</strong> Region in Detail (Variations can be
drawn using &quot;Configure this page&quot; at the left. The menu allows
display of information in Ensembl databases along with external
sources in DAS format such as <a rel="external"
href="http://projects.tcag.ca/variation/">DGV loci</a>.)</li>
</ul>

<p>
Clicking on any variation on an Ensembl page will open a Variation tab
with information about the flanking sequence and source for the
selected variation. Links to linkage disequilibrium (LD) plots,
phenotype information (for human) from EGA, OMIM and <a rel="external"
href="http://www.genome.gov/">NHGRI</a> and Ensembl genes and
transcripts that include the variation can be found at the left of
this tab. You may also view multiple genome alignments of various
species, highlighting the
variation. <a href="/info/docs/compara/analyses.html#ancestral"
class="cp-external">Ancestral sequences</a> are included in this
display. 
</p>

<p>
Variation information can also be accessed
using <a href="/info/data/biomart.html"
class="cp-external">BioMart</a> (gene or variation database), and
the <a href="/info/docs/api/index.html" class="cp-external">Perl
API</a> (<a href="/info/docs/api/variation/index.html"
class="cp-external">variation API</a>).
</p>

<br />

<h2 id="classes">Variation classes</h2>

<p>
We call the class of a variation according to its component alleles and its mapping 
to the reference genome, and then display this information on the website. Internally 
we use <a href="http://www.sequenceontology.org/">Sequence Ontology</a> terms, but we 
map these to our own 'display' terms where common usage differs from the SO 
definition (e.g. our term SNP is closer to the SO term SNV). All the classes we call, 
along with their equivalent SO term are shown in the table below. We also differentiate 
somatic mutations from germline variations in the display term, prefixing the term with
'somatic'. API users can fetch either the SO term or the display term.
</p>

<br />

<table id="variation_classes" class="ss" style="width:50%;">
<tr>
	<th>Ensembl term</th>
	<th>SO term</th>
	<th>SO accession</th>
</tr>
<tr>
	<td>SNP</td>
	<td rowspan="2">SNV</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001483">SO:0001483</a></td>
</tr>
<tr>
	<td>somatic_SNV</td>
</tr>
<tr class="bg2">
	<td>substitution</td>
	<td rowspan="2">substitution</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:1000002">SO:1000002</a></td>
</tr>
<tr class="bg2">
	<td>somatic_substitution</td>
</tr>
<tr>
	<td>CNV</td>
	<td rowspan="2">copy_number_variation</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001019">SO:0001019</a></td>
</tr>
<tr>
	<td>somatic_CNV</td>
</tr>
<tr class="bg2">
	<td>insertion</td>
	<td rowspan="2">insertion</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000667">SO:0000667</a></td>
</tr>
<tr class="bg2">
	<td>somatic_insertion</td>
</tr>
<tr>
	<td>deletion</td>
	<td rowspan="2">deletion</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000159">SO:0000159</a></td>
</tr>
<tr>
	<td>somatic_deletion</td>
</tr>
<tr class="bg2">
	<td>indel</td>
	<td rowspan="2">indel</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:1000032">SO:1000032</a></td>
</tr>
<tr class="bg2">
	<td>somatic_indel</td>
</tr>
<tr>
	<td>tandem_repeat</td>
	<td rowspan="2">tandem_repeat</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0000705">SO:0000705</a></td>
</tr>
<tr>
	<td>somatic_tandem_repeat</td>
</tr>
<tr class="bg2">
	<td>sequence_alteration</td>
	<td rowspan="2">sequence_alteration</td>
	<td rowspan="2"><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001059">SO:0001059</a></td>
</tr>
<tr class="bg2">
	<td>somatic_sequence_alteration</td>
</tr>
</table>


<h3 id="indel">Insertion and Deletion coordinates </h3>

<p> In Ensembl, an insertion is indicated by start coordinate = end coordinate +
1. For example, an insertion of 'C' between nucleotides 12600 and
12601 on the forward strand is indicated with start and end coordinates as
follows:</p>

<pre>
   12601     12600   
</pre>
<p>A deletion is indicated by the exact nucleotide coordinates. For
example, a three base pair deletion of nucleotides 12600, 12601, and
12602 of the reverse strand will have start and end coordinates of :</p>

<pre>
   12600     12602    
</pre>

<br />
<h2 id="consequences">Predicted variation consequences</h2>

<p>
For each variation that is mapped to the reference genome, we identify any
Ensembl transcripts that overlap the variation and use a rule-based approach to
predict the effect that each allele of the variation may have on the transcript.
The set of consequence terms, defined by the <a rel="external"
href="http://www.sequenceontology.org">Sequence Ontology</a> (SO), that can be
currently assigned to each combination of an allele and a transcript is shown
in the table below. Note that each allele of each variation may have a different
effect in different transcripts. 
</p>

<p> 
This approach is applied to all germline variations and somatic mutations
stored in the Ensembl variation databases (though we do not currently calculate
consequences for structural variants). The resulting consequence type calls,
along with information determined as part of the process, such as the cDNA and
CDS coordinates, and the affected codons and amino acids in coding transcripts,
are stored in the variation database and displayed on the website. 
</p>

<p>
Prior to release 62 we used our own internal terms to describe the consequence
types, and we continue to use these by default on the website. You can opt to
see SO terms on some variation views using 'Configure this page', and 
<a href="#VEP">VEP</a> and API users can choose which terms to use. The SO terms 
are more specific than our
display terms but we map all display terms to one or more SO terms, as shown in
the table below. Where the NCBI have an equivalent term we also include it in
this table (and again you can opt to use NCBI terms on some views via 'Configure
this page' and in the VEP and API).
</p>

<p>We follow the SO definition of the term as closely as possible, but in some
cases (e.g. non_synonymous_codon) we have slightly generalised the definition to
include variants that affect multiple codons as well as a single one. We are
working with the SO to refine these definitions. The descriptions shown in the
table are our own, not the SO definitions.
</p>

<p>
The terms in the table are shown in order of severity as estimated by
Ensembl, and this ordering is used on the website summary views. This ordering
is necessarily subjective and API and VEP users can always get the full set of
consequences for each allele and make their own severity judgement.
</p>

<p>
A <a href="consequences.jpg">diagram</a> showing the location of each display 
term relative to the transcript structure is available.
</p>

<br />

<table id="consequence_type_table" class="ss">
<tr>
	<th>Ensembl term</th>
	<th>Ensembl description</th>
	<th>SO term</th>
	<th>SO accession</th>
	<th>NCBI term</th>
</tr>
<tr>
	<td rowspan="2">Essential splice site</td>
	<td rowspan="2">In the first 2 or the last 2 basepairs of an intron</td>
	<td>splice_acceptor_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001574">SO:0001574</a></td>
	<td>splice-3</td>
</tr>
<tr>
	<td>splice_donor_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001575">SO:0001575</a></td>
	<td>splice-5</td>
</tr>
<tr class="bg2">
	<td>Stop gained</td>
	<td>In coding sequence, resulting in the gain of a stop codon</td>
	<td>stop_gained</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001587">SO:0001587</a></td>
	<td>nonsense</td>
</tr>
<tr>
	<td>Stop lost</td>
	<td>In coding sequence, resulting in the loss of a stop codon</td>
	<td>stop_lost</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001578">SO:0001578</a></td>
	<td>-</td>
</tr>
<tr class="bg2">
	<td>Complex in/del</td>
	<td>Insertion or deletion that spans an exon/intron or coding sequence/UTR border</td>
	<td>complex_change_in_transcript</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001577">SO:0001577</a></td>
	<td>-</td>
</tr>
<tr>
	<td rowspan="4">Non-synonymous coding</td>
	<td rowspan="4">In coding sequence and results in an amino acid change in the encoded peptide sequence</td>
	<td>initiator_codon_change</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001582">SO:0001582</a></td>
	<td>-</td>
</tr>
<tr>
	<td>inframe_codon_gain</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001651">SO:0001651</a></td>
	<td>-</td>
</tr>
<tr>
	<td>inframe_codon_loss</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001652">SO:0001652</a></td>
	<td>-</td>
</tr>
<tr>
	<td>non_synonymous_codon</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001583">SO:0001583</a></td>
	<td>missense</td>
</tr>

<tr class="bg2">
	<td>Frameshift coding</td>
	<td>In coding sequence, resulting in a frameshift</td>
	<td>frameshift_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001589">SO:0001589</a></td>
	<td>frameshift</td>
</tr>
<tr>
	<td>Splice site</td>
	<td>1-3 bps into an exon or 3-8 bps into an intron</td>
	<td>splice_region_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001630">SO:0001630</a></td>
	<td>-</td>
</tr>
<tr class="bg2">
	<td>Partial codon</td>
	<td>Located within the final, incomplete codon of a transcript whose end coordinate is unknown</td>
	<td>incomplete_terminal_codon_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001626">SO:0001626</a></td>
	<td>-</td>
</tr>
<tr>
	<td rowspan="2">Synonymous coding</td>
	<td rowspan="2">In coding sequence, not resulting in an amino acid change (silent mutation)</td>
	<td>stop_retained_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001567">SO:0001567</a></td>
	<td>-</td>
</tr>
<tr>
	<td>synonymous_codon</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001588">SO:0001588</a></td>
	<td>cds-synon</td>
</tr>
<tr class="bg2">
	<td>Coding unknown</td>
	<td>In coding sequence with indeterminate effect</td>
	<td>coding_sequence_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001580">SO:0001580</a></td>
	<td>-</td>
</tr>
<tr>
	<td>Within mature miRNA</td>
	<td>Located within a microRNA</td>
	<td>mature_miRNA_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001620">SO:0001620</a></td>
	<td>-</td>
</tr>
<tr class="bg2">
	<td>5 prime UTR</td>
	<td>In 5 prime untranslated region</td>
	<td>5_prime_UTR_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001623">SO:0001623</a></td>
	<td>untranslated_5</td>
</tr>
<tr>
	<td>3 prime UTR</td>
	<td>In 3 prime untranslated region</td>
	<td>3_prime_UTR_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001624">SO:0001624</a></td>
	<td>untranslated_3</td>
</tr>
<tr class="bg2">
	<td>Intronic</td>
	<td>In intron</td>
	<td>intron_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001627">SO:0001627</a></td>
	<td>intron</td>
</tr>
<tr>
	<td>NMD transcript</td>
	<td>Located within a transcript predicted to undergo nonsense-mediated decay</td>
	<td>NMD_transcript_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001621">SO:0001621</a></td>
	<td>-</td>
</tr>
<tr class="bg2">
	<td>Within non-coding gene</td>
	<td>Located within a gene that does not code for a protein</td>
	<td>nc_transcript_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001619">SO:0001619</a></td>
	<td>-</td>
</tr>
<tr>
	<td rowspan="2">Upstream</td>
	<td rowspan="2">Within 5 kb upstream of the 5 prime end of a transcript</td>
	<td>2KB_upstream_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001636">SO:0001636</a></td>
	<td>near-gene-5</td>
</tr>
<tr>
	<td>5KB_upstream_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001635">SO:0001635</a></td>
	<td>-</td>
</tr>
<tr class="bg2">
	<td rowspan="2">Downstream</td>
	<td rowspan="2">Within 5 kb downstream of the 3 prime end of a transcript</td>
	<td>500B_downstream_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001634">SO:0001634</a></td>
	<td>near-gene-3</td>
</tr>
<tr class="bg2">
	<td>5KB_downstream_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001633">SO:0001633</a></td>
	<td>-</td>
</tr>
<tr>
	<td>Regulatory region</td>
	<td>In regulatory region annotated by Ensembl</td>
	<td>regulatory_region_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001566">SO:0001566</a></td>
	<td>-</td>
</tr>
<tr class="bg2">
	<td>Transcription factor binding motif</td>
	<td>Falls in a transcription factor binding motif within an Ensembl regulatory region</td>
	<td>TF_binding_site_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001782">SO:0001782</a></td>
	<td>-</td>
</tr>
<tr>
	<td>Intergenic</td>
	<td>More than 5 kb either upstream or downstream of a transcript</td>
	<td>intergenic_variant</td>
	<td><a rel="external" href="http://www.sequenceontology.org/miso/current_release/term/SO:0001628">SO:0001628</a></td>
	<td>-</td>
</tr>
</table>

<br />
<h2 id="nsSNP">Protein function predictions</h2>

<p>
For human mutations that are predicted to result in an amino acid substitution we also 
provide SIFT and PolyPhen predictions for the effect of this substitution on
protein function. We compute the predictions for each of these tools for all possible 
single amino acid substitutions in the Ensembl human proteome. This means we can provide 
predictions for novel mutations for VEP and API users. We were able to compute predictions
from at least one tool for over 95% of the proteins in Ensembl. We also use the Condel tool
to provide a consensus prediction based on the SIFT and PolyPhen prediction scores.
</p>

<p>
These tools are developed by external groups and we provide a brief explanation of
the approach each takes below, and how we run it in Ensembl. For much more detail
please see the representative papers listed <a href="#references">below</a>, and the 
relevant publications available on each tool's website. We hope to be able to provide 
protein function predictions for species other than human in future releases.
</p>

<h3 id="sift">SIFT</h3>

<p>
<a rel="external" href="http://sift.jcvi.org/">SIFT</a> predicts whether an amino acid 
substitution is likely to affect protein function based on sequence homology and the
physico-chemical similarity between the alternate amino acids. The data we provide for each amino
acid substitution is a score and a qualitative prediction (either 'tolerated' or 'deleterious'). 
The score is the normalized probability that the amino acid change is 
tolerated so scores nearer 0 are more likely to be deleterious.
The qualitative prediction is derived from this score such that substitutions with 
a score &lt; 0.05 are called 'deleterious' and all others are called 'tolerated'.
</p>

<p>
We ran SIFT version 4.0.3 (available 
<a rel="external" href="http://sift.jcvi.org/www/sift4.0.3.tar.gz">here</a>) 
following the instructions from the authors and used SIFT
to choose homologous proteins rather than supplying them ourselves. We used all 
protein sequences available from UniProtKB (both the SwissProt and TrEMBL sets) as 
the protein database. All data was downloaded from UniProt in May 2011.
</p>

<h3 id="polyphen">PolyPhen</h3>

<p>
<a rel="external" href="http://genetics.bwh.harvard.edu/pph2/">PolyPhen-2</a> predicts
the effect of an amino acid substitution on the structure and function of a protein
using sequence homology, Pfam annotations, 3D structures from PDB where available, and
a number of other databases and tools (including DSSP, ncoils etc.). As with SIFT, for each amino
acid substitution where we have been able to calculate a prediction, we provide both a
qualitative prediction (one of 'probably damaging', 'possibly damaging', 'benign'
or 'unknown') and a score. The PolyPhen score represents the probability that a substitution 
is damaging, so values nearer 1 are more confidently predicted to be deleterious (note 
that this the opposite to SIFT). The qualitative prediction is based on the False 
Positive Rate of the classifier model used to make the predictions.
</p>

<p>
We ran PolyPhen-2 version 2.1.0 (available 
<a rel="external" href="http://genetics.bwh.harvard.edu/pph2/dokuwiki/downloads">here</a>)
and again we followed all instructions from the authors, and used the UniProtKB 
UniRef100 non-redundant protein set as the protein database, which was downloaded,
along with PDB structures, and annotations from Pfam and DSSP in May 2011. When 
computing the predictions we used the classifier model trained on the HumDiv training 
set (please refer to the PolyPhen publications for more details of the classification 
system).
</p>

<h3 id="condel">Condel</h3>

<p>
<a rel="external" href="http://bg.upf.edu/group/projects/condel.php">Condel</a> is
a general method for calculating a consensus prediction from the output of tools
designed to predict the effect of an amino acid substitution. It does so by calculating
a weighted average score (WAS) of the scores of each component method. The Condel authors 
provided us with a version specialised for finding a consensus between SIFT and 
PolyPhen and we integrated this into the variation API. Tests run by the authors 
on the HumVar dataset (a test set curated by the PolyPhen team), show that Condel 
can improve both the sensitivity and specificity of predictions compared to either 
SIFT or PolyPhen used alone (please contact the authors for details). The Condel 
score, along with a qualitative prediction (one of 'neutral' or 'deleterious'), 
are available in the VEP and via the API. The Condel score is the consensus probability 
that a substitution is deleterious, so values nearer 1 are predicted with greater 
confidence to affect protein function.
</p>

<h3 id="nsSNP_data_format">Prediction data format</h3>

<p>
The SIFT and PolyPhen predictions are precomputed and stored in the variation
databases and predictions are accessible in the variation API by using the <i>sift_prediction</i>, 
<i>sift_score</i>, <i>polyphen_prediction</i> and <i>polyphen_score</i> methods on a 
<i>Bio::EnsEMBL::Variation::TranscriptVariationAllele</i> object. For users wanting
to access the complete set of predictions from the MySQL database or API, an explanation of
the format used is provided here.
</p>

<p>
The predictions and associated scores are stored as a matrix, with a column for 
each possible alternate amino acid and a row for each position in the translation. Each prediction
for a position and amino acid is stored as a 2-byte value which encodes both the qualitative
prediction and score encoded as described below. A simple example matrix is shown in the figure 
below, though note we only show the decoded score rather than the actual binary value stored
in the database.
</p>

<a rel="external" href="/img/protein_function_prediction_matrix.png">
<img src="/img/protein_function_prediction_matrix.png" style="width:500px;heigth:143px" alt="protein function encoding"/>
</a>

<p>
Prediction matrices can be fetched and manipulated in a user-friendly manner using the variation API, specifically
using the <i>ProteinFunctionPredictionMatrixAdaptor</i> which allows you to fetch a prediction matrix
using either a transcript or a translation stable ID. This adaptor returns a <i>ProteinFunctionPredictionMatrix</i>
object and you can use the <i>get_prediction</i> method to retrieve a prediction for a given position
and amino acid. Users who want to use entire matrices should use the <i>deserialize</i> method to decode
an entire binary formatted matrix into a simple Perl hash. Please refer to the <a
href="/info/docs/api/variation/index.html">API documentation</a> for both 
of these classes for more details. For users who require direct access to the MySQL database (for instance 
because they are accessing the database in a language other than Perl) we provide a description of the 
binary format used below.
</p>

<p>
Prediction matrices for each translation from each of SIFT and PolyPhen are stored in the 
<i>protein_function_predictions</i> table, in the <i>polyphen_predictions</i> or 
<i>sift_predictions</i> columns. The other columns in this table identify the stable ID of 
the Ensembl translation for each
protein coding transcript, the stable ID of the transcript itself, and a hexadecimal 
MD5 hash of the amino acid sequence of the translation. This can be used to ensure 
the predictions apply to the correct translation sequence. The prediction matrices are stored in the 
<i>sift_predictions</i> and <i>polyphen_predictions</i> columns as a gzip compressed binary string.
Once uncompressed, this string contains a 40 byte substring for each row in the matrix
concatenated together in position order. Each row is composed of 20 2-byte predictions, 
one for each possible alternative amino acid in alphabetical order, though note that the 
value for the amino acid that matches the reference amino acid is 
identified as a 2-byte value with all bits set, or <i>0xFFFF</i> in hexadecimal notation. 
The prediction itself is packed as a 16 bit little-endian 
("VAX" order, or "v" format if using the perl <i>pack</i> subroutine) unsigned short value. 
The top 2 bits of this short 
are used to encode the qualitative prediction (PolyPhen has 4 possible values, while SIFT has just
2), and the bottom 10 bits are used to encode the prediction score. To decode the qualitative 
prediction you should mask off all bits except the top 2, and shift the resulting short
right by 14 bits and treat this as an integer between 0 and 3. The corresponding prediction can then 
be looked up in the table below. To decode the prediction score you should mask off the 
top 6 bits and the resulting value can be treated as a number between 0 and 1000, which 
should be divided by 1000 to give a 3 decimal place score (casting to a floating point type
if necessary). Bits 11-14 are not used, except to encode the "same as reference" dummy prediction 
<i>0xFFFF</i>.
</p>

<img src="/img/protein_function_encoding.png" style="width:500px;heigth:143px" alt="protein function encoding"/>

<p>
A diagram of the encoding scheme is shown above. In this example from a polyphen prediction, 
the top prediction bits are <i>0b01</i> which in decimal is the number 1, which corresponds to 
"possibly damaging" in the table below. The score bits are <i>0b1110001010</i> which in decimal 
is the number 906, which when divided by 1000, gives a score of 0.906.
</p>

<table id="prediction_values" class="ss" style="width:50%;">
<tr>
	<th>Tool</th>
	<th>Numerical value</th>
	<th>Qualitative prediction</th>
</tr>
<tr>
	<td>PolyPhen</td>
	<td>0</td>
	<td>"probably damaging"</td>
</tr>
<tr class="bg2">
	<td>PolyPhen</td>
	<td>1</td>
	<td>"possibly damaging"</td>
</tr>
<tr>
	<td>PolyPhen</td>
	<td>2</td>
	<td>"benign"</td>
</tr>
<tr class="bg2">
	<td>PolyPhen</td>
	<td>3</td>
	<td>"unknown"</td>
</tr>
<tr>
  <td>SIFT</td>
	<td>0</td>
	<td>"tolerated"</td>
</tr>
<tr class="bg2">
	<td>SIFT</td>
	<td>1</td>
	<td>"deleterious"</td>
</tr>
</table>

<p>
To retrieve a prediction for a particular amino acid substitution at a given position
in a translation, first you must retrieve the binary matrix from the database and
uncompress it using gzip. You can then calculate the offset into this string by multiplying 
the desired position (starting at 0) by the index of the desired amino acid in an alphabetical ordering of
amino acids (also starting at 0), and then multiply this value by 2 to take into account the fact that each 
prediction uses 2 bytes. Each matrix also includes a 3 byte header used check if the data is corrupted etc. so
you will also need to add 3 to the calculated offset. The 2 bytes at the calculated position 
can then be extracted and the approach described above can be used to retrieve the prediction and score. 
A perl implementation of this scheme can be found in the <i>Bio::EnsEMBL::Variation::ProteinFunctionPredictionMatrix</i>
module in the variation API and should hopefully inform attempts to reimplement this 
scheme in other languages.
</p>

<p>
Condel predictions are very fast to compute and so are not precomputed and stored in the
database, instead we use the <i>get_condel_prediction</i> subroutine provided in the
<i>Bio::EnsEMBL::Variation::Utils::Condel</i> module to calculate the Condel prediction from
a precomputed SIFT and PolyPhen score on the fly.
</p>

<br />

<h2 id="VEP">The Variant Effect Predictor (VEP)</h2>
    
<p>
Users can run all the analyses described above on novel mutations 
using the Ensembl Variant Effect Predictor (VEP). Provided you have a 
list of allele sequences and associated genomic coordinates, you can use 
the VEP to calculate consequences with respect to the Ensembl transcript set for
your species of interest, and (for human) to retrieve SIFT, PolyPhen and Condel predictions 
for any missense variations. The VEP can also identify if a variation is co-located with a 
known variation (i.e. already found in dbSNP or any of the other sources of variation 
data listed <a href="#sources">above</a>), and so can be used to filter your 
data for novel variant loci. The VEP works for any species in Ensembl and not 
just those for which we have an existing variation database.
</p>

<p>
The VEP can either be run online from the <a href="/tools.html">Tools</a> 
section of the website or you can download a script to run yourself on larger
datasets. You can also write custom scripts using the Perl API.
All of these approaches use the same backend code as is used to calculate the
consequences displayed on the website. Please refer to the 
<a href="/info/website/upload/var.html">VEP</a> documentation for more
details on running the VEP, including supported input formats, or the 
<a href="/info/docs/api/variation/index.html">variation API</a> documentation
for help writing custom scripts. The VEP is under active development and
<a href="/info/about/contact/index.html">feedback</a> is welcome!
</p>

<p> See the <a href="/info/docs/variation/vep/index.html">full documentation</a>
for more details. </p>

<br />

<h2 id="variation_sets">Variation sets</h2>
    
<p>
We use the concept of variation sets to group variations that share some property together. For example, we have grouped the variations identified in the three different
1000 Genomes pilot studies into separate variation sets. The sets can be further subdivided into supersets and subsets to reflect hierarchical relationships between them. In the case
of the 1000 Genomes pilot sets, these are divided into subsets based on population. For example, the set representing variations identified in the first 1000 Genomes pilot study is named 
'1000 Genomes - Low coverage' and has three subsets: '1000 Genomes - Low coverage - CEU', '1000 Genomes - Low coverage - CHB+JPT' and '1000 Genomes - Low coverage - YRI'. 
The variation sets can be displayed as separate tracks on the location view. This behaviour is controlled from the 'Germline variations' section on the configuration panel which is accessed by 
clicking the 'Configure this page' link in the left hand side navigation.
</p>

<p>
The sets are constructed during production and are stored in the database. The table below lists the available
variation sets in the Ensembl variation database (subsets are indicated by bullet points).
</p>

<table id="variation_set_table" class="ss">
	<tr>
		<th>Name</th>
		<th>Short name</th>
		<th>Description</th>
	<tr>
	<tr>
		<td>1000 Genomes - High coverage - Trios</td>
		<td>1kg_hct</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data from two family trios (Pilot 2)</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage - Trios - CEU</li></ul></td>
		<td>1kg_hct_ceu</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data from a CEU family trio (two parents and one daughter) (Pilot 2)</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage - Trios - YRI</li></ul></td>
		<td>1kg_hct_yri</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data from a YRI family trio (two parents and one daughter) (Pilot 2)</td>
	<tr>
	<tr class="bg2">
		<td>1000 Genomes - High coverage exons</td>
		<td>1kg_hce</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of 8,140 exons from 906 randomly selected genes in 697 individuals (Pilot 3)</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage exons - CEU</li></ul></td>
		<td>1kg_hce_ceu</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of exons from 90 CEU individuals (Pilot 3)</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage exons - CHB</li></ul></td>
		<td>1kg_hce_chb</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of exons from 109 CHB individuals (Pilot 3)</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage exons - CHD</li></ul></td>
		<td>1kg_hce_chd</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of exons from 107 CHD individuals (Pilot 3)</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage exons - JPT</li></ul></td>
		<td>1kg_hce_jpt</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of exons from 105 JPT individuals (Pilot 3)</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage exons - LWK</li></ul></td>
		<td>1kg_hce_lwk</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of exons from 108 LWK individuals (Pilot 3)</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage exons - TSI</li></ul></td>
		<td>1kg_hce_tsi</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of exons from 66 TSI individuals (Pilot 3)</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - High coverage exons - YRI</li></ul></td>
		<td>1kg_hce_yri</td>
		<td>Variations called by the 1000 Genomes project on high coverage sequence data of exons from 112 YRI individuals (Pilot 3)</td>
	<tr>
	<tr class="bg2">
		<td>1000 Genomes - Low coverage</td>
		<td>1kg_lc</td>
		<td>Variations called by the 1000 Genomes project on low coverage sequence data from 179 unrelated individuals (Pilot 1)</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - Low coverage - CEU</li></ul></td>
		<td>1kg_lc_ceu</td>
		<td>Variations called by the 1000 Genomes project on low coverage sequence data from 60 unrelated CEU individuals (Pilot 1)</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - Low coverage - CHB+JPT</li></ul></td>
		<td>1kg_lc_chb_jpt</td>
		<td>Variations called by the 1000 Genomes project on low coverage sequence data from 60 unrelated CHB individuals and 60 unrelated JPT individuals (Pilot 1)</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">1000 Genomes - Low coverage - YRI</li></ul></td>
		<td>1kg_lc_yri</td>
		<td>Variations called by the 1000 Genomes project on low coverage sequence data from 59 unrelated YRI individuals (Pilot 1)</td>
	<tr>
	<tr class="bg2">
		<td>All failed variations</td>
		<td>fail_all</td>
		<td>Variations that have failed the Ensembl QC checks</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Failed: Alleles do not match reference</li></ul></td>
		<td>fail_nonref</td>
		<td>None of the variant alleles match the reference allele</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Failed: Ambiguous alleles</li></ul></td>
		<td>fail_ambig</td>
		<td>Alleles contain ambiguity codes</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Failed: Genotype frequencies</li></ul></td>
		<td>fail_gt_fq</td>
		<td>Genotype frequencies do not add up to 1</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Failed: Inconsistent mapping</li></ul></td>
		<td>fail_incons_map</td>
		<td>Mapped position is not compatible with reported alleles</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Failed: Multiple mappings</li></ul></td>
		<td>fail_mult_map</td>
		<td>Variation maps to more than 3 different locations</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Failed: No alleles</li></ul></td>
		<td>fail_no_alleles</td>
		<td>Loci with no observed variant alleles in dbSNP</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Failed: No genotypes</li></ul></td>
		<td>fail_no_gt</td>
		<td>Variation has no genotypes</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Failed: No mapping</li></ul></td>
		<td>fail_no_map</td>
		<td>Variation does not map to the genome</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Failed: No sequence</li></ul></td>
		<td>fail_no_seq</td>
		<td>Variation has no associated sequence</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Failed: Non-nucleotide alleles</li></ul></td>
		<td>fail_non_nt</td>
		<td>Alleles contain non-nucleotide characters</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Failed: Too many alleles</li></ul></td>
		<td>fail_mult_alleles</td>
		<td>Variation has more than 3 different alleles</td>
	<tr>
	<tr class="bg2">
		<td>All phenotype-associated variants</td>
		<td>ph_variants</td>
		<td>Variants that have been associated with a phenotype</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">COSMIC phenotype variants</li></ul></td>
		<td>ph_cosmic</td>
		<td>Phenotype annotations of somatic mutations found in human cancers from the COSMIC project</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">HGMD-PUBLIC phenotype variants</li></ul></td>
		<td>ph_hgmd_pub</td>
		<td>Variants with phenotypes annotated by HGMD</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">Johnson & O'Donnell phenotype variants</li></ul></td>
		<td>ph_johnson_et_al</td>
		<td>Johnson & O'Donnell 'An Open Access Database of Genome-wide Association Results' PMID:19161620</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">NHGRI catalog phenotype variants</li></ul></td>
		<td>ph_nhgri</td>
		<td>Variants associated with phenotype data from the NHGRI GWAS catalog [http://www.genome.gov/gwastudies/]</td>
	<tr>
	<tr>
		<td><ul style="margin:0px"><li style="margin:0px">OMIM phenotype variants</li></ul></td>
		<td>ph_omim</td>
		<td>Variations linked to entries in the Online Mendelian Inheritance in Man (OMIM) database</td>
	<tr>
	<tr class="bg2">
		<td><ul style="margin:0px"><li style="margin:0px">Uniprot phenotype variants</li></ul></td>
		<td>ph_uniprot</td>
		<td>Variations with phenotype annotations provided by Uniprot</td>
	<tr>
	<tr>
		<td>Anonymous Irish Male</td>
		<td>ind_irish</td>
		<td>Variants genotyped in an anonymous Irish Male</td>
	<tr>
	<tr class="bg2">
		<td>Anonymous Korean</td>
		<td>ind_ak1</td>
		<td>Variants genotyped in an anonymous Korean individual</td>
	<tr>
	<tr>
		<td>Clinical/LSDB variations from dbSNP</td>
		<td>precious</td>
		<td>Variations that belong to a reserved or "precious" set of clinically associated SNPs from dbSNP [http://www.ncbi.nlm.nih.gov/projects/SNP/]</td>
	<tr>
	<tr class="bg2">
		<td>ENSEMBL:Venter</td>
		<td>ind_venter</td>
		<td>Variants genotyped in Craig Venter</td>
	<tr>
	<tr>
		<td>ENSEMBL:Watson</td>
		<td>ind_watson</td>
		<td>Variants genotyped in James Watson</td>
	<tr>
	<tr class="bg2">
		<td>Failed: Flagged by dbSNP</td>
		<td>fail_dbsnp_suspect</td>
		<td>Flagged as suspect by dbSNP</td>
	<tr>
	<tr>
		<td>HapMap</td>
		<td>hapmap</td>
		<td>Variations which have been assayed by The International HapMap Project [http://hapmap.ncbi.nlm.nih.gov/]</td>
	<tr>
	<tr class="bg2">
		<td>Henry Louis Gates Jr</td>
		<td>ind_gates_jr</td>
		<td>Variants genotyped in Henry Louis Gates Jr</td>
	<tr>
	<tr>
		<td>Henry Louis Gates Sr</td>
		<td>ind_gates_sr</td>
		<td>Variants genotyped in Henry Louis Gates Sr</td>
	<tr>
	<tr class="bg2">
		<td>Marjolein Kriek</td>
		<td>ind_kriek</td>
		<td>Variants genotyped in Marjolein Kriek</td>
	<tr>
	<tr>
		<td>Misha Angrist</td>
		<td>ind_angrist</td>
		<td>Variants genotyped in Misha Angrist</td>
	<tr>
	<tr class="bg2">
		<td>Rosalynn Gill</td>
		<td>ind_gill</td>
		<td>Variants genotyped in Rosalynn Gill</td>
	<tr>
	<tr>
		<td>Saqqaq</td>
		<td>ind_saqqaq</td>
		<td>Variants genotyped in a Palaeo-Eskimo Saqqaq individual</td>
	<tr>
	<tr class="bg2">
		<td>Saqqaq HC</td>
		<td>ind_saqqaq_hc</td>
		<td>Variants genotyped in a Palaeo-Eskimo Saqqaq individual (high confidence SNPs)</td>
	<tr>
	<tr>
		<td>Seong-Jin Kim</td>
		<td>ind_sjk</td>
		<td>Variants genotyped in Seong-Jin Kim</td>
	<tr>
	<tr class="bg2">
		<td>Stephen Quake</td>
		<td>ind_quake</td>
		<td>Variants genotyped in Stephen Quake</td>
	<tr>
	<tr>
		<td>YanHang</td>
		<td>ind_yh</td>
		<td>Variants genotyped in a Han Chinese individual (YanHuang Project)</td>
	<tr>
</table>

  
<br />

<h2 id="notes">Notes</h2>

<p>For more detailed information on the variation specific help see <a href="/Help/View?id=168">here</a>.<br
/>
There is also more information about programmatic access in the <a
href="/info/docs/api/variation/variation_tutorial.html">variation API tutorial</a> page.
</p>


<br />
<h2 id="references">References</h2>

<p>McLaren W, Pritchard B, Rios D, Chen Y, Flicek P, Cunningham F.<br />
<strong>Deriving the consequences of genomic variants with the Ensembl API and SNP Effect Predictor.</strong><br />
<em>BMC Bioinformatics</em>26(16):2069-70(2010)<br />
<a href="http://dx.doi.org/10.1093/bioinformatics/btq330"
rel="external">doi:10.1093/bioinformatics/btq330 </a>
</p> 

<p>
Rios D, McLaren WM, Chen Y, Birney E, Stabenau A, Flicek P, Cunningham F.<br />
<strong>A Database and API for variation, dense genotyping and resequencing data</strong><br />
<em>BMC Bioinformatics</em> 11:238 (2010)<br />
<a href="http://dx.doi.org/10.1186/1471-2105-11-238"
rel="external">doi:10.1186/1471-2105-11-238</a>
</p>

<p>Chen Y, Cunningham F, Rios D, McLaren WM, Smith J, Pritchard B, Spudich GM, Brent S, Kulesha E, Marin-Garcia P, Smedley D, Birney E, Flicek P.<br />
<strong>Ensembl Variation Resources</strong><br />
<em>BMC Genomics</em> 11(1):293 (2010)<br />
<a href="http://dx.doi.org/10.1186/1471-2164-11-293"
rel="external">doi:10.1186/1471-2164-11-293</a>
</p>

<p>Adzhubei IA, Schmidt S, Peshkin L, Ramensky VE, Gerasimova A, Bork P, Kondrashov AS, Sunyaev SR. <br />
<strong>A method and server for predicting damaging missense mutations</strong><br />
<em>Nature Methods</em> 7(4):248-249 (2010)<br />
<a href="http://dx.doi.org/10.1038/nmeth0410-248" rel="external">doi:10.1038/nmeth0410-248</a>
</p> 

<p>Kumar P, Henikoff S, Ng PC.<br />
<strong>Predicting the effects of coding non-synonymous variants on protein function using the SIFT algorithm</strong><br />
<em>Nature Protocols</em> 4(8):1073-1081 (2009)<br />
<a href="http://dx.doi.org/10.1038/nprot.2009.86" rel="external">doi:10.1038/nprot.2009.86</a>
</p> 

<p>Gonzalez-Perez A, Lopez-Bigas N.<br />
<strong>Improving the assessment of the outcome of non-synonymous SNVs with a Consensus deleteriousness score (Condel)</strong><br />
<em>Am J Hum Genet</em> 88(4):440-449 (2011)<br />
<a href="http://dx.doi.org/10.1016/j.ajhg.2011.03.004" rel="external">doi:10.1016/j.ajhg.2011.03.004</a>
</p> 

<p>
Redon, R. <i>et al.</i><br />
<strong>Global variation in copy number in the human
genome</strong><br />
<em>Nature</em> 444:444-454 (2006)<br />
<a href="http://dx.doi.org/10.1038/nature05329"
rel="external">doi:10.1038/nature05329</a>
</p>

<p>
Spencer, C. C. A. <i>et al.</i><br />
<strong>The Influence of Recombination on Human Genetic
Diversity</strong><br />
<em>PLoS Genet.</em> 2(9):e148 (2006)<br />
<a href="http://dx.doi.org/10.1371/journal.pgen.0020148"
rel="external">doi:10.1371/journal.pgen.0020148</a>
</p>

<p>
Venter, J. C. <i>et al.</i><br />
<strong>The Sequence of the Human Genome</strong><br />
<em>Science</em> 291(5507):1304-51 (2001)<br />
<a href="http://dx.doi.org/10.1126/science.1058040"
rel="external">doi:10.1126/science.1058040</a>
</p>

<p>
Wheeler, D. A. <i>et al.</i><br />
<strong>The complete genome of an individual by massively parallel DNA
sequencing</strong><br />
<em>Nature</em> 452:872-876 (2008)<br />
<a href="http://dx.doi.org/10.1038/nature06884"
rel="external">doi:10.1038/nature06884</a>
</p>

</body>
</html>
