<html>
<head>
<meta name="order" content="1" />
<title>Variation</title>
</head>

<body>
<b>
About Ensembl Variation | 
<a href="data_description.html">Data Description</a> | 
<a href="predicted_data.html">Predicted Data</a> | 
<a href="database.html">Database Description</a> | 
<a href="/info/docs/api/variation/index.html">Perl API</a> | 
<a href="vep/index.html">Variant Effect Predictor</a>
</b>
<br /><br />

<h1>About Ensembl Variation</h1>

<p>
The Ensembl database has structural variants and sequence variants
(including Single Nucleotide Polymorphisms (SNPs), insertions,
deletions and somatic mutations) for several <a href="data_description.html#source">species</a>.
<br />
We predict the effects of variants on the Ensembl transcripts and
regulatory features for each species. You can run the same analysis on
your own data using the <a href="vep/index.html">Variant Effect
Predictor</a>.
<br />
This data is integrated with the other data sources in Ensembl and
available on the website and via the Perl API for programmatic access
to the variation data.
</p>


<br />
<p>
The Ensembl Variation documentation is divided in 4 parts:
</p>
<div style="border:1px #555 solid;width:700px">
<ul>
	<li>
		<b>Data</b> : 
		<a href="data_description.html">Description</a> | 
		<a href="sources_documentation.html">Sources</a> | 
		<a href="predicted_data.html">Predictions</a>
	</li>
	<li>
		<b>Database</b> : 
		<a href="database.html">Description</a> | 
		<a href="variation-database-schema.pdf">Schema diagram [PDF]</a> | 
		<a href="variation_schema.html">Tables description</a>
	</li>
	<li>
		<b>Perl API</b> : 
		<a href="/info/docs/Doxygen/variation-api/index.html">Documentation</a> | 
		<a href="/info/docs/api/variation/variation_tutorial.html">Tutorial</a>
	</li>
	<li>
		<b>Variant Effect Predictor (VEP)</b> : 
		<a href="vep/index.html#web">Web documentation</a> | 
		<a href="vep/vep_script.html">Perl script documentation</a>
	</li>
</ul>
</div>


<br />
<h2 id="references">References</h2>

<ul>
	<li>
		<p>Flicek P, Amode MR, Barrell D, Beal K, Brent S, Carvalho-Silva D, Clapham P, Coates G, Fairley S, Fitzgerald S, Gil L, Gordon L, Hendrix M, Hourlier T,<br />
		Johnson N, K&auml;h&auml;ri AK, Keefe D, Keenan S, Kinsella R, Komorowska M, Koscielny G, Kulesha E, Larsson P, Longden I, McLaren W, Muffato M, Overduin B, Pignatelli M,<br />
		Pritchard B, Riat HS, Ritchie GR, Ruffier M, Schuster M, Sobral D, Tang YA, Taylor K, Trevanion S, Vandrovcova J, White S, Wilson M, Wilder SP, Aken BL, Birney E,<br />
		Cunningham F, Dunham I, Durbin R, Fern&aring;ndez-Suarez XM, Harrow J, Herrero J, Hubbard TJ, Parker A, Proctor G, Spudich G, Vogel J, Yates A, Zadissa A, Searle SM.<br />
		<strong>Ensembl 2012</strong><br />
		<em>Nucleic Acids Research</em> 40(D1):D84-D90(2012)<br />
		<a href="http://dx.doi.org/10.1093/nar/gkr991"
		rel="external">doi:10.1093/nar/gkr991</a></p>
	</li>

	<li>
		<p>McLaren W, Pritchard B, Rios D, Chen Y, Flicek P, Cunningham F.<br />
		<strong>Deriving the consequences of genomic variants with the Ensembl API and SNP Effect Predictor.</strong><br />
		<em>BMC Bioinformatics</em> 26(16):2069-70(2010)<br />
		<a href="http://dx.doi.org/10.1093/bioinformatics/btq330"
		rel="external">doi:10.1093/bioinformatics/btq330 </a></p> 
	</li>

	<li>
		<p>
		Rios D, McLaren WM, Chen Y, Birney E, Stabenau A, Flicek P, Cunningham F.<br />
		<strong>A Database and API for variation, dense genotyping and resequencing data</strong><br />
		<em>BMC Bioinformatics</em> 11:238 (2010)<br />
		<a href="http://dx.doi.org/10.1186/1471-2105-11-238"
		rel="external">doi:10.1186/1471-2105-11-238</a></p>
	</li>

	<li>
		<p>Chen Y, Cunningham F, Rios D, McLaren WM, Smith J, Pritchard B, Spudich GM, Brent S, Kulesha E, Marin-Garcia P, Smedley D, Birney E, Flicek P.<br />
		<strong>Ensembl Variation Resources</strong><br />
		<em>BMC Genomics</em> 11(1):293 (2010)<br />
		<a href="http://dx.doi.org/10.1186/1471-2164-11-293"
		rel="external">doi:10.1186/1471-2164-11-293</a></p>
	</li>
</ul>

</body>
</html>
