<html>
<head>
<title>PETs from the Genome Institute of Singapore</title>
</head>
<body>
     
<h2>Genome Institute of Singapore: GIS / ChIP paired-end ditags</h2>
        

<h3>Description</h3>

<p>This track shows the start and end sequence tags of mRNA transcripts
determined by paired-end ditag (PET) sequencing. PETs are composed of
14-21 bases from either end of a cDNA. See the <a href="#refs">References</a>
section below for more details on PET sequencing. The track shows GIS (gene identification signature) 
and ChIP (Chromatin Immunoprecipitation) PETs that were pre-processed and fetched with
the <a href="http://www.gis.a-star.edu.sg/PET_Tool/">PET-Tool</a> (Genome Institute of
Singapore).</p>

<h4>GIS libraries</h4>

<ul>

  <li>SHC12 from human cancer cells MCF7</li>

  <li>SHC13 from human cancer cells HCT116</li>

</ul>

<h4>ChIP libraries</h4>

<ul>

  <li>SHC16 from chromatin immunoprecipitated DNA enriched for
STAT1 binding sites in human HeLa cells
with interferon-&gamma; treatment.</li>

  <li>SHC19 from chromatin immunoprecipitated DNA enriched for
STAT1 binding sites in human HeLa cells without
interferon-&gamma; treatment<br>

  </li>

</ul>

<p>The 5' and 3' ends of the ditag features are represented as small
boxes connected by a horizontal arrow indicating the strand/direction.
Clicking on the feature brings up a menu with information about this
ditag and a link to this more detailed description.</p>

<h3>Alignments</h3>

<p>The ditags were split in half and aligned to the unmasked genome
sequence using <a href="http://www.ebi.ac.uk/~guy/exonerate/">Exonerate</a>.
The hits were filtered using the following strategy:</p>

<ul>

  <li>score &gt;= 75/85, considering only highest scoring hits</li>

  <li>hits for both ends</li>

  <li>both ends on same chromosome, same strand and within 600 kb</li>

  <li>for GIS PETs: 5'/3' orientation has to remain correct</li>

  <li>if the tags align to more than 4 genomic locations, they
were discarded as being unreliable</li>

  <li>sequences with non-(ATGC) characters were removed</li>

  <li>sequences with poly(ATGC) &gt; 5 were removed</li>

</ul>

<p>Identical sequences and hits to the same location (start &amp; end
of ditag) were collapsed into a non-redundant set.</p>

<h3>Credits</h3>

<p>The GIS-PET &amp; ChIP-PET libraries and sequence data for transcriptome analysis were
produced at the <a href="http://www.bii.a-star.edu.sg/index.asp">Genome Institute of Singapore</a> and
processed using their PET-Tool.</p>

<p>The data was mapped and analyzed using code of the <a href="/">EnsEMBL</a> project.</p>

<h3 id="refs">References</h3>

<p>Chiu KP, Wong CH, Chen Q, Ariyaratne P, Ooi HS, Wei CL, Sung WK, Ruan Y.<br />
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=16934139&amp;query_hl=1">PET-Tool: a software suite for comprehensive processing and managing of
Paired-End diTag (PET) sequence data</a>.<br />
<em>BMC Bioinformatics</em>, 2006 Aug 25;7:390.</p>

<p>Ng P, Wei CL, Sung WK, Chiu KP, Lipovich L, Ang CC, Gupta S,
Shahab A, Ridwan A, Wong CH, Liu ET, Ruan Y.<br />
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15782207&amp;query_hl=1">Gene identification signature (GIS) analysis for transcriptome
characterization and genome annotation</a>.<br />
<em>Nat. Methods 2(2)</em>, 105-11 (2005).</p>

<p>Slater GS, Birney E<br />
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=15713233&amp;dopt=Abstract&amp;holding=f1000,f1000m,igbwellib">Automated generation of heuristics for biological sequence comparison</a>.<br />
<em>BMC Bioinformatics 6(1)</em>, 31 (2005).</p>

        
</body>
</html>

