<html>
<head>
  <meta name="navigation" content="Comparative Genomics">
  <title>Pairwise Alignments</title>
  <link rel="stylesheet" type="text/css" media="all" href="85_spreadsheet_table.css" />
  <link rel="stylesheet" type="text/css" media="all" href="95_static_content.css" />
</head>

<body>

<h2>C.intestinalis vs C.savignyi LastZ alignments</h2>

<p>C.intestinalis (<i>Ciona intestinalis</i>, KH) and C.savignyi (<i>Ciona savignyi</i>, CSAV2.0) were aligned using the LastZ alignment algorithm (<a href="http://www.bx.psu.edu/miller_lab/dist/README.lastz-1.02.00/README.lastz-1.02.00a.html">LastZ</a>) in Ensembl release 66. C.intestinalis  was used as the reference species. After running LastZ , the raw LastZ  alignment blocks are chained according to their location in both genomes. During the final netting process, the best sub-chain is chosen in each region on the reference species.</p>

<a href="/info/docs/compara/analyses.html" class="cp-external">Full list of 
pairwise alignments</a>


<h3 class="boxed">Configuration parameters</h3>

<table class="ss tint species-stats">
<tr class="bg2">
<td class="data" style="width:75%">Gap open penalty (O)</td>
<td class="value">400</td>
</tr>
<tr>
<td class="data">Gap extend penalty (E)</td>
<td class="value">30</td>
</tr>
<tr class="bg2">
<td class="data">HSP threshold (K)</td>
<td class="value">3000</td>
</tr>
<tr>
<td class="data">Threshold for gapped extension (L)</td>
<td class="value">3000</td>
</tr>
<tr class="bg2">
<td class="data">Threshold for alignments between gapped alignment blocks (H)</td>
<td class="value">2200</td>
</tr>
<tr>
<td class="data">Masking count (M)</td>
<td class="value">50</td>
</tr>
<tr class="bg2">
<td class="data">Seed and Transition value (T)</td>
<td class="value">1</td>
</tr>
<tr>
<td class="data">Scoring matrix (Q)</td>
<td class="value">Default</td>
</tr>
</table>

<h3>Results</h3>

<p>Number of alignment blocks: 122740</p>

<table style="text-align: left; width: 100%;" border="0" cellpadding="2"
cellspacing="2">
<tbody>
<tr>
<td style="vertical-align: top;">
</td>
<th style="vertical-align: top;">Genome coverage(bp)
</th>
<th style="vertical-align: top;">Coding exon coverage (bp)
</th>
</tr>
<tr>
<th style="vertical-align: middle;">C.intestinalis</th>
<td style="vertical-align: top;"><img style="width: 200px; height: 200px;" alt=""
src="./pie_ref_align_569.png"/>
</td>
<td style="vertical-align: top;"><img style="width: 200px; height: 200px;" alt=""
src="./pie_ref_cod_align_569.png"/>
</td>
</tr>
<tr>
<td></td>
<td style="vertical-align: top;">36,485,577 out of 115,212,710
</td>
<td style="vertical-align: top;">13,625,889 out of 16,105,584
</td>
</tr>
<tr>
<th style="vertical-align: middle;">C.savignyi
</th>
<td style="vertical-align: top;"><img style="width: 200px; height: 200px;" alt=""
src="./pie_non_ref_align_569.png"/>
</td>
<td style="vertical-align: top;"><img style="width: 200px; height: 200px;" alt=""
src="./pie_non_ref_cod_align_569.png"/>
</td>
</tr>
<tr>
<td></td>
<td style="vertical-align: top;">34,534,266 out of 176,989,013
</td>
<td style="vertical-align: top;">11,092,442 out of 13,806,906
</td>
</tr>
</tbody>
</table>


</body>
</html>
