<html>
<head>
<meta name="order" content="99" />
<title>Projects using Ensembl</title>
</head>

<body>

<h1>Projects using Ensembl</h1>


<p>
The Ensembl project is both a source of genome sequence related data
(<a href="http://www.ensembl.org/">http://www.ensembl.org</a>) and an open source 
software system that can be used to organise any such data.
</p>

<h2>User-contributed code</h2>

<p>Whilst we have developed a comprehensive Perl API in-house, we welcome contributions
in other programming languages from the community.</p>

<h3>Ruby</h3>

<p>A Ruby API has been developed by Jan Aerts. A small example, showing what it can do 
(including coordinate transformation and reflecting on the types of associations for a 
given class) can be found on the wiki page.</p>

<ul>
<li><a href="https://github.com/jandot/ruby-ensembl-api">repository</a></li>
<li><a href="https://github.com/jandot/ruby-ensembl-api/wiki">wiki</a></li>
</ul>

<h3>Java</h3>

<p>
An open source Java API (JEnsembl) has been developed by the
Bioinformatics Group at
<a href='http://www.roslin.ed.ac.uk/'>The Roslin Institute</a>. The code has
been designed to be version-aware (a single installation works against
current and previous releases) and is hosted, together with
<a href='http://jensembl.sourceforge.net/apidocs/'>documentation</a>,
<a href='http://jensembl.sourceforge.net/example.html'>examples</a>
and
<a href ='http://jensembl.sourceforge.net/contact.html'>contact
information for the Roslin
group</a>, at <a href='http://jensembl.sourceforge.net'>SourceForge</a>.
</p>

<p>Note that we do not support these user-contributed packages - please contact
the original developer if you have any questions or comments.</p>

<h2>Ensembl-powered Websites</h2>

<p>
Some projects make use of the Ensembl software system, or are
derived from it.
</p>

<p class="center">
<img src="/img/empowered.png" alt="'Powered by Ensembl' logo" width="98" height="20" />
</p>

<h3>Projects based at WTSI or EBI</h3>

<table class="ss tint">
<tr>
    <th>Project</th>
    <th>Description</th>
    <th>Code used</th>
</tr>

<tr>
    <td><strong><a href="http://www.ensemblgenomes.org/">Ensembl Genomes</a></strong></td>
    <td>Websites presenting annotation for bacteria, protists, fungi, plants and invertebrate metazoa</td>
    <td>Web front-end derived from Ensembl webcode, Ensembl schema databases, Ensembl pipelines and Perl API</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://vega.sanger.ac.uk/">Vega</a></strong></td>
    <td>Website to present curated vertebrate annotation</td>
    <td>Web front-end derived from Ensembl webcode, Ensembl schema databases,
	  Ensembl Compara pipeline and Perl API</td>
</tr>

<tr>
    <td><strong><a href="http://www.1000genomes.org/">1000 Genomes</a></strong></td>
    <td>A web browser and ftp sites are provided to access human genetic variation catalogued by the project.</td>
    <td>Web front-end derived from Ensembl webcode </td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://pgpviewer.ensembl.org">Genome Reference Consortium PGPviewer</a></strong></td>
    <td>The GRC is creating reference assemblies of human, mouse and zebrafish that are accessible in the PGP viewer</td>
    <td>Customised webcode, pipelines (not genebuild) and compara</td>
</tr>
</table>

<h3>External projects</h3>
<!-- ADD NEW ENTRIES AT THE TOP, TO KEEP TABLE STRIPING SIMPLE -->
<table class="ss tint">
<tr>
    <th>Project</th>
    <th>Description</th>
    <th>Code used</th>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.berndtlab.pitt.edu/tools/VEP/">mouse Genome Variation Identifier</a></strong></td>
    <td>A simple web interface to the Ensembl VEP for mouse SNPs.</td>
    <td>Variant Effect Predictor script and mouse variation database</td>
</tr>


<tr>
    <td><strong><a href="http://www.autoprime.de">AutoPrime</a></strong></td>
    <td>Automated design of RT primers</td>
    <td>Core API code and gene models</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.pombase.org">PomBase</a></strong></td>
    <td>A resource for <em>Schizosaccharomyces pombe</em> that includes structural and functional annotation, literature curation and access to large-scale data sets.</td>
    <td>Customised webcode, pipeline, database schema and API
</tr>

<tr>
    <td><strong><a href="http://agd.vital-it.ch/index.html">Ashyba Genome Database</a></strong></td>
    <td>A genome/transcriptome database containing gene annotation and expression data for <em>Ashbya gossypii</em> and <em>Saccharomyces cerevisiae</em></td>
    <td>Webcode</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.germonline.org/index.html">GermOnline</a></strong></td>
    <td>Microarray expression database focused on germline development</td>
    <td>Webcode and underlying data</td>
</tr>

<tr>
    <td><strong><a href="http://www.regulatorygenomics.org/index.html">GRiD (Gene Regulators in Disease)</a></strong></td>
    <td>Provides a web browser to investigate gene regulation and human disease.</td>
    <td>Webcode</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.eaglegenomics.com/downloads/sw">Eagle Browser</a></strong></td>
    <td>Provides installation of the Ensembl browser</td>
    <td>Webcode, API</td>
</tr>

<tr>
    <td><strong><a href="http://bgee.unil.ch">Bgee</a></strong></td>
    <td>Comparison of gene expression patterns between species</td>
    <td>Ensembl data and orthologues</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://search.cpan.org/~ggallone/Bio-Homology-InterologWalk/">InterologWalk Perl Modules</a></strong></td>
    <td>Perl modules to determine protein-protein interactions from orthology and interaction data.</td>
    <td>API, compara</td>
</tr>

<tr>
    <td><strong><a href="http://projects.ensembl.org/neandertal/">Neanderthal Genome Browser</a></strong></td>
    <td>Preliminary assembly of <i>Homo sapiens neanderthalensis</i>.</td>
    <td>Web front-end derived from Ensembl webcode, Ensembl schema databases.</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.vectorbase.org/">VectorBase</a></strong></td>
    <td>Bioinformatic Resource Centre for Invertebrate Vectors of Human Pathogens</td>
    <td>Customised Ensembl webcode, Ensembl Compara pipeline, Ensembl annotation pipeline</td>
</tr>


<tr>
    <td><strong><a href="http://www.gramene.org/">Gramene</a></strong></td>
    <td>A Comparative Mapping Resource for Grains</td>
    <td>Customised webcode, Ensembl Compara pipeline,  Ensembl database used for annotation, import of the Arabidopsis Ensembl database from NASC.</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://atensembl.arabidopsis.info/">AtEnsembl</a></strong></td>
    <td><i>Arabidopsis thaliana</i> genome browser</td>
    <td>Customised Ensembl pipeline, API, and webcode</td>
</tr>

<tr>
    <td><strong><a href="http://projects.ensembl.org/heroic/">HEROIC</a></strong></td>
    <td>Functional genomics data for Mouse (Mus musculus)</td>
    <td>Data viewable on Ensembl via DAS.</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.cadre-genomes.org.uk">CADRE</a></strong></td>
    <td>Central Aspergillus Data REpository</td>
    <td></td>
</tr>

<tr>
    <td><strong><a href="http://public-contigbrowser.sigenae.org:9090/">Sigenae</a></strong></td>
    <td>Sigenae EST ContigBrowser</td>
    <td></td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.sanger.ac.uk/genetics/CGP/cosmic/">COSMIC (Catalogue of Somatic Mutations in Cancer)</a></strong></td>
    <td>Web display of somatic sequence variant/mutation data</td>
    <td>Customised Ensembl drawing code</td>
</tr>

<tr>
    <td><strong><a href="http://www.sanger.ac.uk/perl/MVP/mvp/">MVP (Methylation Variable Positions)</a></strong></td>
    <td>Web display of MVPs analysed in the Human Epigenome Project.</td>
    <td>Customised Ensembl drawing code</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://pupasuite.bioinfo.cipf.es/">PupaSuite</a></strong></td>
    <td>Interactive web tool for the selection of optimal sets of SNP for 
        genotyping purposes, with prediction of functional effect and visualisation facilities.</td>
    <td>Web front end to a CGI that uses APIs.</td>
</tr>

<tr>
    <td><strong><a href="http://www.orthomam.univ-montp2.fr">OrthoMaM</a></strong></td>
    <td>A database of alignments and trees based on orthologous exons and CDSs for mammalian species. 
        </td>
    <td>1-to-1 orthologues extracted from Ensembl.</td>
</tr>

<tr class="bg2">
    <td><strong><a href="http://www.cmbi.ru.nl/phylopat/">Phylopat</a></strong></td>
    <td>Web tool to show the presence or absence of certain genes or proteins in a set of 
      eukaryote species. 
        </td>
    <td>Uses orthologue data extracted from EnsMart.</td>
</tr>

<tr>
    <td><strong>DeCIFR legume project</strong> (retired)</td>
    <td>Automated annotation and display of legume genomes</td>
    <td>Customised Ensembl pipeline, API, and webcode</td>
</tr>

<tr class="bg2">
    <td><strong>Otter</strong></td>
    <td>Database backend for interactive curation of annotation</td>
    <td>Otter is an extension of the Ensembl database schema</td>
</tr>

<tr>
    <td><strong><!-- 500 error<a href="http://www.chromhome.org/">-->Chromhome<!--</a>--></strong></td>
    <td>A database of chromosome map data, particularly chromosome painting data 
      between species.</td>
    <td>Ensembl karyotype drawing code.</td>
</tr>

</table>

<p class="center">
<img src="/img/empowered.png" alt="'Powered by Ensembl' logo" width="98" height="20" />
</p>

<p>If you are using Ensembl code in your project, you might like to <a href="/img/empowered_lg.png">download the 'empowered' logo</a> (high-resolution PNG, suitable for use in print).</p>

</body>
</html>
