<html>
<head>
<title>Variant Effect Predictor</title>
</head>

<body>

<h1>Variant Effect Predictor</h1>

About |
<a href="#web">Web version</a> |
<a href="vep_script.html">Perl script</a> |
<a href="vep_formats.html">Data formats</a> |
<a href="vep_faq.html">Frequently asked questions</a>

<h2 id="about"> About </h2>
<p>
Ensembl provides the facility to predict the functional consequences of variants
using the Variant Effect Predictor (VEP). There are three primary ways to use
the functionality of the VEP:
</p>

<ul>
	<li> through the web interface | <a href="#web">Documentation</a> | <a
href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations"
class="modal_link">Launch web interface</a></li>
	<li> using the standalone perl script | <a href="vep_script.html">Documentation</a> | <a href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/variant_effect_predictor.tar.gz?view=tar&root=ensembl&pathrev=branch-ensembl-66">Download latest version</a></li>
	<li> using Ensembl's perl API | <a href="/info/docs/api/variation/index.html">Documentation</a></li>
</ul>

<p>
The web version is suitable for users with small volumes of data or those who
prefer not to use command-line utilities. The script version is the most
flexible of the VEP, and allows users to process large volumes of data using
their own compute resources. The API is suitable for perl programmers looking to
incorporate features of the VEP into their own code.
</p>

<p>
The VEP was formerly known as the SNP Effect Predictor, and was published under
this name. Please reference the following publication:
</p>

<p>McLaren W, Pritchard B, Rios D, Chen Y, Flicek P, Cunningham F.<br />
<strong>Deriving the consequences of genomic variants with the Ensembl API and SNP Effect Predictor.</strong><br />
<em>BMC Bioinformatics</em>26(16):2069-70(2010)<br />
<a href="http://dx.doi.org/10.1093/bioinformatics/btq330"
rel="external">doi:10.1093/bioinformatics/btq330 </a>
</p>

<h2 id="web"> Web version </h2>

<p>
The web version of the VEP can be accessed via the <a
href="/tools.html">Tools</a> link at the top of each Ensembl web page, or via
the "Manage your data" link on any species-specific page.
</p>

<h3> Upload form </h3>

<p> When you reach the VEP web interface, you will be presented with a form to
enter <a href="vep_formats.html#input">your data</a>. Data can be uploaded in
one of three ways:</p>

<ul>
	<li><b>File upload</b> - click the "Choose file" button and locate the file on your system</li>
	<li><b>Paste file</b> - simply copy and paste the contents of your file into the large text box</li>
	<li><b>File URL</b> - point the VEP to a file hosted on a publically
	accessible address. This can be either a 'http://' or 'ftp://' address.</li>
</ul>

<p>
It is then possible to configure other options. The VEP will run fine with the
default options; click the blue "Next" button at the bottom of the panel to
continue when you are happy with the options.
</p>

<ul>
    <li><b>File format</b> - ensure you select the correct file format for your uploaded data</li>
	<li><b>Species</b> - ensure you have selected the correct species for your data!</li>
	<li><b>Get regulatory region consequences</b> - the VEP can check for
	overlaps with known regulatory features, and also check if a variation falls
	in a high information part of a transcription factor binding site.</li>
	<li><b>Type of consequences to display</b> - the VEP can output consequences
	as described by <a
	href="/info/docs/variation/predicted_data.html#consequences">Ensembl</a>,
	the <a href="http://www.sequenceontology.org/">Sequence Ontology (SO)</a> or
	the <a href="http://www.ncbi.nlm.nih.gov/projects/SNP/">NCBI</a>.</li>
	<li><b>Check for existing co-located variants</b> - in species with an
	Ensembl Variation database, the VEP can check for existing variants
	co-located with your input - the identifiers of these variants appear in the
	output</li>
	<li><b>Return results in coding regions only</b> - by selecting this
	checkbox the VEP will filter out any results that do not fall in a protein
	coding region of a transcript, for example those in the introns or upstream
	of a transcript.</li>
    <li><b>Show HGNC identifier for genes where available</b> - add the HGNC
    identifier for the overlapping gene to the Extra column of the output (the
    default output shows only the Ensembl Gene ID e.g. ENSG00000000345).</li>
	<li><b>Show Ensembl protein identifiers where available</b> - adds the
	Ensembl protein identifier for the transcript (e.g. ENSP00000411206) to the
	Extra column of the output.</li>
	<li><b>Show HGVS identifiers for variants where available</b> - adds <a
	href="http://www.hgvs.org/mutnomen/" target="_blank">HGVS</a> nomenclature
	based on Ensembl stable identifiers to the output. Coding and/or protein
	sequence names can be added where appropriate.</li>
</ul>

<p>
The following options are currently available for human only.
</p>

<ul>
    <li>For non-synonymous SNPs, the VEP can provide additional predictions on
    protein productst using the following external tools (all tools output the
    prediction term, score or both).
		<ul>
			<li><b>SIFT predictions</b> - <a href="http://sift.jcvi.org/"
			target="_blank">SIFT</a> predicts whether an amino acid substitution
			affects protein function based on sequence homology and the physical
			properties of amino acids.</li>
            <li><b>PolyPhen predictions</b> - <a
            href="http://genetics.bwh.harvard.edu/pph2/"
            target="_blank">PolyPhen</a> is a tool which predicts possible
            impact of an amino acid substitution on the structure and function
            of a human protein using straightforward physical and comparative
            considerations.</li>
			<li><b>Condel consensus predictions</b> - <a
			href="https://bg.upf.edu/forge/projects/condel/wiki/Condel">Condel</a>
			computes a weighed average of the scores (WAS) of several
			computational tools aimed at classifying missense mutations as
			likely deleterious or likely neutral. The VEP currently presents a
			Condel WAS from SIFT and PolyPhen.</li>
		</ul>
	</li>
</ul>

<h3> Select output format </h3>

<p>
After clicking "Next", you are then asked to select either HTML or Text output.
Both formats contain the same information:
</p>

<ul>
	<li><b>Text</b> format is useful if you wish to use the output as the input
	for any other tools.</li>
	<li><b>HTML</b> presents the same information as the Text format, but
	formatted for the web with links to genes, transcripts and locations in the
	Ensembl browser. Links are also provided to genes, transcripts and
	co-located variations. SIFT, PolyPhen and Condel predictions are coloured
	according to severity, with red representing high severity, green low
	severity and blue unknown.</li>
</ul>

<h3> Viewing your results in the browser </h3>

<p>
Any data uploaded via the VEP web tool can be viewed on the Ensembl location
view; to view your data, either click a link in the Location column of the HTML
output, or switch on the track on location view (click "Configure this page",
then "User attached data" on Region in detail view to see uploaded tracks).
</p>

<h3> Issues with larger datasets </h3>

<p> The web interface to the VEP has a hard limit of 750 variants in your
uploaded file. However, it is possible that the tool will not work with fewer
variants than this, depending on the content of your data and the features you
switch on. For example, a relatively small file (e.g. 100 variants) may fail to
return results if every variant in the file falls in a different gene and those
genes are spread across many chromosomes. Contrastingly, a file containing 500
variants may return results quickly if those variants all fall in just a few
genes. </p>

<p> To mitigate this issue, users should consider splitting up their input by
chromosome and uploading each chromosome's variants as a separate file. The
problem can also be solved by using the <a href="vep_script.html">VEP script</a>
- it is a command line tool, but not as hard to use as you might think! It also
offers many more features than the web interface and is a generally much more
powerful tool. </p>

</body>
</html>
