<html>
<head>
  <title>Gene Families</title>
</head>

<body>

<h1>Gene Families in Compara</h1>

<h2 id="Introduction">Introduction</h2>

<p>
Ensembl families are determined through clustering of all Ensembl proteins along with 
metazoan sequences from <a rel="external" href="http://www.ebi.ac.uk/uniprot/">UniProtKB</a>.  
It therefore provides a way of exploring orthologues and closely related homologues across a range of animal species.
</p>


<h2 id="EnsemblProteinFamilyMembers">Ensembl Protein Family Report</h2>

<dl>

<dt id="FamilyID">Family ID</dt>
<dd>
Please refer to the <a href="stable_ids.html">Family and GeneTree stable ID page</a> for more info on stable ID assignation.
</dd>

<dt id="ConsensusAnnotation">Consensus Annotation</dt>
<dd>
For each cluster obtained, a consensus annotation is automatically generated from the
<a rel="external" href="http://www.ebi.ac.uk/swissprot/">UniProt/Swiss-Prot</a> and
<a rel="external" href="http://www.ebi.ac.uk/trembl/">UniProt/TrEMBL</a> description lines of all
<a rel="external" href="http://www.ebi.ac.uk/uniprot/">UniProtKB</a> members using the following approach:
<br />
If the description covers less than 40% of UniProt members in the cluster, the family description 
is assigned <strong>'AMBIGUOUS'</strong>.  If the annotation confidence score, described below, is zero, 
<strong>'UNKNOWN'</strong> is assigned.  Be aware that <strong>'UNCHARACTERIZED'</strong> is a UniProt 
description for a protein, and does not reflect the score.
<br />
The <strong>annotation confidence score</strong> is the percentage of UniProtKB family members with this 
description, or part of it. Note that only family members with 'informative' UniProt descriptions are taken into account.</dd>

<dt id="PredictionMethod">Prediction Method</dt>
<dd>A brief summary of the protein family <a href="#Clustering">clustering algorithm</a> is given.</dd>

<dt id="MultipleAlignments">Multiple Alignments</dt>
<dd>Ensembl provides pre-calculated multiple sequence alignments of all members for each cluster.
<br />
If the Java runtime environment, <a rel="external" href="http://www.oracle.com/technetwork/java/javase/index.html">JRE</a>, 
is properly installed on your computer, then buttons will produce a new window with 
<strong>multiple alignments</strong> of family members displayed in
<a rel="external" href="http://www.jalview.org/">JalView</a>.
The first option includes Ensembl protein prediction from the current, as well as all other 
species supported by Ensembl. The second option also includes
<a rel="external" href="http://www.ebi.ac.uk/uniprot/">UniProtKB</a> members.
You can also export a text file with the alignments of all the family members - a wide range of 
formats is available from the control panel.
<br />
Alternatively, export alignments using the <a href="/info/docs/api/compara/index.html">Compara Perl API</a>.
</dd>

</dl>

<h2 id="Clustering">Clustering</h2>

<p>
The protein family database is generated by running the Markov Clustering
(<a rel="external" href="http://micans.org/mcl/">MCL</a>) algorithm
[<a href="#nb1" name="nh1">1</a>, <a href="#nb2" name="nh2">2</a>, <a href="#nb3" name="nh3">3</a>, 
<a href="#nb4" name="nh4">4</a>] as initially proposed by A.J. Enright, S. van Dongen and C.A. Ouzounis
[<a href="#nb5" name="nh5">5</a>].
Prior to the clustering process, an all-against-all BLASTP sequence similarity search is run 
on the super-set of all Ensembl protein predictions of all species, together with all metazoan sequences from
<a rel="external" href="http://www.ebi.ac.uk/swissprot/">UniProt/Swiss-Prot</a> and
<a rel="external" href="http://www.ebi.ac.uk/trembl/">UniProt/TrEMBL</a>, to establish similarities.
Using these similarities, protein family <strong>clusters</strong> are established running the
<a rel="external" href="http://micans.org/mcl/">MCL</a> algorithm.
</p>

<h2 id="References">References</h2>

<p>
<a href="#nh1" name="nb1"><strong>1</strong></a>.
Stijn van Dongen
<br />
<strong>Graph Clustering by Flow Simulation.</strong>
<br />
PhD thesis, University of Utrecht, May 2000
<br />
<a rel="external" href="http://www.library.uu.nl/digiarchief/dip/diss/1895620/inhoud.htm">Full text</a>
</p>

<p>
<a href="#nh2" name="nb2"><strong>2</strong></a>.
Stijn van Dongen
<br />
<strong>A cluster algorithm for graphs.</strong>
<br />
Technical Report INS-R0010,
National Research Institute for Mathematics and Computer Science in the Netherlands,
Amsterdam,
May 2000
<br />
<a rel="external" href="http://repository.cwi.nl/search/fullrecord.php?publnr=4463">Record 4463</a>
<!--<a rel="external" href="http://www.cwi.nl/ftp/CWIreports/INS/INS-R0010.ps.Z">PostScript file</a>-->
<a rel="external" href="http://persistent-identifier.nl/?identifier=urn:nbn:nl:ui:18-4463">urn:NBN:nl:ui:18-4463</a>
</p>

<p>
<a href="#nh3" name="nb3"><strong>3</strong></a>.
Stijn van Dongen
<br />
<strong>A stochastic uncoupling process for graphs.</strong>
<br />
Technical Report INS-R0011,
National Research Institute for Mathematics and Computer Science in the Netherlands,
Amsterdam,
May 2000
<br />
<a rel="external" href="http://repository.cwi.nl/search/fullrecord.php?publnr=4462">Record 4462</a><br />
<!--<a rel="external" href="http://www.cwi.nl/ftp/CWIreports/INS/INS-R0011.ps.Z">PostScript file</a>-->
<a rel="external" href="http://persistent-identifier.nl/?identifier=urn:nbn:nl:ui:18-4462">urn:NBN:nl:ui:18-4462</a>
</p>

<p>
<a href="#nh4" name="nb4"><strong>4</strong></a>.
Stijn van Dongen
<br />
<strong>Performance criteria for graph clustering and Markov cluster experiments.</strong>
<br />
Technical Report INS-R0012,
National Research Institute for Mathematics and Computer Science in the Netherlands,
Amsterdam,
May 2000
<br />
<a rel="external" href="http://repository.cwi.nl/search/fullrecord.php?publnr=4461">Record 4461</a><br />
<!--<a rel="external" href="http://www.cwi.nl/ftp/CWIreports/INS/INS-R0012.ps.Z">PostScript file</a>-->
<a rel="external" href="http://persistent-identifier.nl/?identifier=urn:nbn:nl:ui:18-4461">urn:NBN:nl:ui:18-4461</a>
</p>

<p>
<a href="#nh5" name="nb5"><strong>5</strong></a>.
Anton J. Enright,
Stijn van Dongen and
Christos A. Ouzounis
<br />
<strong>An efficient algorithm for large-scale detection of protein families.</strong>
<br />
Nucleic Acids Res. 2002 Apr 1;30,7,:1575-1584.
<br />
<a rel="external" href="http://nar.oxfordjournals.org/cgi/content/abstract/30/7/1575">Abstract</a>
<a rel="external" href="http://nar.oxfordjournals.org/cgi/content/full/30/7/1575">Full text</a>
</p>

</body>
</html>
