<h3>Annotation</h3>

<p class="space-below">
The Ensembl mouse automatic gene annotations were vastly improved 
in release 61 (1 February 2011) by using updated Ensembl genebuild
pipeline code and incorporating new data resources which have become 
available since the last NCBIM37 genebuild (April 2007). The new 
resources include an updated mouse-specific repeat library, 
additional <a href="http://www.ncbi.nlm.nih.gov/RefSeq">RefSeq</a> 
and <a href="http://www.uniprot.org/protein">Uniprot</a> 
protein sequence data for annotating the coding regions of 
protein-coding genes, as well as new cDNAs and ESTs for annotating 
untranslated regions (UTRs) of protein-coding genes. Extensive data 
quality checks have been performed to remove gene/transcript models 
with erroneous structures (e.g. interlocking transcripts with long 
introns on the same strand) or supported by dubious evidence (e.g. 
cDNA fragments with short, wrongly annotated open-reading frames).  
In release 62 (13 April 2011), the Ensembl automatic gene 
annotations were patched to correct gene/transcript models previously 
truncated due to the presence of selenocysteine residues (encoded by 
the UGA codon) in their translations.  Since release 62, the Ensembl 
automatic annotations have been frozen.

</p>

<p class="space-below">
As in all previous releases since October 2007, in Ensembl release 
64, we provide a combined Ensembl-Vega merged gene set.  Specifically, 
the frozen Ensembl annotations from April 2011 were merged with the 
latest <a href="http://Vega.sanger.ac.uk/Mus_musculus/Info/Index">Vega</a> 
manual annotations (as of 16 May 2011) at the transcript level. The 
Vega annotations were generated mainly by the HAVANA team at the 
<a href="http://www.sanger.ac.uk">Wellcome Trust Sanger Institute</a>. 
Transcripts from the two annotation sources were merged if they shared 
the same internal exon-intron boundaries (i.e. had identical splicing 
pattern) with slight differences in the terminal exons allowed. 
Importantly, all Vega source transcripts (regardless of merge status) 
were included in the final merged gene set.  In addition, the gene 
set was also updated in release 64 to maintain its consistency with 
the 22147 <a href="<http://www.ncbi.nlm.nih.gov/projects/CCDS/CcdsBrowse.cgi">CCDS</a> gene models at the time of production (19 May 2011).

</p>

<p class="space-below">
As a result, the release 64 gene set consists of 37681 genes and 95883 
transcripts. Of the 95883 transcripts, 18.00% (17257) were the result of 
merging Ensembl and Vega annotations, 21.65% (20756) originated from 
Ensembl, 54.18% (51953) originated from Vega, and a remaining ~6.1% were 
incorporated from other sources (e.g. immunoglobulin gene 
segments/transcripts imported from 
<a href="http://imgt.cines.fr/">IMGT</a>
data). Most of the non-merged Vega transcripts were alternative splice 
variants and/or non-coding transcripts complementing Ensembl 
annotations which focus on providing a conservative set of 
protein-coding genes/transcripts.
</p>

<ul>
<li><a href="/Mus_musculus/2011_09_mouse_genebuild.pdf">Detailed information on genebuild</a> (PDF)</li>
</ul>

<p class="space-below">
<a href="http://vega.sanger.ac.uk/Mus_musculus/"> <img src="/img/vega_small.gif" alt="Vega logo" style="width:83px;height:30px" title="Vega - Vertebrate Genome Annotation database" /></a>
Additional manual annotation of this genome can be found in
<a href="http://vega.sanger.ac.uk/Mus_musculus/">Vega</a></p>

<a href="http://www.heroic-ip.eu/"><img src="/img/credits/heroic_logo.jpg" alt="HEROIC" style="width:100px;height:26px;float:left;margin-right:5px" /></a>
<p>Additional functional genomics data produced by the HEROIC project (High-throughput Epigenetic 
Regulatory Organisation In Chromatin) is available to download from the 
<a href="http://projects.ensembl.org/heroic/">Ensembl Projects HEROIC portal</a>.</p>
