<HTML>

<head>
<meta name="order" content="2" />
  <meta name="navigation" content="Regulation">
  <title>Regulation</title>
</head>


<BODY LANG="en-GB" DIR="LTR">

[[INCLUDE::/info/docs/funcgen/doc_links.inc]]
[[INCLUDE::/info/docs/api/funcgen/doc_links.inc]]

<!-- <a href="/info/docs/funcgen/regulatory_build.html">Regulatory Build</a> |
<a href="/info/docs/funcgen/regulation_sources.html">Regulation Sources</a> |
<a href="/info/docs/funcgen/regulatory_segmentation.html">Regulatory Segmentation</a> |
<a href="/info/docs/microarray_probe_set_mapping.html">Microarray Probeset Mapping</a> |
<a href="/info/docs/api/funcgen/index.html">Perl API</a> |
<a href="/info/docs/api/funcgen/funcgen_schema.html">Database Description</a> 
<br /><br /><br />-->

<H1>Regulation Summary</H1>
<br>
<h2 id="regbuild">Regulatory Features</h2>
<p>
Ensembl stores regulatory features for human and mouse, inferred from experimental data through the 
<a href="/info/docs/funcgen/regulatory_build.html">Regulatory Build</a> process. It also provides the
experimental evidence supporting the regulatory features, which is derived from publicly
available <a href="/info/docs/funcgen/regulation_sources.html">data sources</a>. These data inlude:
<ul>
<li>Open Chromatin: Dnase1 Seq & <a href='http://www.ncbi.nlm.nih.gov/pubmed/19303047'>FAIRE</a>
<li>Transcription Factor ChIP-Seq
<li>Histone Modification  ChIP-Seq
<li>Transcription Factor Binding Site Annotations (<a href="regulatory_build.html#tfbs">PWMs</a>)
</ul>
</p>
<p>
The <a href="/info/docs/funcgen/regulatory_build.html">Regulatory Build</a> produces a special 'MultiCell' summary set of features, and cell specific feature sets with more detailed structures and classifications, such as: 
<ul>
<li>Promoter associated
<li>Gene associated
<li>Non-gene associated
<li>Polymerase III associated
<li>Unclassified
</ul>
Follow the link for more details about the <a href="regulatory_build.html#regannot">classification process</a>.
</p>

<br>
<p>
<h2 id="segmentation">Regulatory Segmentation</h2>
Genome segmentation analyses from the <a href='http://encodeproject.org'>ENCODE</a> project, providing genome wide summary of regulatory function/status. See <a href="/info/docs/funcgen/regulatory_segmentation.html">here</a> for more details. 
</p>
<br>

<h2 id="other">Other Regulatory Data</h2>
<p>
Ensembl Regulation databases also store data directly imported from external sources (identified as <i>'Other regulatory regions'</i>):
<ul>
<li>Micro RNA target predictions for human and mouse (a stringent subset of MiRanda described <a rel="external" href='http://www.ebi.ac.uk/enright-srv/microcosm/htdocs/targets/v5/consset.html'>here</a>).
<li>Experimentally validated human non-coding fragments from the <a rel="external" href='http://enhancer.lbl.gov/'>VISTA Enhancer Browser</a>.
<li>Genome-wide regulatory module and element prediction from <a rel="external" href='http://www.cisred.org'>cisRED</a> for human and mouse.
<li>Human DNA Methylation <a href='/info/data/ensembl_das.html'>DAS</a> tracks, including:
<ul><li>MeDIP-Chip methylation for 17 cell lines and tissues (<a rel="external" href='http://www.ncbi.nlm.nih.gov/pubmed?term=18577705'>Rakyan et al (2008)</a>)
<li>
RRBS (Reduced Representation Bisulfite Sequencing) for 44 cell lines (<a rel="external" href='http://www.genome.gov/10005107'>ENCODE</a>).<br>
<!-- remove this if there is a help item/legend --> 
Replicates were merged using mean percentage methylation, and a minimum 20x read coverage in the combined set.<br>
Methylation state is indicated by a color gradient. Dark blue indicates highly methylated areas, green intermediate and yellow which shows low methylation.
</ul>
</ul>
</p>
<br>

<h2 id="microarray">Microarray Probe Mappings</h2>
<p>
Ensembl stores <a href="/info/docs/funcgen/microarray_probe_set_mapping.html">microarray probe mappings</a> for several species and technologies, including:
<ul>
<li>Affymetrix: IVT & ST arrays
<li>Codelink: Whole genome arrays
<li>Agilent: Whole genome and CGH arrays <!-- add Sureprint here for 66-->
<li>Illumina: Whole genome and Infinium methylation arrays
<li>Phalanx
</ul>
</p>
<br>

<h2 id="displays">Regulation displays</h2>
<p>
Regulation data can be viewed in the browser through pages such as:
</p>

<ul>
<li>
<strong>Gene:</strong> Regulatory Regions surrounding a gene (e.g. <a href="/Homo_sapiens/Gene/Regulation?db=core;g=ENSG00000159197">for all regulatory features around KCNE2</a>).
Note that even if a regulatory region is near or within a gene, this does not mean that it is acting on that gene.
</li>

<li><strong>Location:</strong> Region in Detail (by default 'MultiCell' regulatory features are displayed.
Cell-specific annotations and evidence tracks can also be drawn using &quot;Configure this page&quot; at the left. 
The menu allows display of information in Ensembl databases along with external sources in DAS format such as 
Methylation data.)</li>
</ul>

<p>
Clicking on any regulatory feature on an Ensembl page will open a Regulation tab with information about the 
evidence supporting the selected regulatory feature, as well as cell-specific annotations.
You can choose different views in the Regulation tab:

<ul>
<li>
The <i>'Details by cell line'</i> view is the default view, where you get the generic 'MultiCell' regulatory feature 
and cell-specific classifications, as well as part of their supporting evidence displayed as tracks. To display more evidence tracks, 
you can go to the <i>'Configure this page'</i> link and select tracks within the <i>'Regulation'</i> subsection.
'DNA Methylation' and external sources ('Other regulatory regions') are not used as evidence for regulatory regions.
</li>
<li>
The <i>'Summary'</i> view will display the 'MultiCell' regulatory feature, along with it's core evidence and cell-specific annotations.
By default, this view does not display cell-specific evidence.
</li>
<li>
The <i>'Feature Context'</i> view displays the regulatory features in a wider context around the chosen regulatory feature. 
</li>
<li>
The complete list of supporting core evidence (TF binding and open chromatin) for a regulatory feature can be obtained 
via the <i>'Evidence'</i> view. 
</li>
</ul>
</p>

<h3>Display configuration</h3>

Configuration of the various regulation displays is available under the 'Regulation' section, with the following subsections:

<ul>
<li><b>Regulatory Features</b> These are the output of the Regulatory Build. You can specifiy regulatory feature tracks for individual cell lines</li>
<li><b>Open Chromatin and TFBS</b> Here you can configure the evidence tracks used in the Regulatory Build to build regulatory features. </li>
<li><b>Histones and Polymerases</b> Here you can configure the evidence tracks used in the Regulatory Build to establish cell-specific boundaries for regulatory features. </li>
<li><b>DNA Methylation</b> Here you can configure the methylation tracks you wish to visualise. </li>
<li><b>Other regulatory regions</b> Here you can configure the external tracks you wish to visualise. </li>
</ul>

<p>
When diplaying the regulatory feature evidence, there are two visualisations available:<br/>
<ul>
<li><b>Peaks</b> - Shows significantly enriched 'peak' regions</li>
<li><b>Signal</b> - Shows a plot or wiggle representing the raw signal.</li>
</ul>
</p>

<p> 
Species-specific microarray probe mappings can be visualized on the Location view by turning on tracks from the
<i>'Oligo Probes'</i> section in the configuration panel.
</p>

<h2 id="access">Data Access</h2>

<p>
Regulation information stored in Ensembl databases can be accessed through the website, but also using <a href="/info/data/biomart.html"
class="cp-external">BioMart</a> (regulation database), the <a href="/info/docs/api/index.html" class="cp-external">Perl API</a> 
(<a href="/info/docs/api/funcgen/index.html">regulation API</a>) or by directly accessing the databases. 
More information about programmatic access is available in the <a href="/info/docs/api/funcgen/regulation_tutorial.html">regulation API tutorial</a> page.
A more detailed description of the regulation databases can be seen <a href='/info/docs/api/funcgen/funcgen_schema.html'>here</a>.
</p>


<P><BR><BR>
</P>
</BODY>
</HTML>
