<html>
<head>
<title>Variant Effect Predictor</title>
</head>

<body>

<h1 id="top">Frequently asked questions</h1>

<a href="index.html#about">About</a> |
<a href="index.html#formats">Data formats</a> |
<a href="index.html#web">Web version</a> |
<a href="vep_script.html">Perl script</a> |
Frequently asked questions

<ul>
    <li><a href="#general">General questions</a></li>
    <li><a href="#web">Web VEP questions</a></li>
    <li><a href="#script">VEP script questions</a></li>
</ul>

<h2 id="general"> General questions </h2>

<p style="font-weight:bold;"> Q: Why don't I see any co-located variations when
using species X? </p>

<p> A: Ensembl only has variation databases for a subset of all Ensembl species
- see <a href="/info/docs/variation/index.html">this document</a> for
details.</p>
<p>&nbsp;</p>


<p style="font-weight:bold;"> Q: Why are some in/dels in my Pileup format file
being given incorrect coordinates and/or allele string in the output? </p>

<p> A: This is a known issue with the Pileup parser - it is under investigation
</p>
<p>&nbsp;</p>

<a href="#top">[Back to top]</a>

<h2 id="web"> Web VEP questions </h2>

<p style="font-weight:bold;"> Q: How do I access the web version of the Variant
Effect Predictor? </p>

<p> A: You can find the web VEP on the <a href="/tools.html">Tools</a> page, or
from any species-specific page by clicking the blue "Manage your data" link
under the left-hand menu, then "Variant Effect Predictor". </p>

<p>&nbsp;</p>


<p style="font-weight:bold;"> Q: I have selected a VCF file in the file upload
field, but nothing happens when I click the blue "Next" button. Why? </p>

<p> A: Ensure that you have selected VCF as the input file format, and that your
VCF file is formatted <a
href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40"
target="_blank">correctly</a>.</p>

<p>&nbsp;</p>


<p style="font-weight:bold;"> Q: I uploaded a file with 1000 variants, but some
of them seem to be missing from the output? </p>

<p> A: Due to a limitation in the servers underlying Ensembl, only 750 variants
in one file can be processed at once. Consider splitting your file into smaller
chunks, or using the <a href="vep_script.html">standalone perl script</a>.</p>

<p>&nbsp;</p>



<p style="font-weight:bold;"> Q: Why is the output I get for my input file
different when I use the web VEP and the VEP script? </p>

<p> A: Ensure that you are passing equivalent arguments to the script that you
are using in the web version. If you are sure this is still a problem, please
report it on the Ensembl developers mailing list, <a
href="mailto:dev@ensembl.org">dev@ensembl.org</a>.</p>

<p>&nbsp;</p>

<a href="#top">[Back to top]</a>

<h2 id="script"> VEP script questions </h2>

<p style="font-weight:bold;"> Q: Why do I see the following error? </p>

<pre class="code">
Could not connect to database homo_sapiens_core_63_37 as user anonymous using [DBI:mysql:database=homo_sapiens_core_63_37;host=ensembldb.ensembl.org;port=5306] as a locator:
Unknown MySQL server host 'ensembldb.ensembl.org' (2) at $HOME/src/ensembl/modules/Bio/EnsEMBL/DBSQL/DBConnection.pm line 290.

-------------------- EXCEPTION --------------------
MSG: Could not connect to database homo_sapiens_core_63_37 as user anonymous using [DBI:mysql:database=homo_sapiens_core_63_37;host=ensembldb.ensembl.org;port=5306] as a locator:
Unknown MySQL server host 'ensembldb.ensembl.org' (2)
</pre>

<p> A: By default the VEP script is configured to connect to the public MySQL
server at ensembldb.ensembl.org. Occasionally the server may break connection
with your script, which causes this error. This can happen when the server is
busy, or due to various network issues. Consider using a <a
href="vep_script.html#local">local copy of the database</a>, or the new <a
href="vep_script.html#cache">caching system</a>. </p>

<p>&nbsp;</p>

<a href="#top">[Back to top]</a>

</body>
</html>