<html>
<head>
<title>PETs from the FANTOM Consortium</title>
</head>
<body>
     
<h2>FANTOM Consortium: paired-end ditags of PolyA+ RNA</h2>
        
<h3 class="boxed">Mapping of transcript tags (ditags)</h3>

<h4>Description</h4>

<p>This track shows the start and end sequence tags of mRNA transcripts determined by paired-end ditag (PET) sequencing. PETs are composed of 14-21 bases from either end of a cDNA. See the <a href="#refs">References</a> section below for more details on PET sequencing. The track shows GSC (gene signature cloning) and GIS (gene identification signature) PETs.</p>

<p>GSC libraries:</p>
<ul>
  <li>ZZ11: Mouse Embryo (11, 14.5 and 17 days Embryo mixture)</li>
  <li>ZZ12: Mouse Thymus (6 and 7 days Neonate mixture)</li>
  <li>ZZ13: Mouse Brain (Adult Pitutary, Cerebrum, Cortex, Hippocampus)</li>
  <li>ZZ14: Mouse Testis (Adult)</li>
</ul>

<p>GIS libraries:</p>
<ul>
  <li>SME005: Mouse embryonic stem cell line E14</li>
  <li>SME006: E17.5 whole mouse embryo</li>
  <li>SMN001: Melan A2 mouse melanocyte cell line</li>
  <li>SMT001: B16F1 mouse melanoma cell line</li>
</ul>

<p>The 5' and 3' ends of the ditag features are represented as small boxes connected by a horizontal arrow indicating the strand/direction. Clicking on the feature brings up a menu with information about this ditag and a link to this more detailed description.</p>

<h4>Alignments</h4>

<p>The ditags were split in half and aligned to the unmasked genome
sequence using <a href="http://www.ebi.ac.uk/~guy/exonerate/">Exonerate</a>.
The hits were filtered using the following strategy:</p>
<ul>

  <li>score &gt;= 75/85, considering only higest scoring hits</li>

  <li>hits for both ends</li>

  <li>both ends on same chromosome, same strand and within 600 kb</li>

  <li>for GIS PETs: 5'/3' orientation has to remain correct</li>

  <li>if the tags align to more than 4 genomic locations, they
were discarded
as being unreliable</li>

  <li>sequences with non-(ATGC) characters were removed</li>

  <li>sequences with poly(ATGC) &gt; 5 were removed</li>

</ul>

<p>Identical sequences and hits to the same location (start &amp; end
of ditag) were collapsed into a non-redundant set.</p>

<h4>Credits</h4>

<p>The ditag libraries were downloaded from the <a href="ftp://fantom.gsc.riken.jp/FANTOM3/">FANTOM consortium
site</a> and processed using code of the <a href="http://www.ensembl.org">EnsEMBL</a> project.</p>


<h4 id="refs">References</h4>


<p>Carninci P. et al.
<br />
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=16141072&amp;query_hl=2&amp;itool=pubmed_docsum">The
transcriptional landscape of the mammalian genome.</a>.
<br />
<em>Science 309(5740)</em>, 1559-63 (2005).</p>


<p>Ng P, Wei CL, Sung WK, Chiu KP, Lipovich L, Ang CC, Gupta S,
Shahab A, Ridwan A, Wong CH, Liu ET, Ruan Y.
<br />
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15782207&amp;query_hl=1">Gene
identification signature (GIS) analysis for transcriptome
characterization and genome annotation</a>.
<br />
<em>Nat. Methods 2(2)</em>, 105-11 (2005).</p>


<p>Slater GS, Birney E
<br />
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=15713233&amp;dopt=Abstract&amp;holding=f1000,f1000m,igbwellib">Automated
generation of heuristics for biological sequence comparison</a>.
<br />
<em>BMC Bioinformatics 6(1)</em>, 31 (2005).</p>

</body>
</html>

