<p>The zebrafish Zv9 assembly was annotated using a modified
      Ensembl pipeline. Predictions from zebrafish proteins have been given
      priority over predictions from other non-mammalian vertebrate species. All
      Uniprot proteins were filtered to remove predictions ( PE levels 3 and
      above ). Aligned zebrafish cDNAs have been used to add UTR regions.8,374
      RNASeq models made from a range of zebrafish developmental stages and
      tissues were added into the gene build where they added a novel model or
      splice variant.Genes are named based on the alignment of their coding
      regions to known entries in public databases; ZFIN genes have priority in
      this process.</p>

    <p>The Ensembl annotations were then merged with Vega annotations at the
      transcript level. Transcripts were merged if they shared the same internal
      exon-intron boundaries (i.e. had identical splicing pattern) with slight
      differences in the terminal exons allowed. Importantly, all Vega source
      transcripts (regardless of merge status) were included in the final merged
      gene set.</p>

<ul>
<li><a href="/info/docs/genebuild/2012_04_zebrafish_zv9_genebuild.pdf">Detailed information on genebuild</a> (PDF)</li>
</ul>

<p><a href="http://vega.sanger.ac.uk/Danio_rerio/"> <img src="/img/vega_small.gif" alt="Vega logo" style="width:83px;height:30px" title="Vega - Vertebrate Genome Annotation database" /></a>
Additional manual annotation of this genome can be found in
<a href="http://vega.sanger.ac.uk/Danio_rerio/">Vega</a></p>
