<html>
<head>
<meta name="order" content="6" />
<title>How Ensembl uses DAS</title>
</head>

<body>
  
<h1>The Distributed Annotation System, DAS</h1>

<p>
Ensembl makes use of the <a href="/info/docs/das/index.html">DAS</a>
in two ways:
</p>

<ul>

<li>External data may be integrated into the website.</li>

<li>Ensembl data may be integrated into other applications via the
Ensembl DAS server.</li>

</ul>

<h2>Integrating external DAS data</h2>

<p>
The <a rel="external"
href="http://www.biodas.org/wiki/Main_Page">Distributed Annotation
System</a> (DAS) is a client-server system that has been initially
designed for the exchange of biological annotation information based
on a standard protocol.  DAS provides thus a standardised method to
serve custom annotation information and to integrate data sets for
display in other resources. Ensembl allows attachment and
configuration of external DAS sources to several Ensembl genome
browser displays:
</p>

<ul>
<li>Location tab &gt; Region overview page &gt; cytoview panel</li>
<li>Location tab &gt; Region in detail page &gt; contigviewtop panel</li>
<li>Location tab &gt; Region in detail page &gt; contigviewbottom panel</li>
<li>Gene tab &gt; Gene summary page &gt; gene_summary panel</li>
<li>Gene tab &gt; External data page</li>
<li>Transcript tab &gt; External data page</li>
</ul>

<p>
Several DAS resources provided by the
<a rel="external" href="http://www.ebi.ac.uk/">European Bioinformatics
Institute</a> and the <a rel="external"
href="http://www.sanger.ac.uk/">Wellcome Trust Sanger Institute</a>
are already available and pre-configured by default. To turn these on
and off:
</p>

<ol>
<li>Navigate to the appropriate tab and view. For example the <em>Location</em>
tab and <em>Region in detail</em> section.</li>
<li>Click the "Configure this page" link.</li>
<li>Choose a track to enable from the available categories.</li>
</ol>

<p>
Further annotation information resources can be integrated into
Ensembl by attaching a valid data source provided by a DAS annotation
server. To attach a DAS source:
</p>

<ol>
<li>Navigate to the appropriate tab and view.</li>
<li>Click on the "Add custom data to page" link at the left.</li>
<li>Click on "Attach DAS" at the left, and follow the instructions.</li>
<li>Save and close the window. The source will be automatically enabled.</li>
</ol>

<h3>Further details</h3>

<p>
  Whilst DAS was originally designed to exchange annotations of reference
  sequences in chromosome or clone coordinate systems, GeneDAS and ProteinDAS
  are extensions to the DAS protocol used to exchange gene and protein
  annotations independent of genomic location information. Currently, Ensembl
  supports annotations based on several different coordinate systems. Some
  coordinate systems allow annotations which are <em>positional</em> in nature
  (i.e. refer to a location within a sequence), whereas others are concerned
  with text-based <em>non-positional</em> annotations.
</p>

<p>
  The views that a particular DAS source may be displayed on depends on the data
  being served and the coordinate system of its annotations:
</p>

<ul>
<li>In genomic coordinate systems (e.g. chromosome, clone, contig), annotations are positional.</li>
<li>In gene coordinate systems (Ensembl, Entrez, HUGO or MGI), annotations are non-positional.</li>
<li>In protein coordinate systems (Ensembl, UniProt or IPI), annotations may be positional OR non-positional.</li>
</ul>

<p>Ensembl supports DAS on the following sections of the website:</p>

<ul>
<li>Location tab &gt; Region overview page &gt; cytoview panel (positional annotations)</li>
<li>Location tab &gt; Region in detail page &gt; contigviewtop panel (positional annotations)</li>
<li>Location tab &gt; Region in detail page &gt; contigviewbottom panel (positional annotations)</li>
<li>Gene tab &gt; Gene summary page &gt; gene_summary panel (positional annotations)</li>
<li>Gene tab &gt; External data page (non-positional annotations)</li>
<li>Transcript tab &gt; External data page (non-positional annotations)</li>
</ul>

<p>
  The extended DAS specification supported by Ensembl is described in
  <a href="http://www.biomedcentral.com/1471-2105/9/S8/S3">Jenkinson AM et al.,
  BMC Bioinformatics, 2008;9(Suppl 8):S3</a>.
</p>

<h2>The Ensembl DAS reference server</h2>

<p>
Ensembl provides a DAS reference server which gives access to
genomic sequences, the latest Ensembl gene predictions, and for
some species, karyotypes and ditags. A list of the sources
currently served from the Ensembl DAS reference server may be
found as XML documents at:
</p>

<ul>
<li><a href="http://www.ensembl.org/das/dsn">http://www.ensembl.org/das/dsn</a> 
(listing generated from the standard DAS data source name (DSN request)</li>
<li><a href="http://www.ensembl.org/das/sources">http://www.ensembl.org/das/sources</a>
(a listing with extended information)</li>
</ul>
    
<h3>Example requests</h3>

<p>DAS request URLs have a specific format:</p>
    
<blockquote>
  <var>protocol</var>://<var>site-prefix</var>/das/<var>data-source</var>/<var>command</var>?<var>arguments</var>
</blockquote>
    
<p>For example:</p>
    
<dl>
<dt><a href="http://www.ensembl.org/das/Homo_sapiens.GRCh37.transcript/features?segment=13:32889611,32973347">http://www.ensembl.org/das/Homo_sapiens.GRCh37.transcript/features?segment=13:32889611,32973347</a></dt>
<dd>Request all transcripts (exons really) in the region [32889611,32973347] on human chromosome 13 
(this is where the gene BRCA2 is located in the GRCh37 assembly).</dd>
<dt><a href="http://www.ensembl.org/das/Gallus_gallus.WASHUC2.reference/sequence?segment=1:1,1000">http://www.ensembl.org/das/Gallus_gallus.WASHUC2.reference/sequence?segment=1:1,1000</a></dt>
<dd>Request the first 1000&nbsp;bp of the first chicken chromosome.</dd>
</dl>

<p>For more information on the DAS protocol, see our main <a href="/info/docs/das/">DAS documentation</a>.</p>  


</body>
</html>
