<HTML>

<head>
<meta name="order" content="2" />
  <meta name="navigation" content="Regulation Sources">
  <title>Regulation Sources</title>
</head>


<BODY LANG="en-GB" DIR="LTR">


[[INCLUDE::/info/docs/funcgen/doc_links.inc]]
[[INCLUDE::/info/docs/api/funcgen/doc_links.inc]]


<H1>Regulation Sources</H1>

<ul>
<b><a href="#source">Source Data</a></b><br>
<b><a href="#tfgenehuman">Correspondence of Transcription Factors to ENSEMBL Genes and Jaspar Matrices (Human)</a></b><br>
<b><a href="#tfgenemouse">Correspondence of Transcription Factors to ENSEMBL Genes and Jaspar Matrices (Mouse)</a></b><br>
<b><a href="#humanregbuild">Human Regulatory Build Detailed Data</a></b><br>
<b><a href="#mouseregbuild">Mouse Regulatory Build Detailed Data</a></b><br>
<b><a href="#datarefs">References for Datasets</a></b><br>
</ul>

<H2 id='source' CLASS="western">Source Data</H2>
<P> Regulatory features are generated by using a variety of genome wide epigenomic data sets. The vast majority of features are derived from ChIP-seq data available from large scale sequencing projects. The experimental data for these can be visualised via the experiment view:

<ul>
<a href ="/Homo_sapiens/Experiment?db=core;ex=project-Roadmap%20Epigenomics-">Roadmap Epigenomics</a>(human)<br>
<a href ="/Homo_sapiens/Experiment?db=core;ex=project-ENCODE-">ENCODE</a> (human).<br>
<a href ="/Mus_musculus/Experiment?db=core;ex=project-ENCODE-">ENCODE</a> (mouse).<br>
</ul>
Publication reference for these data can be found at the in the <a href="#datarefs">References for Datasets</a> section.

<p>
To maintain a standardised peak calling methodology, we start our analyses with raw reads from each experiment. 
Reads (replicates are pooled) are aligned to the genome using the <a href='http://www.ncbi.nlm.nih.gov/pubmed/19451168'>BWA</a> algorithm (using default parameters).  All matches to mitochondria are filtered out to avoid alignment anomalies due to similarites with autosomal regions.
</p>
<p>Peak calling is performed using two algorithms SWEMBL (S. Wilder et al, in preparation) and  <a href="http://cmb.gis.a-star.edu.sg/ChIPSeq/paperCCAT.htm">CCAT</a> 
<a href="http://www.ncbi.nlm.nih.gov/pubmed?term=20371496">(Xu et al, 2010)</a>:: 

<ul>
<li>SWEMBL - Strict parameters (-f 150 -R 0.015) are used for most epxeriments. These were obtained using CTCF as a reference dataset. 
<li>SWEMBL - Less strict set (-f 150 -R 0.0025) is used for data that has a broader distribution of reads. This is generally used with DNAse1 datasets to enable a greater set of potential regulatory regions. 
<li>CCAT - Specialised for broad peak calling of features such as H3K36me3. Parameters were set for histone calling using a 1kb window.
</ul>

The ENCODE <a href="http://hgwdev.cse.ucsc.edu/cgi-bin/hgFileUi?db=hg19&g=wgEncodeMapability">
DAC Blacklist regions</a> are then used to filter the resultant peaks. The black list regions
are also stored in the core database as 'encode_excluded' MiscFeatures.

</P>

<H3 id='tfgenehuman'>Correspondence of Transcription Factors to ENSEMBL Genes and Jaspar Matrices (Human)</H3>
<table class="ss tint" style="width:50%;">
<tr><th>Transcription Factor</th><th class="center">Ensembl Gene</th><th class="center">Jaspar Matrix(ces)</th></tr>
<tr class="bg2"><td>ATF3</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000162772">ENSG00000162772</a></td><td></td></td></tr>
<tr ><td>Ap2alpha</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000137203">ENSG00000137203</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0003.1&rm=present&collection=CORE">MA0003.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0085.1&rm=present&collection=CORE">PB0085.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0189.1&rm=present&collection=CORE">PB0189.1</a> </td></td></tr>
<tr class="bg2"><td>Ap2gamma</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000087510">ENSG00000087510</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0087.1&rm=present&collection=CORE">PB0087.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0191.1&rm=present&collection=CORE">PB0191.1</a> </td></td></tr>
<tr ><td>BAF155</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000173473">ENSG00000173473</a></td><td></td></td></tr>
<tr class="bg2"><td>BAF170</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000139613">ENSG00000139613</a></td><td></td></td></tr>
<tr ><td>BATF</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000156127">ENSG00000156127</a></td><td></td></td></tr>
<tr class="bg2"><td>BCL11A</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000119866">ENSG00000119866</a></td><td></td></td></tr>
<tr ><td>BCL3</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000069399">ENSG00000069399</a></td><td></td></td></tr>
<tr class="bg2"><td>BHLHE40</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000134107">ENSG00000134107</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0007.1&rm=present&collection=CORE">PB0007.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0111.1&rm=present&collection=CORE">PB0111.1</a> </td></td></tr>
<tr ><td>Bdp1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000145734">ENSG00000145734</a></td><td></td></td></tr>
<tr class="bg2"><td>Brf1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000185024">ENSG00000185024</a></td><td></td></td></tr>
<tr ><td>Brf2</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000104221">ENSG00000104221</a></td><td></td></td></tr>
<tr class="bg2"><td>Brg1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000127616">ENSG00000127616</a></td><td></td></td></tr>
<tr ><td>CBP</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000005339">ENSG00000005339</a></td><td></td></td></tr>
<tr class="bg2"><td>CTCF</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000102974">ENSG00000102974</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0139.1&rm=present&collection=CORE">MA0139.1</a> </td></td></tr>
<tr ><td>Cfos</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000170345">ENSG00000170345</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0099.1&rm=present&collection=CORE">MA0099.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0099.2&rm=present&collection=CORE">MA0099.2</a> </td></td></tr>
<tr class="bg2"><td>Cjun</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000177606">ENSG00000177606</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0099.2&rm=present&collection=CORE">MA0099.2</a> </td></td></tr>
<tr ><td>Cmyc</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000136997">ENSG00000136997</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0059.1&rm=present&collection=CORE">MA0059.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0147.1&rm=present&collection=CORE">MA0147.1</a> </td></td></tr>
<tr class="bg2"><td>E2F1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000101412">ENSG00000101412</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0024.1&rm=present&collection=CORE">MA0024.1</a> </td></td></tr>
<tr ><td>E2F4</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000205250">ENSG00000205250</a></td><td></td></td></tr>
<tr class="bg2"><td>E2F6</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000169016">ENSG00000169016</a></td><td></td></td></tr>
<tr ><td>EBF</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000164330">ENSG00000164330</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0154.1&rm=present&collection=CORE">MA0154.1</a> </td></td></tr>
<tr class="bg2"><td>Egr1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000120738">ENSG00000120738</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0162.1&rm=present&collection=CORE">MA0162.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0010.1&rm=present&collection=CORE">PB0010.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0114.1&rm=present&collection=CORE">PB0114.1</a> </td></td></tr>
<tr ><td>FOSL2</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000075426">ENSG00000075426</a></td><td></td></td></tr>
<tr class="bg2"><td>GTF2B</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000137947">ENSG00000137947</a></td><td></td></td></tr>
<tr ><td>Gabp</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000154727">ENSG00000154727</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0062.1&rm=present&collection=CORE">MA0062.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0062.2&rm=present&collection=CORE">MA0062.2</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0020.1&rm=present&collection=CORE">PB0020.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0124.1&rm=present&collection=CORE">PB0124.1</a> </td></td></tr>
<tr class="bg2"><td>Gata1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000102145">ENSG00000102145</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0035.1&rm=present&collection=CORE">MA0035.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0035.2&rm=present&collection=CORE">MA0035.2</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0140.1&rm=present&collection=CORE">MA0140.1</a> </td></td></tr>
<tr ><td>Gata2</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000179348">ENSG00000179348</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0036.1&rm=present&collection=CORE">MA0036.1</a> </td></td></tr>
<tr class="bg2"><td>HDAC1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000116478">ENSG00000116478</a></td><td></td></td></tr>
<tr ><td>HDAC2</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000196591">ENSG00000196591</a></td><td></td></td></tr>
<tr class="bg2"><td>HDAC3</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000171720">ENSG00000171720</a></td><td></td></td></tr>
<tr ><td>HDAC6</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000094631">ENSG00000094631</a></td><td></td></td></tr>
<tr class="bg2"><td>HEY1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000164683">ENSG00000164683</a></td><td></td></td></tr>
<tr ><td>IRF4</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000137265">ENSG00000137265</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0034.1&rm=present&collection=CORE">PB0034.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0138.1&rm=present&collection=CORE">PB0138.1</a> </td></td></tr>
<tr class="bg2"><td>Ini1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000099956">ENSG00000099956</a></td><td></td></td></tr>
<tr ><td>Jund</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000130522">ENSG00000130522</a></td><td></td></td></tr>
<tr class="bg2"><td>Max</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000125952">ENSG00000125952</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0058.1&rm=present&collection=CORE">MA0058.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0043.1&rm=present&collection=CORE">PB0043.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0147.1&rm=present&collection=CORE">PB0147.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PL0007.1&rm=present&collection=CORE">PL0007.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PL0014.1&rm=present&collection=CORE">PL0014.1</a> </td></td></tr>
<tr ><td>MOF</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000103510">ENSG00000103510</a></td><td></td></td></tr>
<tr class="bg2"><td>NELFe</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000204356">ENSG00000204356</a></td><td></td></td></tr>
<tr ><td>NFKB</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000109320">ENSG00000109320</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0105.1&rm=present&collection=CORE">MA0105.1</a> </td></td></tr>
<tr class="bg2"><td>Nfe2</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000123405">ENSG00000123405</a></td><td></td></td></tr>
<tr ><td>Nfya</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000001167">ENSG00000001167</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0060.1&rm=present&collection=CORE">MA0060.1</a> </td></td></tr>
<tr class="bg2"><td>Nfyb</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000120837">ENSG00000120837</a></td><td></td></td></tr>
<tr ><td>Nrf1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000106459">ENSG00000106459</a></td><td></td></td></tr>
<tr class="bg2"><td>Nrsf</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000084093">ENSG00000084093</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0138.2&rm=present&collection=CORE">MA0138.2</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0138.1&rm=present&collection=CORE">MA0138.1</a> </td></td></tr>
<tr ><td>PCAF</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000114166">ENSG00000114166</a></td><td></td></td></tr>
<tr class="bg2"><td>POU2F2</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000028277">ENSG00000028277</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PH0144.1&rm=present&collection=CORE">PH0144.1</a> </td></td></tr>
<tr ><td>PU1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000066336">ENSG00000066336</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0080.1&rm=present&collection=CORE">MA0080.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0080.2&rm=present&collection=CORE">MA0080.2</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0058.1&rm=present&collection=CORE">PB0058.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0162.1&rm=present&collection=CORE">PB0162.1</a> </td></td></tr>
<tr class="bg2"><td>Pax5</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000196092">ENSG00000196092</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0014.1&rm=present&collection=CORE">MA0014.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0239.1&rm=present&collection=CORE">MA0239.1</a> </td></td></tr>
<tr ><td>Pbx3</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000167081">ENSG00000167081</a></td><td></td></td></tr>
<tr class="bg2"><td>RPC155</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000148606">ENSG00000148606</a></td><td></td></td></tr>
<tr ><td>RXRA</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000186350">ENSG00000186350</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0016.1&rm=present&collection=CORE">MA0016.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0057.1&rm=present&collection=CORE">PB0057.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0161.1&rm=present&collection=CORE">PB0161.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0065.2&rm=present&collection=CORE">MA0065.2</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0074.1&rm=present&collection=CORE">MA0074.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0115.1&rm=present&collection=CORE">MA0115.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0159.1&rm=present&collection=CORE">MA0159.1</a> </td></td></tr>
<tr class="bg2"><td>Rad21</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000164754">ENSG00000164754</a></td><td></td></td></tr>
<tr ><td>SETDB1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000143379">ENSG00000143379</a></td><td></td></td></tr>
<tr class="bg2"><td>SIX5</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000177045">ENSG00000177045</a></td><td></td></td></tr>
<tr ><td>SP1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000185591">ENSG00000185591</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0079.1&rm=present&collection=CORE">MA0079.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0079.2&rm=present&collection=CORE">MA0079.2</a> </td></td></tr>
<tr class="bg2"><td>SRebp1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000072310">ENSG00000072310</a></td><td></td></td></tr>
<tr ><td>SRebp2</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000198911">ENSG00000198911</a></td><td></td></td></tr>
<tr class="bg2"><td>Sin3Ak20</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000169375">ENSG00000169375</a></td><td></td></td></tr>
<tr ><td>Sirt6</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000077463">ENSG00000077463</a></td><td></td></td></tr>
<tr class="bg2"><td>Srf</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000112658">ENSG00000112658</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0083.1&rm=present&collection=CORE">MA0083.1</a>  <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0078.1&rm=present&collection=CORE">PB0078.1</a> </td></td></tr>
<tr ><td>TAF1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000147133">ENSG00000147133</a></td><td></td></td></tr>
<tr class="bg2"><td>Tip60</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000172977">ENSG00000172977</a></td><td></td></td></tr>
<tr ><td>TFIIIC-110</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000115207">ENSG00000115207</a></td><td></td></td></tr>
<tr class="bg2"><td>Tcf12</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000140262">ENSG00000140262</a></td><td></td></td></tr>
<tr ><td>Tr4</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000177463">ENSG00000177463</a></td><td></td></td></tr>
<tr class="bg2"><td>USF1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000158773">ENSG00000158773</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0093.1&rm=present&collection=CORE">MA0093.1</a> </td></td></tr>
<tr ><td>XRCC4</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000152422">ENSG00000152422</a></td><td></td></td></tr>
<tr class="bg2"><td>Yy1</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000100811">ENSG00000100811</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0095.1&rm=present&collection=CORE">MA0095.1</a> </td></td></tr>
<tr ><td>ZBTB33</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000177485">ENSG00000177485</a></td><td></td></td></tr>
<tr class="bg2"><td>ZNF274</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000171606">ENSG00000171606</a></td><td></td></td></tr>
<tr ><td>ZZZ3</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000036549">ENSG00000036549</a></td><td></td></td></tr>
<tr class="bg2"><td>Znf263</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000006194">ENSG00000006194</a></td><td></td></td></tr>
<tr ><td>p300</td><td><a href="/homo_sapiens/Gene/Summary?g=ENSG00000100393">ENSG00000100393</a></td><td></td></td></tr>

</table>

<H3 id='tfgenemouse'>Correspondence of Transcription Factors to ENSEMBL Genes and Jaspar Matrices (Mouse)</H3>
<table class="ss tint" style="width:50%;">
<tr><th>Transcription Factor</th><th class="center">Ensembl Gene</th><th class="center">Jaspar Matrix(ces)</th></tr>
<tr class="bg2"><td>cMyb</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000019982">ENSMUSG00000019982</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0100.1&rm=present&collection=CORE">MA0100.1</a> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0149.1&rm=present&collection=CORE">PB0149.1</a> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0045.1&rm=present&collection=CORE">PB0045.1</a></td></td></tr>
<tr ><td>CTCF</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000005698">ENSMUSG00000005698</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0139.1&rm=present&collection=CORE">MA0139.1</a> </td></td></tr>
<tr class="bg2"><td>Cmyc</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000022346">ENSMUSG00000022346</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0147.1&rm=present&collection=CORE">MA0147.1</a> </td></td></tr>
<tr ><td>E2F1</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000027490">ENSMUSG00000027490</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0024.1&rm=present&collection=CORE">MA0024.1</a> </td></td></tr>
<tr class="bg2"><td>Esrrb</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000021255">ENSMUSG00000021255</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0141.1&rm=present&collection=CORE">MA0141.1</a> </td></td></tr>
<tr ><td>Klf4</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000003032">ENSMUSG00000003032</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0039.2&rm=present&collection=CORE">MA0039.2</a> </td></td></tr>
<tr class="bg2"><td>Max</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000059436">ENSMUSG00000059436</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0147.1&rm=present&collection=CORE">PB0147.1</a> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PB0043.1&rm=present&collection=CORE">PB0043.1</a> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=PL0007.1&rm=present&collection=CORE">PL0007.1</a></td></td></tr>
<tr ><td>Nanog</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000012396">ENSMUSG00000012396</a></td><td></td></td></tr>
<tr class="bg2"><td>NELFe</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000024369">ENSMUSG00000024369</a></td><td></td></td></tr>
<tr ><td>nMyc</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000037169">ENSMUSG00000037169</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0104.2&rm=present&collection=CORE">MA0104.2</a> </td></td></tr>
<tr class="bg2"><td>Oct4</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000012396">ENSMUSG00000012396</a></td><td></td></td></tr>
<tr ><td>p300</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000055024">ENSMUSG00000055024</a></td><td></td></td></tr>
<tr class="bg2"><td>Rad21</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000022314">ENSMUSG00000022314</a></td><td></td></td></tr>
<tr ><td>Smad1</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000031681">ENSMUSG00000031681</a></td><td></td></td></tr>
<tr class="bg2"><td>Sox2</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000074637">ENSMUSG00000074637</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0143.1&rm=present&collection=CORE">MA0143.1</a> </td></td></tr>
<tr ><td>STAT3</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000004040">ENSMUSG00000004040</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0144.1&rm=present&collection=CORE">MA0144.1</a> </td></td></tr>
<tr class="bg2"><td>Suz12</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000017548">ENSMUSG00000017548</a></td><td></td></td></tr>
<tr ><td>Tcfcp2l1</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000026380">ENSMUSG00000026380</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0145.1&rm=present&collection=CORE">MA0145.1</a> </td></td></tr>
<tr class="bg2"><td>USF2</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000058239">ENSMUSG00000058239</a></td><td></td></td></tr>
<tr ><td>Zfx</td><td><a href="/mus_musculus/Gene/Summary?g=ENSMUSG00000079509">ENSMUSG00000079509</a></td><td> <a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0146.1&rm=present&collection=CORE">MA0146.1</a> </td></td></tr>
</table>

</br>
</br>
<p>
The current release comprises of the following datasets: 
</P>
<H3 id='humanregbuild'>Human Regulatory Build version 11</H3>

<table class="ss tint" style="width:50%;">

<tr><th colspan="3">CD4</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>CBP</td><td>ChIP-Seq</td><td>12</td></tr>
<tr><td>CTCF</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>HDAC1</td><td>ChIP-Seq</td><td>12</td></tr>
<tr><td>HDAC2</td><td>ChIP-Seq</td><td>12</td></tr>
<tr class="bg2"><td>HDAC3</td><td>ChIP-Seq</td><td>12</td></tr>
<tr><td>HDAC6</td><td>ChIP-Seq</td><td>12</td></tr>
<tr class="bg2"><td>MOF</td><td>ChIP-Seq</td><td>12</td></tr>
<tr><td>p300</td><td>ChIP-Seq</td><td>12</td></tr>
<tr class="bg2"><td>PCAF</td><td>ChIP-Seq</td><td>12</td></tr>
<tr><td>Tip60</td><td>ChIP-Seq</td><td>12</td></tr>

<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>H2AK5ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H2AK9ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H2AZ</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H2BK120ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H2BK12ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H2BK20ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H2BK5ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H2BK5me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3K14ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H3K18ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H3K23ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H3K27ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H3K27me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3K27me2</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3K36ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H3K36me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3K4ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H3K4me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3K4me2</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3K79me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3K79me2</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3K79me3</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3K9ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H3K9me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3K9me2</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3K9me3</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H3R2me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H3R2me2</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H4K12ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H4K16ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H4K20me1</td><td>ChIP-Seq</td><td>3</td></tr>
<tr ><td>H4K20me3</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>H4K5ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H4K8ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr class="bg2"><td>H4K91ac</td><td>ChIP-Seq</td><td>4</td></tr>
<tr ><td>H4R3me2</td><td>ChIP-Seq</td><td>3</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>3</td></tr>

<tr><th colspan="3">GM06990</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>5</td></tr>

<tr><th colspan="3">GM12878</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>BATF</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>BCL11A</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>BCL3</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>Cfos</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Cmyc</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>DNase1</td><td>Dnase-Seq</td><td>6</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr ><td>EBF</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Egr1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>FAIRE</td><td>FAIRE-Seq</td><td>6</td></tr>
<tr class="bg2"><td>Gabp</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>IRF4</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Jund</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Max</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>NFKB</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Nrsf</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>POU2F2</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>PU1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Pax5</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>Pbx3</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Rad21</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>SP1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Sin3Ak20</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>Srf</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>TAF1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>Tcf12</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Tr4</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>USF1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Yy1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>ZBTB33</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>ZZZ3</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>p300</td><td>ChIP-Seq</td><td>10</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>H3K27ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K4me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me2</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K9ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H4K20me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>PolIII</td><td>ChIP-Seq</td><td>9</td></tr>

<tr><th colspan="3">H1ESC</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>Cmyc</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>6</td></tr>
<tr ><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr class="bg2"><td>FAIRE</td><td>FAIRE-Seq</td><td>6</td></tr>
<tr ><td>Nrsf</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>TAF1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>H3K27ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K4me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me1</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K4me1</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K4me1</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K4me2</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K9ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K9ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K9ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K9ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K9me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K9me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K9me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H4K20me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>6</td></tr>

<tr><th colspan="3">HUVEC</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>Cjun</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Cmyc</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>6</td></tr>
<tr ><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr class="bg2"><td>FAIRE</td><td>FAIRE-Seq</td><td>6</td></tr>
<tr ><td>Max</td><td>ChIP-Seq</td><td>9</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>H3K27ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K4me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me2</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K9ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K9me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H4K20me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>9</td></tr>

<tr><th colspan="3">HeLa</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>Ap2alpha</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Ap2gamma</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>BAF155</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>BAF170</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Bdp1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Brf1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Brf2</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Brg1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>Cfos</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Cjun</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Cmyc</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>Cmyc</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>DNase1</td><td>Dnase-Seq</td><td>6</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr ><td>E2F1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>E2F4</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>E2F6</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>FAIRE</td><td>FAIRE-Seq</td><td>6</td></tr>
<tr ><td>Gabp</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Ini1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Jund</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Max</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Nrf1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>RPC155</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>TAF1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>TFIIIC-110</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Tr4</td><td>ChIP-Seq</td><td>9</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>9</td></tr>

<tr><th colspan="3">HepG2</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>BHLHE40</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>Cmyc</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>DNase1</td><td>Dnase-Seq</td><td>6</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr ><td>FAIRE</td><td>FAIRE-Seq</td><td>6</td></tr>
<tr class="bg2"><td>FOSL2</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>Gabp</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>HEY1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>Jund</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>RXRA</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>SRebp1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>SRebp2</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Sin3Ak20</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>USF1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>ZBTB33</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>p300</td><td>ChIP-Seq</td><td>10</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>H3K27ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K4me2</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>H3K9ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H4K20me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>9</td></tr>

<tr><th colspan="3">IMR90</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>11</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>H2AK5ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H2BK120ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H2BK12ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H2BK15ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H2BK20ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K14ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K18ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K23ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K27ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K4ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K4me1</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K4me2</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K56ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K79me1</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K79me2</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H3K9ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H3K9me3</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H4K20me1</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H4K5ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr ><td>H4K8ac</td><td>ChIP-Seq</td><td>11</td></tr>
<tr class="bg2"><td>H4K91ac</td><td>ChIP-Seq</td><td>11</td></tr>

<tr><th colspan="3">K562</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>ATF3</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Bdp1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Brf1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Brf2</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Brg1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>Cfos</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Cjun</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Cmyc</td><td>ChIP-Seq</td><td>6</td></tr>
<tr class="bg2"><td>Cmyc</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>DNase1</td><td>Dnase-Seq</td><td>6</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr ><td>Egr1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>FAIRE</td><td>FAIRE-Seq</td><td>6</td></tr>
<tr ><td>GTF2B</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Gabp</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>HEY1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Ini1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Jund</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Max</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>NELFe</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Nfe2</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Nfya</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Nfyb</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Nrsf</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>PU1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>Rad21</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>SIX5</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>SP1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>Sin3Ak20</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>Sirt6</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Srf</td><td>ChIP-Seq</td><td>10</td></tr>
<tr ><td>TAF1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>TFIIIC-110</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>USF1</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>XRCC4</td><td>ChIP-Seq</td><td>9</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>H3K27ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K4me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me2</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr ><td>H3K9ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K9me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H4K20me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>10</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>PolIII</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Znf263</td><td>ChIP-Seq</td><td>9</td></tr>

<tr><th colspan="3">K562b (no Regulatory Features built, but data is available).</th></tr>
<tr class="bg2"><td>E2F4</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>E2F6</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Gata1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Gata2</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>SETDB1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>Tr4</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Yy1</td><td>ChIP-Seq</td><td>9</td></tr>
<tr ><td>ZNF274</td><td>ChIP-Seq</td><td>9</td></tr>
<tr class="bg2"><td>Znf263</td><td>ChIP-Seq</td><td>9</td></tr>

<tr><th colspan="3">NHEK</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>6</td></tr>
<tr ><td>CTCF</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>6</td></tr>
<tr ><td>DNase1</td><td>Dnase-Seq</td><td>5</td></tr>
<tr class="bg2"><td>FAIRE</td><td>FAIRE-Seq</td><td>6</td></tr>
<tr><th>Attribute Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr ><td>H3K27ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K27me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K36me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>H3K4me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me2</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K4me3</td><td>ChIP-Seq</td><td>5</td></tr>
<tr class="bg2"><td>H3K9ac</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>H3K9me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr class="bg2"><td>H4K20me1</td><td>ChIP-Seq</td><td>7</td></tr>
<tr ><td>PolII</td><td>ChIP-Seq</td><td>7</td></tr>
</table>

<H3 id='mouseregbuild'>Mouse Regulatory Build version 7</H3>

<table class="ss tint" style="width:50%;"><tr><th colspan="3">ES</th></tr>

<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>13</td></tr>
<tr><td>CTCF</td><td>ChIP-Seq</td><td>14</td></tr>
<tr class="bg2"><td>Cmyc</td><td>ChIP-Seq</td><td>14</td></tr>
<tr><td>E2F1</td><td>ChIP-Seq</td><td>14</td></tr>
<tr class="bg2"><td>Esrrb</td><td>ChIP-Seq</td><td>14</td></tr>
<tr><td>Klf4</td><td>ChIP-Seq</td><td>14</td></tr>
<tr class="bg2"><td>Nanog</td><td>ChIP-Seq</td><td>14</td></tr>
<tr><td>Oct4</td><td>ChIP-Seq</td><td>14</td></tr>
<tr class="bg2"><td>STAT3</td><td>ChIP-Seq</td><td>14</td></tr>
<tr><td>Smad1</td><td>ChIP-Seq</td><td>14</td></tr>
<tr class="bg2"><td>Sox2</td><td>ChIP-Seq</td><td>14</td></tr>
<tr><td>Suz12</td><td>ChIP-Seq</td><td>14</td></tr>
<tr class="bg2"><td>Tcfcp2I1</td><td>ChIP-Seq</td><td>14</td></tr>
<tr><td>Zfx</td><td>ChIP-Seq</td><td>14</td></tr>
<tr class="bg2"><td>nMyc</td><td>ChIP-Seq</td><td>14</td></tr>
<tr><td>p300</td><td>ChIP-Seq</td><td>14</td></tr>

<tr><th>Attribute Sets</th><th>Data type</th><th>Reference</th></tr>
<tr class="bg2"><td>H3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H3K4me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr class="bg2"><td>H3K9me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H3K27me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr class="bg2"><td>H3K36me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H4K20me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr class="bg2"><td>PolII</td><td>ChIP-Seq</td><td>15</td></tr>

<tr><th colspan="3">ES Hybrid</th></tr>

<tr><th>Attribute Sets</th><th>Data type</th><th>Reference</th></tr>
<tr><td>H3K36me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H3K9me3</td><td>ChIP-Seq</td><td>15</td></tr>

<tr><th colspan="3">MEF</th></tr>

<tr><th>Attribute Sets</th><th>Data type</th><th>Reference</th></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H3K36me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H3K9me3</td><td>ChIP-Seq</td><td>15</td></tr>

<tr><th colspan="3">MEL</th></tr>
<tr><th>Focus Sets</th><th class="center">Data type</th><th class="center">Reference</th></tr>
<tr class="bg2"><td>DNase1</td><td>Dnase-Seq</td><td>16</td></tr>
<tr><td>cMyb</td><td>ChIP-Seq</td><td>17</td></tr>
<tr class="bg2"><td>CTCF</td><td>ChIP-Seq</td><td>17</td></tr>
<tr><td>Max</td><td>ChIP-Seq</td><td>17</td></tr>
<tr class="bg2"><td>NELFe</td><td>ChIP-Seq</td><td>17</td></tr>
<tr><td>p300</td><td>ChIP-Seq</td><td>17</td></tr>
<tr class="bg2"><td>Rad21</td><td>ChIP-Seq</td><td>17</td></tr>
<tr><td>USF2</td><td>ChIP-Seq</td><td>17</td></tr>

<tr><th colspan="3">NPC</th></tr>

<tr><th>Attribute Sets</th><th>Data type</th><th>Reference</th></tr>
<tr class="bg2"><td>H3K27me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H3K36me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr class="bg2"><td>H3K4me3</td><td>ChIP-Seq</td><td>15</td></tr>
<tr><td>H3K9me3</td><td>ChIP-Seq</td><td>15</td></tr>

</TABLE>

<h3 id='datarefs'>References for Datasets</h3>
<p>
1. Genome-wide identification of DNaseI hypersensitive sites was
performed by Greg Crawford and Terry Furey (Duke University) using a
whole genome DNase-sequencing protocol (Crawford et al., Genome
Research 2006).<br />DNase-sequencing was performed using the Illumina
(Solexa) sequencing by synthesis method from a DNase treated library
generated from the GM06990 cell line (Crawford and Furey,
unpublished).
</p>

<p>
2. Kim, T.H.; Abdullaev, Z.K.; Smith, A.D.; Ching, K.A.; Loukinov,
D.I.; Green, R.D.; Zhang, M.Q.; Lobanenkov, V.V. &amp; Ren, B.<br />
Analysis of the vertebrate insulator protein CTCF-binding sites in the
human genome.<br />Cell, 2007 , 128 , 1231-1245
</p>

<p>
3. A. Barski, S. Cuddapah, K. Cui, T.Y. Roh, D.E. Schones, Z. Wang,
G. Wei, I. Chepelev and K. Zhao, (2007). High-resolution profiling of
histone methylations in the human genome, Cell 129 (2007),
pp. 823-837. <a href="http://www.ncbi.nlm.nih.gov/pubmed?term=17512414">PMID: 17512414</a>
</p>

<p>
4. Wang Z, Zang C, Rosenfeld JA, Schones DE, Barski A, Cuddapah S, Cui
K, Roh TY, Peng W, Zhang MQ, Zhao K. Combinatorial patterns of histone
acetylations and methylations in the human genome. Nat Genet. 2008
Jul;40(7):897-903. Epub 2008 Jun 15. <a href="http://www.ncbi.nlm.nih.gov/pubmed?term=18552846">PMID: 18552846</a>
</p>


<p>
5. This data was produced as part of the ENCODE project and is used in accordance to their <a href='http://genome.ucsc.edu/ENCODE/terms.html'>data release policy</a>. These data were 
generated by the UW ENCODE group.
More information <A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?g=wgEncodeUwDnaseSeq&db=hg18'>here</A> and 
<A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?g=wgEncodeUwChIPSeq&db=hg18'>here</A>
</p>

<p>
6. This data was produced as part of the ENCODE project and is used in accordance to their <a href='http://genome.ucsc.edu/ENCODE/terms.html'>data release policy</a>. These data and 
annotations were created by a collaboration of multiple institutions.
More information <A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?g=wgEncodeChromatinMap&db=hg18'>here</A> 
</p>

<p>
7. This data was produced as part of the ENCODE project and is used in accordance to their <a href='http://genome.ucsc.edu/ENCODE/terms.html'>data release policy</a>. The ChIP-seq data were 
generated at the Broad Institute and in the Bradley E. Bernstein lab at the 
Massachusetts General Hospital/Harvard Medical School. 
More information <A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?g=wgEncodeBroadChipSeq&db=hg18'>here</A> 
</p>

<p>
8. Raha D, Wang Z, Moqtaderi Z, Wu L et al. Close association of RNA polymerase 
II and many transcription factors with Pol III genes. Proc Natl Acad Sci USA 
2010 Feb 23;107(8):3639-44. <a href="http://www.ncbi.nlm.nih.gov/pubmed?term=20139302">PMID: 20139302</a> 
</p>

<p>
9. This data was produced as part of the ENCODE project and is used in accordance to their <a href='http://genome.ucsc.edu/ENCODE/terms.html'>data release policy</a>. These data were generated 
and analyzed by the labs of Michael Snyder, Mark Gerstein and Sherman Weissman at 
Yale University; Peggy Farnham at UC Davis; and Kevin Struhl at Harvard.    
More information <A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?g=wgEncodeYaleChIPseq&db=hg18'>here</A> 
</p>

<p>
10. This data was produced as part of the ENCODE project and is used in accordance to their <a href='http://genome.ucsc.edu/ENCODE/terms.html'>data release policy</a>. These data were 
provided by the Myers Lab at the HudsonAlpha Institute for Biotechnology. 
More information <A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?g=wgEncodeHudsonalphaChipSeq&db=hg18'>here</A> 
</p>

<p>
11. This data was produced as part of the Epigenomics Roadmap and is used in accordance to their 
<a href='http://www.nida.nih.gov/about/roadmap/epigenomics/data_access_policies.html'>data release policy</a>. 
More information in here: <A HREF='http://nihroadmap.nih.gov/epigenomics/'>http://nihroadmap.nih.gov/epigenomics/</A> 
</p>

<p>
12. Wang Z, Zang C, Cui K, Schones DE, Barski A, Peng W, Zhao K.
Genome-wide mapping of HATs and HDACs reveals distinct functions in active and inactive genes.
Cell. 2009 Sep 4;138(5):1019-31. <a href="http://www.ncbi.nlm.nih.gov/pubmed?term=18555785">PMID: 18555785</a>
</p>

<p>
13. Dnase1-sequencing was produced as a collaboration between Ensembl,
David Adams (Wellcome Trust Sanger Institute), and Greg Crawford (Duke
University).
</p>

<p>
14. Chen X, Xu H, Yuan P, Fang F, Huss M, Vega VB, Wong E, Orlov YL, Zhang W, 
Jiang J, Loh YH, Yeo HC, Yeo ZX, Narang V, Govindarajan KR, Leong B, 
Shahab A, Ruan Y, Bourque G, Sung WK, Clarke ND, Wei CL, Ng HH.
Integration of external signaling pathways with the core transcriptional 
network in embryonic stem cells. Cell. 2008 Jun 13;133(6):1106-17. 
<a href="http://www.ncbi.nlm.nih.gov/pubmed?term=18555785">PMID: 18555785</a>
</p>

<p>
15. Mikkelsen TS, Ku M, Jaffe DB, Issac B, Lieberman E, Giannoukos G,
Alvarez P, Brockman W, Kim TK, Koche RP, Lee W, Mendenhall E,
O'Donovan A, Presser A, Russ C, Xie X, Meissner A, Wernig M, Jaenisch
R, Nusbaum C, Lander ES, Bernstein BE. Genome-wide maps of chromatin
state in pluripotent and lineage-committed cells. Nature. 2007 Aug
2;448(7153):548-9. <a href="http://www.ncbi.nlm.nih.gov/pubmed?term=17603471">PMID: 17603471</a>
</p>

<p>
16. This data was produced as part of the ENCODE project and is used in accordance to their <a href='http://genome.ucsc.edu/ENCODE/terms.html'>data release policy</a>. 
These data were generated by the UW ENCODE group. 
More information <A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm9&g=wgEncodeUwDnase'>here</A> 
</p>

<p>
17. This data was produced as part of the ENCODE project and is used in accordance to their <a href='http://genome.ucsc.edu/ENCODE/terms.html'>data release policy</a>. 
These data were generated and analyzed by the labs of Michael Snyder at Stanford University and Sherman Weissman at Yale University. 
More information <A HREF='http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm9&g=wgEncodeSydhTfbs'>here</A> 
</p>

<P><BR><BR>
</P>
</BODY>
</HTML>
