<html>
<head>
<meta name="order" content="1" />
<title>Gene Annotation</title>
</head>

<body>

<h1>The Ensembl Annotation Process</h1>

<h2>Protein-coding gene annotation</h2>

<p>
Protein-coding genes are automatically annotated using Ensembl's genebuild
pipeline. All transcripts are based on mRNA and proteins in public
scientific databases. 
</p>

<p>
The human gene set is used as the <a rel="external"
href="http://www.gencodegenes.org/">GENCODE</a> gene set. 
The human and mouse gene sets include all <a href="ccds.html">CCDS</a> transcripts.
</p>

<p>
See the <a href="genome_annotation.html">annotation article</a> for
more about the Ensembl genebuild pipeline, gene names and annotation.
</p>  

<p><a href="2x_genomes.html">Low-coverage genomes</a> are annotated
using a modified pipeline which attempts to locate genes across
multiple scaffolds.</p>


<h2>More genes</h2>

<p>
The Ensembl gene set also includes automatically-annotated pseudogenes 
and <a href="ncrna.html">non-coding RNAs</a>. 
For human and mouse, we include annotation from <a rel="external"
href="http://www.imgt.org/">IMGT</a> for <a href="ig_tcr.html">Ig genes</a>.
</p>

<p>EST-based genes are predicted and displayed on the website 
but are not included in the final gene set.</p>

<p>Paired-end Illumina RNASeq data have been used to generate transcript models 
for <a href="rnaseq_annotation.html">human and zebrafish</a>.</p>

<h2>Alternative Splicing</h2>

<p>
Ensembl includes automatically-annotated <a href="ase_annotation.html">Alternative splicing events</a> for model organisms.
</p>

</body>
</html>
