<HTML>

<head>
<meta name="order" content="2" />
  <meta name="navigation" content="Regulation Sources">
  <title>Regulation Sources</title>
</head>


<BODY LANG="en-GB" DIR="LTR">


[[INCLUDE::/info/docs/funcgen/doc_links.inc]]
[[INCLUDE::/info/docs/api/funcgen/doc_links.inc]]


<P> 
Regulatory features are generated by using a variety of genome wide epigenomic data sets. The vast majority of features are derived from ChIP-seq data available from large scale sequencing projects. Meta data and links to source data for these can be visualised via the experiment view:

<ul>
<a href ="/Homo_sapiens/Experiment?db=core;ex=project-Roadmap%20Epigenomics-">Roadmap Epigenomics</a>(human)<br>
<a href ="/Homo_sapiens/Experiment?db=core;ex=project-ENCODE-">ENCODE</a> (human).<br>
<a href ="/Mus_musculus/Experiment?db=core;ex=project-ENCODE-">ENCODE</a> (mouse).<br>
</ul>

Other published data, can also be visualised via the above links, by removing the project filters. The results table also shows mappings between transcription factors, Ensembl genes and Jaspar matrices which have been used in the regulatory build.

<p>
To maintain a standardised peak calling methodology, we start our analyses with raw reads from each experiment. 
Reads (replicates are pooled) are aligned to the genome using the <a href='http://www.ncbi.nlm.nih.gov/pubmed/19451168'>BWA</a> algorithm (using default parameters).  All matches to mitochondria are filtered out to avoid alignment anomalies due to similarites with autosomal regions.
</p>
<p>Peak calling is performed using two algorithms SWEMBL (S. Wilder et al, in preparation) and  <a href="http://cmb.gis.a-star.edu.sg/ChIPSeq/paperCCAT.htm">CCAT</a> 
<a href="http://www.ncbi.nlm.nih.gov/pubmed?term=20371496">(Xu et al, 2010)</a>:

<ul>
<li>SWEMBL - Strict parameters (-f 150 -R 0.015) are used for most experiments. These were obtained using CTCF as a reference dataset. 
<li>SWEMBL - Less strict settings (-f 150 -R 0.0025) are used for data that has a broader distribution of reads. This is generally used with DNAse1 datasets to enable a greater set of potential regulatory regions. 
<li>CCAT - Specialised for broad peak calling of features such as H3K36me3. Parameters were set for histone calling using a 1kb window.
</ul>

The ENCODE <a href="http://hgwdev.cse.ucsc.edu/cgi-bin/hgFileUi?db=hg19&g=wgEncodeMapability">
DAC Blacklist regions</a> are then used to filter the resultant peaks. The black list regions
are also stored in the core database as 'encode_excluded' MiscFeatures.

</P>
</BODY>
</HTML>
